// @synchro-charts/core/dist/esm/line-chart-viewport-change.entry.js
import { r as registerInstance, h } from './index-0c943849.js';

const VIEWPORT_GROUP = 'some-group';
const NARROW_VIEWPORT = {
    start: new Date(2000, 0, 0),
    end: new Date(2000, 0, 1),
    group: VIEWPORT_GROUP,
};
const WIDE_VIEWPORT = {
    start: new Date(2000, 0, 0),
    end: new Date(2001, 0, 0),
    group: VIEWPORT_GROUP,
};
const FIVE_MINUTE_VIEWPORT = {
    duration: '5m',
};
const THIRTY_MINUTE_VIEWPORT = {
    duration: '30m',
};
const LineChartViewportChange = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.viewport = NARROW_VIEWPORT;
        this.setViewPort = viewport => {
            this.viewport = viewport;
        };
    }
    render() {
        return (h("div", { class: "synchro-chart-tests" }, h("button", { id: "toggle-narrow-view-port", onClick: () => this.setViewPort(NARROW_VIEWPORT) }, "use narrow viewport"), h("button", { id: "toggle-wide-view-port", onClick: () => this.setViewPort(WIDE_VIEWPORT) }, "use wide viewport"), h("button", { id: "toggle-five-minute-view-port", onClick: () => this.setViewPort(FIVE_MINUTE_VIEWPORT) }, "use 5 minute viewport"), h("button", { id: "toggle-thirty-moinute-view-port", onClick: () => this.setViewPort(THIRTY_MINUTE_VIEWPORT) }, "use 30 minute viewport"), h("br", null), h("br", null), h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, h("sc-line-chart", { widgetId: "widget-id", dataStreams: [], viewport: this.viewport })), h("sc-webgl-context", null)));
    }
};

export { LineChartViewportChange as line_chart_viewport_change };
