// @synchro-charts/core/dist/esm/sc-straight-line-segment-colored.entry.js
import { r as registerInstance, h, g as getElement } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import './time-b732648c.js';
import './utils-96fe4147.js';
import './predicates-67461267.js';
import './number-a7331d73.js';
import './getDataPoints-6f8b4e89.js';
import './three.module-06da257a.js';
import { w as webGLRenderer } from './webglContext-da311040.js';
import './_commonjsHelpers-1d681858.js';
import './v4-9366210a.js';
import './index-2e17ff6c.js';
import './clipSpaceConversion-8273b9c5.js';
import { C as CHART_SIZE } from './chartSize-6ceb3800.js';
import './pointMesh-a5aff2b1.js';
import { c as chartScene } from './chartScene-5da77983.js';
import './common-a5fcad18.js';

// viewport boundaries
const X_MIN = new Date(2000, 0, 0);
const X_MAX = new Date(2000, 0, 1);
const Y_MIN = 0;
const Y_MAX = 100;
const WIDTH = X_MAX.getTime() - X_MIN.getTime();
const TEST_DATA_POINT_1 = {
    x: X_MIN.getTime() + WIDTH / 3,
    y: (Y_MIN + Y_MAX) / 2,
};
const TEST_DATA_POINT_2 = {
    x: X_MIN.getTime() + WIDTH * (2 / 3),
    y: (Y_MIN + Y_MAX) / 2,
};
const ScStraightLineSegment = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    componentDidLoad() {
        const container = this.el.querySelector('#test-container');
        const scene = chartScene({
            viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: Y_MIN,
                yMax: Y_MAX,
            },
            dataStreams: [
                {
                    id: 'test-stream',
                    name: 'test-stream-name',
                    color: 'purple',
                    data: [TEST_DATA_POINT_1, TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: DataType.NUMBER,
                },
            ],
            chartSize: CHART_SIZE,
            container,
            minBufferSize: 100,
            bufferFactor: 2,
            thresholdOptions: {
                showColor: false,
            },
            thresholds: [],
        });
        webGLRenderer.addChartScene({ manager: scene });
        const rect = container.getBoundingClientRect();
        webGLRenderer.setChartRect(scene.id, Object.assign({ density: 1 }, rect.toJSON()));
    }
    render() {
        return (h("sc-webgl-context", null, h("div", { id: "test-container", style: { width: `${CHART_SIZE.width}px`, height: `${CHART_SIZE.height}px` } })));
    }
    get el() { return getElement(this); }
};

export { ScStraightLineSegment as sc_straight_line_segment_colored };
