// @synchro-charts/core/dist/esm/sc-webgl-bar-chart-unsupported-data-types.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
import { Y as Y_VALUE_STRING } from './constants-9d36da8a.js';

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(2000, 0);
const X_MAX = new Date(2001, 3);
const LEFT_X = new Date(2000, 3).getTime();
const MIDDLE_X = new Date(2000, 6).getTime();
const RIGHT_X = new Date(2000, 9).getTime();
const ScWebglBarChartUnsupportedDataTypes = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    render() {
        return (h("div", { class: "synchro-chart-tests" }, h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test-name',
                    aggregates: {},
                    data: [
                        {
                            x: LEFT_X,
                            y: Y_VALUE_STRING,
                        },
                        {
                            x: MIDDLE_X,
                            y: Y_VALUE_STRING,
                        },
                        {
                            x: RIGHT_X,
                            y: Y_VALUE_STRING,
                        },
                    ],
                    resolution: MONTH_IN_MS,
                    dataType: DataType.STRING,
                },
            ], size: {
                width: 500,
                height: 500,
            }, widgetId: "widget-id", viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            } })), h("sc-webgl-context", null)));
    }
};

export { ScWebglBarChartUnsupportedDataTypes as sc_webgl_bar_chart_unsupported_data_types };
