// @synchro-charts/core/dist/esm/sc-webgl-chart-no-annotations.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { X as X_MIN, b as Y_MIN, d as X_MAX, c as Y_MAX } from './constants-9d36da8a.js';

const ScWebglChartNoAnnotations = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    render() {
        return (h("div", null, h("sc-line-chart", { widgetId: "widget-id", dataStreams: [], annotations: {
                x: [
                    {
                        value: X_MIN,
                        label: {
                            text: 'x label',
                            show: true,
                        },
                        showValue: true,
                        color: 'red',
                    },
                ],
                y: [
                    {
                        value: Y_MIN,
                        label: {
                            text: 'y label',
                            show: true,
                        },
                        showValue: true,
                        color: 'blue',
                    },
                ],
                show: false,
            }, size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }), h("sc-webgl-context", null)));
    }
};

export { ScWebglChartNoAnnotations as sc_webgl_chart_no_annotations };
