// @synchro-charts/core/dist/esm/sc-webgl-bar-chart-dynamic-data-streams.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
import { a as Y_VALUE } from './constants-9d36da8a.js';

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(2000, 0);
const X_MAX = new Date(2001, 3);
const LEFT_X = new Date(2000, 3).getTime();
const MIDDLE_X = new Date(2000, 6).getTime();
const RIGHT_X = new Date(2000, 9).getTime();
const ScWebglBarChartDynamicDataStreams = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.dataStreams = [];
        this.colorIndex = 0;
        this.colors = ['#264653', '#2a9d8f', '#e9c46a', '#f4a261', '#e76f51'];
        this.increaseColorIndex = () => {
            this.colorIndex += 1;
        };
        this.getColor = () => {
            // Modding the  will cycle through the colors array when the color index becomes greater than
            // the colors array length
            return this.colors[this.colorIndex % this.colors.length];
        };
        this.addStream = () => {
            const leftPoint = {
                x: LEFT_X,
                y: Y_VALUE,
            };
            const middlePoint = {
                x: MIDDLE_X,
                y: Y_VALUE,
            };
            const rightPoint = {
                x: RIGHT_X,
                y: Y_VALUE,
            };
            const streamId = `stream-${this.dataStreams.length + 1}`;
            this.dataStreams = [
                ...this.dataStreams,
                {
                    id: streamId,
                    color: this.getColor(),
                    name: `${streamId}-name`,
                    aggregates: {
                        [MONTH_IN_MS]: [leftPoint, middlePoint, rightPoint],
                    },
                    data: [],
                    resolution: MONTH_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ];
            this.increaseColorIndex();
        };
        this.removeStream = () => {
            this.dataStreams.pop();
            this.dataStreams = [...this.dataStreams];
            this.colorIndex -= 1;
        };
    }
    render() {
        return (h("div", { class: "synchro-chart-tests" }, h("button", { id: "add-stream", onClick: this.addStream }, "Add Stream"), h("button", { id: "remove-stream", onClick: this.removeStream }, "Remove Stream"), h("br", null), h("br", null), h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, h("sc-bar-chart", { dataStreams: this.dataStreams, size: {
                width: 500,
                height: 500,
            }, widgetId: "widget-id", viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            } })), h("sc-webgl-context", null)));
    }
};

export { ScWebglBarChartDynamicDataStreams as sc_webgl_bar_chart_dynamic_data_streams };
