// @synchro-charts/core/dist/esm/status-timeline-multiple-data-streams.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
import './utils-96fe4147.js';
import './predicates-67461267.js';
import './number-a7331d73.js';
import './dataFilters-2772c214.js';
import { T as TEST_DATA_POINT_STANDARD, b as Y_MIN, c as Y_MAX, X as X_MIN, d as X_MAX } from './constants-9d36da8a.js';

const StatusTimelineMultipleDataStreams = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    render() {
        return (h("div", null, h("sc-status-timeline", { alarms: { expires: MINUTE_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream 1',
                    aggregates: { [MINUTE_IN_MS]: [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 70 })] },
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
                {
                    id: 'test2',
                    color: 'blue',
                    name: 'test stream 2',
                    aggregates: { [MINUTE_IN_MS]: [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 170 })] },
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
                {
                    id: 'test3',
                    color: 'red',
                    name: 'test stream 3',
                    aggregates: { [MINUTE_IN_MS]: [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 60 })] },
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    }
};

export { StatusTimelineMultipleDataStreams as status_timeline_multiple_data_streams };
