// @synchro-charts/core/dist/collection/testing/test-routes/charts/sc-webgl-bar-chart/sc-webgl-bar-chart-standard.js
import { Component, h } from '@stencil/core';
import { MINUTE_IN_MS } from '../../../../utils/time';
import { TEST_DATA_POINT_STANDARD, Y_MAX, Y_MIN, X_MIN, X_MAX } from '../constants';
import { DataType } from '../../../..';
/**
 * Testing route for the webGL rendering without being fully coupled to the chart.
 *
 * Tests that a single point renders as a bar correctly
 */
export class ScWebglBarChartStandard {
    render() {
        return (h("div", null,
            h("sc-bar-chart", { dataStreams: [
                    {
                        id: 'test',
                        color: 'black',
                        name: 'test stream',
                        data: [],
                        aggregates: {
                            [MINUTE_IN_MS]: [TEST_DATA_POINT_STANDARD],
                        },
                        resolution: MINUTE_IN_MS,
                        dataType: DataType.NUMBER,
                    },
                ], widgetId: "test-id", size: {
                    width: 500,
                    height: 500,
                }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }),
            h("sc-webgl-context", null)));
    }
    static get is() { return "sc-webgl-bar-chart-standard"; }
}
