// @synchro-charts/core/dist/collection/testing/testing-ground/sc-box/sc-box.js
import { h, Component, Prop } from '@stencil/core';
export class ScBox {
    constructor() {
        this.size = { width: 0, height: 0 };
    }
    render() {
        const dataProps = this.someObject
            ? Object.keys(this.someObject).reduce((obj, key) => (Object.assign(Object.assign({}, obj), { [`data-${key}`]: this.someObject[key] })), {})
            : {};
        return (h("div", Object.assign({}, dataProps, { class: "box-container", style: {
                boxSizing: 'border-box',
                border: '1px solid black',
                backgroundColor: 'lightgrey',
                width: `${this.size.width}px`,
                height: `${this.size.height}px`,
            } }),
            h("ul", null,
                h("li", null,
                    this.size.width,
                    " width"),
                h("li", null,
                    this.size.height,
                    " height"))));
    }
    static get is() { return "sc-box"; }
    static get properties() { return {
        "size": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "MinimalSizeConfig",
                "resolved": "MinimalSizeConfig",
                "references": {
                    "MinimalSizeConfig": {
                        "location": "import",
                        "path": "../../../utils/dataTypes"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "{ width: 0, height: 0 }"
        },
        "someObject": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Object",
                "resolved": "Object | undefined",
                "references": {
                    "Object": {
                        "location": "global"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
}
