// @synchro-charts/core/dist/collection/components/charts/sc-scatter-chart/sc-scatter-chart.js
import { Component, Prop, h } from '@stencil/core';
import { chartScene, updateChartScene } from './chartScene';
import { DEFAULT_CHART_CONFIG } from '../sc-webgl-base-chart/chartDefaults';
import { DATA_ALIGNMENT } from '../common/constants';
import { validate } from '../../common/validator/validate';
import { DataType } from '../../../constants';
// The initial size of buffers. The larger this is, the more memory allocated up front per chart.
// The lower this number is, more likely that charts will have to re-initialize there buffers which is
// a slow operation (CPU bound).
const DEFAULT_MIN_BUFFER_SIZE = 1000;
const DEFAULT_BUFFER_FACTOR = 2;
const tooltip = (props) => (h("sc-tooltip", Object.assign({}, props, { visualizesAlarms: false, supportString: false, dataAlignment: DATA_ALIGNMENT.EITHER })));
export class ScScatterChart {
    constructor() {
        this.gestures = true;
        /** Status */
        this.isEditing = false;
        /** Memory Management */
        this.bufferFactor = DEFAULT_BUFFER_FACTOR;
        this.minBufferSize = DEFAULT_MIN_BUFFER_SIZE;
    }
    componentWillRender() {
        validate(this);
    }
    render() {
        return (h("sc-size-provider", { size: this.size, renderFunc: (rect) => (h("sc-webgl-base-chart", { supportedDataTypes: [DataType.NUMBER], axis: this.axis, gestures: this.gestures, configId: this.widgetId, legend: this.legend, annotations: this.annotations, trends: this.trends, updateChartScene: updateChartScene, createChartScene: chartScene, size: Object.assign(Object.assign(Object.assign({}, DEFAULT_CHART_CONFIG.size), this.size), rect), dataStreams: this.dataStreams, alarms: this.alarms, viewport: this.viewport, minBufferSize: this.minBufferSize, bufferFactor: this.bufferFactor, isEditing: this.isEditing, renderTooltip: tooltip, supportString: false, visualizesAlarms: false, messageOverrides: this.messageOverrides })) }));
    }
    static get is() { return "sc-scatter-chart"; }
    static get properties() { return {
        "viewport": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "MinimalViewPortConfig",
                "resolved": "MinimalLiveViewport | MinimalStaticViewport",
                "references": {
                    "MinimalViewPortConfig": {
                        "location": "import",
                        "path": "../../../utils/dataTypes"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Chart API"
            }
        },
        "movement": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "MovementConfig",
                "resolved": "MovementConfig | undefined",
                "references": {
                    "MovementConfig": {
                        "location": "import",
                        "path": "../common/types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "scale": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ScaleConfig",
                "resolved": "ScaleConfig | undefined",
                "references": {
                    "ScaleConfig": {
                        "location": "import",
                        "path": "../common/types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "layout": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "LayoutConfig",
                "resolved": "LayoutConfig | undefined",
                "references": {
                    "LayoutConfig": {
                        "location": "import",
                        "path": "../common/types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "legend": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "LegendConfig",
                "resolved": "LegendConfig | undefined",
                "references": {
                    "LegendConfig": {
                        "location": "import",
                        "path": "../common/types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "size": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "MinimalSizeConfig",
                "resolved": "MinimalSizeConfig | undefined",
                "references": {
                    "MinimalSizeConfig": {
                        "location": "import",
                        "path": "../../../utils/dataTypes"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "widgetId": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": true,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "widget-id",
            "reflect": false
        },
        "dataStreams": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "DataStream[]",
                "resolved": "DataStream<Primitive>[]",
                "references": {
                    "DataStream": {
                        "location": "import",
                        "path": "../../../utils/dataTypes"
                    }
                }
            },
            "required": true,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "alarms": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "AlarmsConfig",
                "resolved": "undefined | { expires?: number | undefined; }",
                "references": {
                    "AlarmsConfig": {
                        "location": "import",
                        "path": "../../../utils/dataTypes"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "gestures": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "gestures",
            "reflect": false,
            "defaultValue": "true"
        },
        "annotations": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Annotations",
                "resolved": "Annotations",
                "references": {
                    "Annotations": {
                        "location": "import",
                        "path": "../common/types"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "trends": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Trend[]",
                "resolved": "LinearRegression[]",
                "references": {
                    "Trend": {
                        "location": "import",
                        "path": "../common/trends/types"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "axis": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Axis.Options",
                "resolved": "Options | undefined",
                "references": {
                    "Axis": {
                        "location": "import",
                        "path": "../common/types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "messageOverrides": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "MessageOverrides",
                "resolved": "undefined | { liveTimeFrameValueLabel?: string | undefined; historicalTimeFrameValueLabel?: string | undefined; noDataStreamsPresentHeader?: string | undefined; noDataStreamsPresentSubHeader?: string | undefined; noDataPresentHeader?: string | undefined; noDataPresentSubHeader?: string | undefined; liveModeOnly?: string | undefined; unsupportedDataTypeHeader?: string | undefined; unsupportedDataTypeSubHeader?: string | undefined; supportedTypes?: string | undefined; }",
                "references": {
                    "MessageOverrides": {
                        "location": "import",
                        "path": "../../../utils/dataTypes"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "isEditing": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Status"
            },
            "attribute": "is-editing",
            "reflect": false,
            "defaultValue": "false"
        },
        "bufferFactor": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Memory Management"
            },
            "attribute": "buffer-factor",
            "reflect": false,
            "defaultValue": "DEFAULT_BUFFER_FACTOR"
        },
        "minBufferSize": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "min-buffer-size",
            "reflect": false,
            "defaultValue": "DEFAULT_MIN_BUFFER_SIZE"
        }
    }; }
}
