// @synchro-charts/core/dist/collection/testing/test-routes/charts/status-timeline/status-timeline-standard.js
import { Component, h } from '@stencil/core';
import { MINUTE_IN_MS } from '../../../../utils/time';
import { TEST_DATA_POINT_STANDARD, Y_MAX, Y_MIN, X_MIN, X_MAX } from '../constants';
import { DataType } from '../../../../utils/dataConstants';
export class StatusTimelineStandard {
    render() {
        return (h("div", null,
            h("sc-status-timeline", { alarms: { expires: MINUTE_IN_MS }, dataStreams: [
                    {
                        id: 'test',
                        color: 'black',
                        name: 'test stream',
                        data: [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 70 })],
                        resolution: 0,
                        dataType: DataType.NUMBER,
                    },
                ], widgetId: "test-id", size: {
                    width: 500,
                    height: 500,
                }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }),
            h("sc-webgl-context", null)));
    }
    static get is() { return "status-timeline-standard"; }
}
