// @synchro-charts/core/dist/esm-es5/chartDefaults-f5689a69.js
import { S as ScaleType, L as LEGEND_POSITION } from './constants-35570903.js';
var DEFAULT_CHART_CONFIG = {
    widgetId: 'fake-id',
    viewport: {
        start: new Date(1995, 0, 0, 0),
        end: new Date(2020, 1, 0, 0),
        yMin: 0,
        yMax: 10000,
    },
    // width is width - marginLeft - marginRight
    size: { width: 400 + 50 + 25, height: 350, marginLeft: 50, marginRight: 50, marginTop: 24, marginBottom: 30 },
    movement: {
        enableXScroll: true,
        enableYScroll: false,
        zoomMax: Infinity,
        zoomMin: 0.00001,
    },
    layout: {
        xGridVisible: false,
        yGridVisible: true,
        xTicksVisible: true,
        yTicksVisible: true,
    },
    scale: {
        xScaleType: ScaleType.TimeSeries,
        yScaleType: ScaleType.Linear,
        xScaleSide: 'bottom',
        yScaleSide: 'left',
    },
    dataStreams: [],
    legend: {
        position: LEGEND_POSITION.BOTTOM,
        width: 170,
    },
};
var DEFAULT_THRESHOLD_OPTIONS = {
    showColor: true,
};
var DEFAULT_THRESHOLD_OPTIONS_OFF = {
    showColor: false,
};
export { DEFAULT_CHART_CONFIG as D, DEFAULT_THRESHOLD_OPTIONS as a, DEFAULT_THRESHOLD_OPTIONS_OFF as b };
