// @synchro-charts/core/dist/esm-es5/line-chart-viewport-change.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
var VIEWPORT_GROUP = 'some-group';
var NARROW_VIEWPORT = {
    start: new Date(2000, 0, 0),
    end: new Date(2000, 0, 1),
    group: VIEWPORT_GROUP,
};
var WIDE_VIEWPORT = {
    start: new Date(2000, 0, 0),
    end: new Date(2001, 0, 0),
    group: VIEWPORT_GROUP,
};
var FIVE_MINUTE_VIEWPORT = {
    duration: '5m',
};
var THIRTY_MINUTE_VIEWPORT = {
    duration: '30m',
};
var LineChartViewportChange = /** @class */ (function () {
    function LineChartViewportChange(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.viewport = NARROW_VIEWPORT;
        this.setViewPort = function (viewport) {
            _this.viewport = viewport;
        };
    }
    LineChartViewportChange.prototype.render = function () {
        var _this = this;
        return (h("div", { class: "synchro-chart-tests" }, h("button", { id: "toggle-narrow-view-port", onClick: function () { return _this.setViewPort(NARROW_VIEWPORT); } }, "use narrow viewport"), h("button", { id: "toggle-wide-view-port", onClick: function () { return _this.setViewPort(WIDE_VIEWPORT); } }, "use wide viewport"), h("button", { id: "toggle-five-minute-view-port", onClick: function () { return _this.setViewPort(FIVE_MINUTE_VIEWPORT); } }, "use 5 minute viewport"), h("button", { id: "toggle-thirty-moinute-view-port", onClick: function () { return _this.setViewPort(THIRTY_MINUTE_VIEWPORT); } }, "use 30 minute viewport"), h("br", null), h("br", null), h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, h("sc-line-chart", { widgetId: "widget-id", dataStreams: [], viewport: this.viewport })), h("sc-webgl-context", null)));
    };
    return LineChartViewportChange;
}());
export { LineChartViewportChange as line_chart_viewport_change };
