// @synchro-charts/core/dist/esm-es5/line-chart-unsupported-data-types.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import { Y as Y_VALUE_STRING } from './constants-9d36da8a.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(1998, 0, 0);
var X_MAX = new Date(2000, 0, 1);
var WIDTH = X_MAX.getTime() - X_MIN.getTime();
var LineChartUnsupportedDataTypes = /** @class */ (function () {
    function LineChartUnsupportedDataTypes(hostRef) {
        registerInstance(this, hostRef);
    }
    LineChartUnsupportedDataTypes.prototype.render = function () {
        return (h("div", { class: "synchro-chart-tests" }, h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [
                        {
                            x: new Date(X_MIN.getTime() + WIDTH / 4).getTime(),
                            y: Y_VALUE_STRING,
                        },
                        {
                            x: new Date(X_MIN.getTime() + WIDTH / 4).getTime(),
                            y: Y_VALUE_STRING,
                        },
                    ],
                    resolution: 0,
                    dataType: DataType.STRING,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX }, bufferFactor: 1, minBufferSize: 1 }), h("sc-webgl-context", null))));
    };
    return LineChartUnsupportedDataTypes;
}());
export { LineChartUnsupportedDataTypes as line_chart_unsupported_data_types };
