// @synchro-charts/core/dist/esm-es5/sc-annotations-draggable-multi.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
var X_MIN = new Date(1998, 0, 0);
var X_MAX = new Date(2001, 0, 1);
var ANNOTATIONS = {
    y: [
        {
            id: 'annotation-1',
            color: 'red',
            isEditable: true,
            showValue: true,
            value: 1000,
            label: {
                text: 'annotation-1',
                show: true,
            },
        },
        {
            id: 'annotation-2',
            color: 'blue',
            isEditable: true,
            showValue: true,
            value: 0,
            label: {
                text: 'annotation-2',
                show: true,
            },
        },
    ],
};
var SIZE = {
    height: 340,
    width: 340,
};
var ScAnnotationsDraggableMulti = /** @class */ (function () {
    function ScAnnotationsDraggableMulti(hostRef) {
        registerInstance(this, hostRef);
    }
    ScAnnotationsDraggableMulti.prototype.render = function () {
        return (h("div", null, h("div", { style: { width: '500px', height: '500px' } }, h("sc-line-chart", { widgetId: "widget-id", dataStreams: [], annotations: ANNOTATIONS, size: SIZE, viewport: { start: X_MIN, end: X_MAX } })), h("div", { style: { width: '500px', height: '500px' } }, h("sc-line-chart", { widgetId: "widget-id-2", dataStreams: [], annotations: ANNOTATIONS, size: SIZE, viewport: { start: X_MIN, end: X_MAX } })), h("sc-webgl-context", null)));
    };
    return ScAnnotationsDraggableMulti;
}());
export { ScAnnotationsDraggableMulti as sc_annotations_draggable_multi };
