// @synchro-charts/core/dist/esm-es5/sc-box.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
var ScBox = /** @class */ (function () {
    function ScBox(hostRef) {
        registerInstance(this, hostRef);
        this.size = { width: 0, height: 0 };
    }
    ScBox.prototype.render = function () {
        var _this = this;
        var dataProps = this.someObject
            ? Object.keys(this.someObject).reduce(function (obj, key) {
                var _a;
                return (Object.assign(Object.assign({}, obj), (_a = {}, _a["data-" + key] = _this.someObject[key], _a)));
            }, {})
            : {};
        return (h("div", Object.assign({}, dataProps, { class: "box-container", style: {
                boxSizing: 'border-box',
                border: '1px solid black',
                backgroundColor: 'lightgrey',
                width: this.size.width + "px",
                height: this.size.height + "px",
            } }), h("ul", null, h("li", null, this.size.width, " width"), h("li", null, this.size.height, " height"))));
    };
    return ScBox;
}());
export { ScBox as sc_box };
