// @synchro-charts/core/dist/esm-es5/sc-grid_4.entry.js
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { r as registerInstance, h, g as getElement, e as createEvent } from './index-0c943849.js';
import { D as DATA_ALIGNMENT } from './constants-35570903.js';
import { S as StreamType } from './dataConstants-856cb1e9.js';
import './terms-d11f73d5.js';
import './index-b3f1c3a1.js';
import { p as parseDuration } from './time-b732648c.js';
import { g as getThresholds } from './utils-96fe4147.js';
import { a as isDefined, b as isMinimalStaticViewport } from './predicates-67461267.js';
import './number-a7331d73.js';
import './dataFilters-2772c214.js';
import { a as activePoints, b as breachedThreshold } from './breachedThreshold-500f9d58.js';
import './getDataPoints-6f8b4e89.js';
import './three.module-06da257a.js';
import { w as webGLRenderer } from './webglContext-da311040.js';
import './_commonjsHelpers-1d681858.js';
import { g as getDataStreamForEventing } from './common-a5fcad18.js';
import { l as lodash_throttle, D as DATE_RANGE_EMIT_EVENT_MS } from './constants-24b6dac6.js';
import { v as viewportStartDate, a as viewportEndDate, i as isInLiveMode } from './viewPort-8a84bb2e.js';
import { v as validate } from './validate-b6246900.js';
import { D as DEFAULT_MESSAGE_OVERRIDES } from './dataTypes-91d8b311.js';
import { V as Value } from './Value-9ca575b7.js';
import { t as tippy, T as TIPPY_SETTINGS } from './toolTipSettings-fa1c4ac6.js';
import { u as updateName } from './helper-8e4c02dd.js';
var scGridCss = "sc-grid .grid{--widget-min-width:190px;--widget-min-height:100px;display:grid;grid-template-columns:repeat(auto-fit, minmax(var(--widget-min-width), 1fr));grid-auto-rows:minmax(var(--widget-min-height), 1fr);grid-gap:10px;margin:10px}sc-grid .grid-wrapper{display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;-ms-flex-pack:center;justify-content:center;-ms-flex-line-pack:center;align-content:center}";
var ScGrid = /** @class */ (function () {
    function ScGrid(hostRef) {
        registerInstance(this, hostRef);
    }
    ScGrid.prototype.render = function () {
        return (h("div", { class: "grid-wrapper" }, h("div", { class: "grid" }, h("slot", null))));
    };
    return ScGrid;
}());
ScGrid.style = scGridCss;
var ScGridTooltip = /** @class */ (function () {
    function ScGridTooltip(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.displayToolTip = function () {
            var container = _this.el.querySelector('.tooltip-container');
            var tooltip = _this.el.querySelector('.cell-tooltip');
            if (tooltip != null && container != null) {
                tooltip.style.display = 'block';
                _this.tooltip = tippy(container, Object.assign(Object.assign({}, TIPPY_SETTINGS), { placement: 'left', content: tooltip }));
            }
        };
    }
    ScGridTooltip.prototype.componentDidLoad = function () {
        this.displayToolTip();
    };
    ScGridTooltip.prototype.disconnectedCallback = function () {
        if (this.tooltip) {
            this.tooltip.destroy();
        }
    };
    ScGridTooltip.prototype.render = function () {
        var thereIsSomeData = this.propertyPoint != null || this.alarmPoint != null;
        var color = this.breachedThreshold ? this.breachedThreshold.color : undefined;
        var displaysMoreThanTitle = thereIsSomeData && this.isEnabled;
        return (h("div", { class: "tooltip-container" }, h("div", { class: "cell-tooltip awsui-util-container awsui" }, h("div", { class: { 'awsui-util-container-header': true, 'awsui-util-mb-m': displaysMoreThanTitle } }, h("h3", null, this.title)), displaysMoreThanTitle && (h("div", null, h("div", { class: "awsui-util-spacing-v-s" }, this.propertyPoint && (h("div", null, h("div", { class: "awsui-util-label" }, "Latest value:"), h("div", null, h("strong", { style: { color: color } }, h(Value, { value: this.propertyPoint.y })), ' ', "at", ' ', new Date(this.propertyPoint.x).toLocaleString('en-US', {
            hour12: true,
            minute: 'numeric',
            hour: 'numeric',
            year: 'numeric',
            month: 'numeric',
            day: 'numeric',
        })))), this.alarmPoint && (h("div", null, h("div", { class: "awsui-util-label" }, "Status:"), h("div", null, h("strong", { style: { color: color } }, h(Value, { value: this.alarmPoint.y })), ' ', "since", ' ', new Date(this.alarmPoint.x).toLocaleString('en-US', {
            hour12: true,
            minute: 'numeric',
            hour: 'numeric',
            year: 'numeric',
            month: 'numeric',
            day: 'numeric',
        }), this.breachedThreshold && this.breachedThreshold.description && (h("div", null, "(", this.breachedThreshold.description, ")"))))))))), h("slot", null)));
    };
    Object.defineProperty(ScGridTooltip.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: false,
        configurable: true
    });
    return ScGridTooltip;
}());
/* eslint-disable max-len */
var QuestionMarkIcon = function () { return (h("svg", { width: "25px", height: "25px", viewBox: "0 0 30 30" }, h("defs", null, h("path", { d: "M14.8742857,24.7485714 C20.3277031,24.7485714 24.7485714,20.3277031 24.7485714,14.8742857 C24.7485714,9.4208683 20.3277031,5 14.8742857,5 C9.4208683,5 5,9.4208683 5,14.8742857 C5,20.3277031 9.4208683,24.7485714 14.8742857,24.7485714 Z M13.64,18.5771429 L16.1566605,18.5771429 L16.1566605,21.0457143 L13.64,21.0457143 L13.64,18.5771429 Z M10.7019512,12.6821943 C10.7134712,12.0601112 10.8200302,11.4898769 11.0216312,10.9714743 C11.2232322,10.4530717 11.5054693,10.0037962 11.8683512,9.62363429 C12.231233,9.24347238 12.6689886,8.94683535 13.1816312,8.73371429 C13.6942737,8.52059322 14.267388,8.41403429 14.9009912,8.41403429 C15.7189152,8.41403429 16.4014684,8.52635316 16.9486712,8.75099429 C17.4958739,8.97563541 17.9365095,9.25499262 18.2705912,9.58907429 C18.6046728,9.92315596 18.8437104,10.2831524 18.9877112,10.6690743 C19.1317119,11.0549962 19.2037112,11.4149926 19.2037112,11.7490743 C19.2037112,12.3020371 19.1317119,12.7570725 18.9877112,13.1141943 C18.8437104,13.4713161 18.6651522,13.776593 18.4520312,14.0300343 C18.2389101,14.2834756 18.0027525,14.4994734 17.7435512,14.6780343 C17.4843499,14.8565952 17.2395523,15.0351534 17.0091512,15.2137143 C16.77875,15.3922752 16.5742721,15.5967531 16.3957112,15.8271543 C16.2171503,16.0575554 16.1546517,16.3455526 16.1085714,16.6911543 L16.1085714,17.3477943 L13.64,17.3477943 L13.64,16.5701943 C13.64,15.9718011 13.7557976,15.6706595 13.9113184,15.3365778 C14.0668391,15.0024962 14.3480301,14.7068355 14.5553912,14.4706743 C14.7627522,14.2345131 14.98163,14.0300351 15.2120312,13.8572343 C15.4424323,13.6844334 15.6555502,13.5116351 15.8513912,13.3388343 C16.0472321,13.1660334 16.2056306,12.9759553 16.3265912,12.7685943 C16.4475518,12.5612332 16.5022712,12.3020358 16.4907512,11.9909943 C16.4907512,11.4610716 16.3611525,11.0693956 16.1019512,10.8159543 C15.8427499,10.562513 15.4827535,10.4357943 15.0219512,10.4357943 C14.7109096,10.4357943 14.4430723,10.4962737 14.2184312,10.6172343 C13.99379,10.7381949 13.8094719,10.8994733 13.6654712,11.1010743 C13.5214704,11.3026753 13.4149115,11.5388329 13.3457912,11.8095543 C13.2766708,12.0802756 13.2421112,12.3711527 13.2421112,12.6821943 L10.7019512,12.6821943 Z", id: "path-1" })), h("g", { fill: "none" }, h("rect", { x: "0", y: "0", width: "30", height: "30" }), h("mask", { id: "mask-2", fill: "white" }, h("use", { xlinkHref: "#path-1" })), h("g", { id: "Oval" }), h("g", { mask: "url(#mask-2)", fill: "#DEE0E2" }, h("rect", { x: "0", y: "0", width: "30", height: "30" }))))); };
var ScHelpTooltip = /** @class */ (function () {
    function ScHelpTooltip(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.displayToolTip = function () {
            var container = _this.el.querySelector('.help-icon');
            var tooltip = _this.el.querySelector('[role="tooltip"]');
            if (tooltip != null && container != null) {
                tooltip.style.display = 'block';
                _this.tooltip = tippy(container, Object.assign(Object.assign({}, TIPPY_SETTINGS), { content: tooltip }));
            }
        };
    }
    ScHelpTooltip.prototype.disconnectedCallback = function () {
        if (this.tooltip) {
            this.tooltip.destroy();
        }
    };
    ScHelpTooltip.prototype.render = function () {
        return (h("div", { class: "help-icon", tabIndex: -1, onMouseOver: this.displayToolTip, onFocus: this.displayToolTip }, h(QuestionMarkIcon, null), h("div", { role: "tooltip", class: "awsui-util-container awsui", style: { display: 'none' } }, h("div", { class: "awsui-util-spacing-v-s" }, h("p", null, this.message)))));
    };
    Object.defineProperty(ScHelpTooltip.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: false,
        configurable: true
    });
    return ScHelpTooltip;
}());
/**
 * Given a list of infos, return the ones that are to be visualized.
 *
 * This will remove any alarms that don't have an associated property info.
 */
var removePairedAlarms = function (streams) {
    var alarmInfos = streams.filter(function (_c) {
        var streamType = _c.streamType;
        return streamType === StreamType.ALARM;
    });
    var propertyInfos = streams.filter(function (_c) {
        var streamType = _c.streamType;
        return streamType !== StreamType.ALARM;
    });
    // If an alarm is not 'part' of any property info, it is a stray and can be visualized.
    var isStrayAlarm = function (_c) {
        var id = _c.id;
        return !propertyInfos.some(function (_c) {
            var _d = _c.associatedStreams, associatedStreams = _d === void 0 ? [] : _d;
            return associatedStreams.some(function (a) { return a.id === id; });
        });
    };
    var strayAlarmInfos = alarmInfos.filter(isStrayAlarm);
    var visualizedInfoIds = __spreadArrays(propertyInfos, strayAlarmInfos).map(function (_c) {
        var id = _c.id;
        return id;
    });
    // Want to maintain the original order, so we will filter out what isn't included from our original input.
    return streams.filter(function (_c) {
        var id = _c.id;
        return visualizedInfoIds.includes(id);
    }).filter(isDefined);
};
/**
 * Returns alarm/property pairs.
 *
 * For instance, if you have one property with 3 alarms associated with it, this will return you 3 pairs in total. One pair for each alarm.
 */
var streamPairs = function (dataStreams) {
    var primaryInfos = removePairedAlarms(dataStreams);
    return primaryInfos
        .map(function (stream) {
        if (stream.streamType === StreamType.ALARM) {
            // if it's an alarm an not removed, that means it's not pair, and has no associated property.
            return [{ alarm: stream }];
        }
        var hasNoAssociatedAlarms = stream.associatedStreams == null || !stream.associatedStreams.some(function (_c) {
            var type = _c.type;
            return type === StreamType.ALARM;
        });
        if (hasNoAssociatedAlarms) {
            // No alarms, just report back the property
            return [{ property: stream }];
        }
        // Return on cell info for each alarm associated
        return (stream.associatedStreams || [])
            .map(function (_c) {
            var associatedId = _c.id;
            return dataStreams.find(function (_c) {
                var id2 = _c.id;
                return id2 === associatedId;
            });
        })
            .filter(isDefined)
            .map(function (alarmInfo) { return ({
            property: stream,
            alarm: alarmInfo,
        }); });
    })
        .flat();
};
var scWidgetGridCss = "sc-widget-grid .help-icon-container{z-index:100;position:absolute;right:0;top:0}sc-widget-grid .container{display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;height:100%;overflow:auto;position:relative;-ms-overflow-style:none;scrollbar-width:none}sc-widget-grid .container::-webkit-scrollbar{display:none}";
var title = function (_c) {
    var alarm = _c.alarm, property = _c.property;
    if (property) {
        return property.detailedName || property.name;
    }
    if (alarm) {
        return alarm.detailedName || alarm.name;
    }
    return '';
};
var ScWidgetGrid = /** @class */ (function () {
    function ScWidgetGrid(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.dateRangeChange = createEvent(this, "dateRangeChange", 7);
        this.widgetUpdated = createEvent(this, "widgetUpdated", 7);
        this.collapseVertically = true;
        this.isEditing = false;
        this.messageOverrides = {};
        /** Widget data stream names */
        this.names = [];
        /** Active Viewport */
        this.start = viewportStartDate(this.viewport);
        this.end = viewportEndDate(this.viewport);
        this.duration = !isMinimalStaticViewport(this.viewport)
            ? parseDuration(this.viewport.duration)
            : undefined;
        this.onDateRangeChange = lodash_throttle(function (_c) {
            var start = _c[0], end = _c[1], from = _c[2];
            _this.dateRangeChange.emit([start, end, from]);
        }, DATE_RANGE_EMIT_EVENT_MS, {
            leading: true,
            trailing: true,
        });
        this.onUpdate = function (_c) {
            var start = _c.start, end = _c.end, duration = _c.duration;
            var hasViewPortChanged = viewportStartDate(_this.viewport).getTime() !== start.getTime() ||
                viewportEndDate(_this.viewport).getTime() !== end.getTime();
            var inLiveMode = isInLiveMode(_this.viewport);
            if (hasViewPortChanged && !inLiveMode) {
                _this.onDateRangeChange([start, end, _this.viewport.group]);
            }
            // Update active viewport
            _this.start = start;
            _this.end = end;
            _this.duration = duration;
        };
        /**
         * Emit the current widget configuration
         */
        this.emitUpdatedWidgetConfiguration = function (dataStreams) {
            var configUpdate = {
                movement: undefined,
                scale: undefined,
                layout: undefined,
                legend: undefined,
                annotations: _this.annotations,
                axis: undefined,
                widgetId: _this.widgetId,
                dataStreams: dataStreams ? getDataStreamForEventing(dataStreams) : _this.dataStreams,
            };
            _this.widgetUpdated.emit(configUpdate);
        };
        this.onChangeLabel = function (_c) {
            var streamId = _c.streamId, name = _c.name;
            _this.names = updateName(_this.names, name, streamId);
            _this.onWidgetUpdated();
        };
        this.getPoints = function () { return activePoints({
            viewport: {
                start: _this.start,
                end: _this.end,
            },
            dataStreams: _this.dataStreams,
            selectedDate: _this.end,
            allowMultipleDates: true,
            dataAlignment: DATA_ALIGNMENT.EITHER,
        }); };
        this.getBreachedThreshold = function (point, dataStream) { return breachedThreshold({
            value: point && point.y,
            date: isMinimalStaticViewport(_this.viewport) ? new Date(_this.viewport.end) : new Date(),
            dataStreams: _this.dataStreams,
            dataStream: dataStream,
            thresholds: getThresholds(_this.annotations),
        }); };
    }
    ScWidgetGrid.prototype.componentWillRender = function () {
        validate(this);
    };
    ScWidgetGrid.prototype.componentDidLoad = function () {
        webGLRenderer.addChartScene({
            manager: {
                id: this.widgetId,
                viewportGroup: this.viewport.group,
                updateViewPort: this.onUpdate,
            },
            duration: this.duration,
        });
    };
    ScWidgetGrid.prototype.onViewPortChange = function (newViewPort) {
        this.onUpdate(Object.assign(Object.assign({}, newViewPort), { duration: !isMinimalStaticViewport(newViewPort) ? parseDuration(newViewPort.duration) : undefined, start: viewportStartDate(this.viewport), end: viewportEndDate(this.viewport) }));
    };
    ScWidgetGrid.prototype.disconnectedCallback = function () {
        // necessary to make sure that the allocated memory is released, and nothing is incorrectly rendered.
        webGLRenderer.removeChartScene(this.widgetId);
    };
    /**
     * On Widget Updated - Persist `DataStreamInfo`
     *
     * Emits an event which persists the current `NameValue[]` state into the
     * data stream info.
     */
    ScWidgetGrid.prototype.onWidgetUpdated = function () {
        var _this = this;
        // Construct the config update with the new names specified.
        var updatedDataStreams = this.dataStreams.map(function (dataStream) {
            var nameValue = _this.names.find(function (_c) {
                var nameId = _c.id;
                return dataStream.id === nameId;
            });
            var name = nameValue != null ? nameValue.name : dataStream.name;
            return Object.assign(Object.assign({}, dataStream), { name: name });
        });
        this.emitUpdatedWidgetConfiguration(updatedDataStreams);
    };
    ScWidgetGrid.prototype.render = function () {
        var _this = this;
        var _a, _b;
        var isEnabled = this.duration != null;
        var points = this.getPoints();
        var pairs = streamPairs(this.dataStreams);
        var isMiniVersion = pairs.length > 1;
        return (h("div", { class: { tall: !this.collapseVertically } }, !isEnabled && (h("div", { class: "help-icon-container" }, h("sc-help-tooltip", { message: (_b = (_a = this.messageOverrides) === null || _a === void 0 ? void 0 : _a.liveModeOnly) !== null && _b !== void 0 ? _b : DEFAULT_MESSAGE_OVERRIDES.liveModeOnly }))), h("sc-grid", null, pairs.map(function (_c) {
            var alarm = _c.alarm, property = _c.property;
            var stream = alarm || property;
            if (stream == null) {
                return undefined;
            }
            var alarmPointWrapper = alarm && points.find(function (p) { return p.streamId === alarm.id; });
            var propertyPointWrapper = property && points.find(function (p) { return p.streamId === property.id; });
            var alarmPoint = alarmPointWrapper ? alarmPointWrapper.point : undefined;
            var propertyPoint = propertyPointWrapper ? propertyPointWrapper.point : undefined;
            var pointToEvaluateOn = alarmPoint || propertyPoint;
            var infoToEvaluateOn = alarm || property;
            var threshold = pointToEvaluateOn && infoToEvaluateOn && _this.getBreachedThreshold(pointToEvaluateOn, infoToEvaluateOn);
            var alarmStream = alarm && _this.dataStreams.find(function (s) { return s.id === alarm.id; });
            var primaryStream = alarm ? alarmStream : property;
            return (h("sc-grid-tooltip", { title: title({ alarm: alarm, property: property }), propertyPoint: propertyPoint, alarmPoint: alarmPoint, breachedThreshold: threshold, isEnabled: isEnabled }, _this.renderCell({
                isEnabled: isEnabled,
                trendStream: property,
                propertyStream: property,
                propertyPoint: propertyPoint,
                alarmStream: alarmStream,
                alarmPoint: alarmPoint,
                breachedThreshold: threshold,
                isEditing: _this.isEditing,
                viewport: { start: _this.start, end: _this.end },
                miniVersion: isMiniVersion,
                onChangeLabel: _this.onChangeLabel,
                messageOverrides: _this.messageOverrides,
                labelsConfig: _this.labelsConfig,
                icon: threshold ? threshold.icon : undefined,
                valueColor: threshold ? threshold.color : undefined,
                error: primaryStream ? primaryStream.error : undefined,
                isLoading: primaryStream ? primaryStream.isLoading || false : false,
                isRefreshing: primaryStream ? primaryStream.isRefreshing || false : false,
            })));
        }))));
    };
    Object.defineProperty(ScWidgetGrid, "watchers", {
        get: function () {
            return {
                "viewport": ["onViewPortChange"]
            };
        },
        enumerable: false,
        configurable: true
    });
    return ScWidgetGrid;
}());
ScWidgetGrid.style = scWidgetGridCss;
export { ScGrid as sc_grid, ScGridTooltip as sc_grid_tooltip, ScHelpTooltip as sc_help_tooltip, ScWidgetGrid as sc_widget_grid };
