// @synchro-charts/core/dist/esm-es5/sc-scatter-chart-unsupported-data-types.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
import { Y as Y_VALUE_STRING } from './constants-9d36da8a.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(1998, 0, 0);
var X_MAX = new Date(2001, 0, 1);
var TEST_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: Y_VALUE_STRING,
};
var TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: Y_VALUE_STRING,
};
var ScScatterChartUnsupportedDataTypes = /** @class */ (function () {
    function ScScatterChartUnsupportedDataTypes(hostRef) {
        registerInstance(this, hostRef);
    }
    ScScatterChartUnsupportedDataTypes.prototype.render = function () {
        var _a;
        return (h("div", null, h("sc-scatter-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    name: 'test stream',
                    color: 'black',
                    aggregates: (_a = {}, _a[MONTH_IN_MS] = [TEST_DATA_POINT, TEST_DATA_POINT_2], _a),
                    data: [],
                    resolution: MONTH_IN_MS,
                    dataType: DataType.STRING,
                },
            ], annotations: {}, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }), h("sc-webgl-context", null)));
    };
    return ScScatterChartUnsupportedDataTypes;
}());
export { ScScatterChartUnsupportedDataTypes as sc_scatter_chart_unsupported_data_types };
