// @synchro-charts/core/dist/esm-es5/sc-scatter-chart-threshold-no-coloration.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { C as COMPARISON_OPERATOR } from './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(1998, 0, 0);
var X_MAX = new Date(2001, 0, 1);
var TEST_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: 2000,
};
var TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: 4000,
};
var ScScatterChartThresholdNoColoration = /** @class */ (function () {
    function ScScatterChartThresholdNoColoration(hostRef) {
        registerInstance(this, hostRef);
    }
    ScScatterChartThresholdNoColoration.prototype.render = function () {
        var _a;
        return (h("div", null, h("sc-scatter-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    name: 'test stream',
                    color: 'black',
                    aggregates: (_a = {}, _a[MONTH_IN_MS] = [TEST_DATA_POINT, TEST_DATA_POINT_2], _a),
                    data: [],
                    resolution: MONTH_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], annotations: {
                y: [
                    {
                        value: 3000,
                        label: {
                            text: 'y label',
                            show: true,
                        },
                        showValue: true,
                        color: 'red',
                        comparisonOperator: COMPARISON_OPERATOR.GREATER_THAN_EQUAL,
                    },
                ],
                thresholdOptions: { showColor: false },
            }, size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }), h("sc-webgl-context", null)));
    };
    return ScScatterChartThresholdNoColoration;
}());
export { ScScatterChartThresholdNoColoration as sc_scatter_chart_threshold_no_coloration };
