// @synchro-charts/core/dist/esm-es5/sc-size-provider-standard.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
var DEFAULT_SIZE = {
    width: 500,
    height: 600,
};
var SHIFT_X_DIFF = 50;
var ScSizeProviderStandard = /** @class */ (function () {
    function ScSizeProviderStandard(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.marginLeft = 0;
        this.onShiftRight = function () {
            _this.marginLeft += SHIFT_X_DIFF;
        };
    }
    ScSizeProviderStandard.prototype.render = function () {
        return (h("div", { class: "synchro-chart-tests", style: { width: '2000px', height: '2000px' } }, h("div", null, h("button", { id: "shift-right", onClick: this.onShiftRight }, "Shift Right")), h("div", { id: "container", style: {
                marginLeft: this.marginLeft + "px",
                height: DEFAULT_SIZE.height + "px",
                width: DEFAULT_SIZE.width + "px",
            } }, h("sc-size-provider", { renderFunc: function (size) { return h("sc-box", { someObject: size, size: size }); } }))));
    };
    return ScSizeProviderStandard;
}());
export { ScSizeProviderStandard as sc_size_provider_standard };
