// @synchro-charts/core/dist/esm-es5/sc-scatter-chart-trend-line-with-legend.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { L as LEGEND_POSITION } from './constants-35570903.js';
import { D as DataType, T as TREND_TYPE } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { S as SECOND_IN_MS } from './time-b732648c.js';
// viewport boundaries
var X_MIN = new Date(2018, 0, 0);
var X_MAX = new Date(2020, 0, 0);
var Y_MIN = 0;
var Y_MAX = 5000;
// test data point dead center of the viewport
var TEST_DATA_POINTS = [
    {
        x: new Date(2018, 6, 0).getTime(),
        y: 1000,
    },
    {
        x: new Date(2019, 0, 0).getTime(),
        y: 4000,
    },
    {
        x: new Date(2019, 6, 0).getTime(),
        y: 3000,
    },
];
var ScScatterChartTrendLineWithLegend = /** @class */ (function () {
    function ScScatterChartTrendLineWithLegend(hostRef) {
        registerInstance(this, hostRef);
    }
    ScScatterChartTrendLineWithLegend.prototype.render = function () {
        var _a;
        return (h("div", null, h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    aggregates: (_a = {}, _a[SECOND_IN_MS] = TEST_DATA_POINTS, _a),
                    data: [],
                    resolution: SECOND_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: Y_MIN,
                yMax: Y_MAX,
            }, legend: {
                position: LEGEND_POSITION.BOTTOM,
                width: 300,
            }, trends: [
                {
                    dataStreamId: 'test',
                    type: TREND_TYPE.LINEAR,
                },
            ] }), h("sc-webgl-context", null)));
    };
    return ScScatterChartTrendLineWithLegend;
}());
export { ScScatterChartTrendLineWithLegend as sc_scatter_chart_trend_line_with_legend };
