// @synchro-charts/core/dist/esm-es5/sc-single-colored-bar.entry.js
import { r as registerInstance, h, g as getElement } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { D as DAY_IN_MS } from './time-b732648c.js';
import './utils-96fe4147.js';
import './predicates-67461267.js';
import './number-a7331d73.js';
import './getDataPoints-6f8b4e89.js';
import './three.module-06da257a.js';
import { w as webGLRenderer } from './webglContext-da311040.js';
import './_commonjsHelpers-1d681858.js';
import './v4-9366210a.js';
import './index-2e17ff6c.js';
import './clipSpaceConversion-8273b9c5.js';
import './getDistanceFromDuration-5c7da5d2.js';
import { C as CHART_SIZE } from './chartSize-6ceb3800.js';
import { c as chartScene } from './chartScene-028556fc.js';
// viewport boundaries
var X_MIN = new Date(2000, 0, 0);
var X_MAX = new Date(2000, 0, 1);
var Y_MIN = 0;
var Y_MAX = 100;
var WIDTH = X_MAX.getTime() - X_MIN.getTime();
var HEIGHT = Y_MAX - Y_MIN;
var TEST_DATA_POINT_1 = {
    x: new Date(X_MIN.getTime() + WIDTH / 3).getTime(),
    y: Y_MIN + HEIGHT / 2,
};
var ScSingleColoredBar = /** @class */ (function () {
    function ScSingleColoredBar(hostRef) {
        registerInstance(this, hostRef);
    }
    ScSingleColoredBar.prototype.componentDidLoad = function () {
        var _a;
        var container = this.el.querySelector('#test-container');
        var scene = chartScene({
            viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: Y_MIN,
                yMax: Y_MAX,
            },
            dataStreams: [
                {
                    id: 'test-stream',
                    name: 'test-stream-name',
                    color: 'red',
                    resolution: DAY_IN_MS,
                    data: [],
                    aggregates: (_a = {},
                        _a[DAY_IN_MS] = [TEST_DATA_POINT_1],
                        _a),
                    dataType: DataType.NUMBER,
                },
            ],
            container: container,
            chartSize: CHART_SIZE,
            minBufferSize: 100,
            bufferFactor: 2,
            thresholdOptions: {
                showColor: false,
            },
            thresholds: [],
        });
        webGLRenderer.addChartScene({ manager: scene });
        var rect = container.getBoundingClientRect();
        webGLRenderer.setChartRect(scene.id, Object.assign({ density: 1 }, rect.toJSON()));
    };
    ScSingleColoredBar.prototype.render = function () {
        return (h("sc-webgl-context", null, h("div", { id: "test-container", style: { width: CHART_SIZE.width + "px", height: CHART_SIZE.height + "px" } })));
    };
    Object.defineProperty(ScSingleColoredBar.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: false,
        configurable: true
    });
    return ScSingleColoredBar;
}());
export { ScSingleColoredBar as sc_single_colored_bar };
