// @synchro-charts/core/dist/esm-es5/sc-webgl-chart-large-viewport.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
var VIEWPORT = {
    yMin: 0,
    yMax: 5000,
    start: new Date(2000, 0, 0),
    end: new Date(2001, 0, 0),
};
// test data point dead center of the viewport
var TEST_DATA_POINT = {
    x: new Date((VIEWPORT.start.getTime() + VIEWPORT.end.getTime()) / 2).getTime(),
    y: (VIEWPORT.yMin + VIEWPORT.yMax) / 2,
};
var ScWebglChartLargeViewport = /** @class */ (function () {
    function ScWebglChartLargeViewport(hostRef) {
        registerInstance(this, hostRef);
    }
    ScWebglChartLargeViewport.prototype.render = function () {
        return (h("div", null, h("sc-line-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_DATA_POINT],
                    resolution: 0,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "widget-id", size: {
                height: 500,
                width: 500,
            }, viewport: VIEWPORT }), h("sc-webgl-context", null)));
    };
    return ScWebglChartLargeViewport;
}());
export { ScWebglChartLargeViewport as sc_webgl_chart_large_viewport };
