// @synchro-charts/core/dist/esm-es5/sc-webgl-bar-chart-dynamic-buffer.entry.js
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(1998, 0, 0);
var X_MAX = new Date(2000, 0, 1);
/**
 * Testing route for the webGL rendering without being fully coupled to the chart.
 *
 * Used to test the behavior of a bar chart when adding/removing data points
 */
var WIDTH = X_MAX.getTime() - X_MIN.getTime();
var ScWebglBarChartDynamicBuffer = /** @class */ (function () {
    function ScWebglBarChartDynamicBuffer(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.data = [];
        this.addDataPoint = function () {
            var dataPoint = {
                // To generate new bar with half of the distance.
                x: new Date(X_MIN.getTime() + WIDTH / (2 + _this.data.length)).getTime(),
                y: 2500,
            };
            _this.data = __spreadArrays([dataPoint], _this.data);
        };
    }
    ScWebglBarChartDynamicBuffer.prototype.render = function () {
        var _a;
        return (h("div", { class: "synchro-chart-tests" }, h("button", { id: "add-data-point", onClick: this.addDataPoint }, "Add Data Point"), h("div", { id: "chart-container", style: { height: '500px', width: '500px', marginTop: '20px' } }, h("sc-bar-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'red',
                    name: 'test stream',
                    aggregates: (_a = {},
                        _a[MONTH_IN_MS] = this.data,
                        _a),
                    data: [],
                    resolution: MONTH_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            }, bufferFactor: 1, minBufferSize: 1 }), h("sc-webgl-context", null))));
    };
    return ScWebglBarChartDynamicBuffer;
}());
export { ScWebglBarChartDynamicBuffer as sc_webgl_bar_chart_dynamic_buffer };
