// @synchro-charts/core/dist/esm-es5/sc-webgl-bar-chart-start-from-zero.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
var X_MIN = new Date(2000, 0, 0, 0, 0);
var X_MAX = new Date(2000, 0, 0, 0, 10);
var ScWebglBarChartStartFromZero = /** @class */ (function () {
    function ScWebglBarChartStartFromZero(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.testData = [
            {
                x: new Date((X_MIN.getTime() + X_MAX.getTime()) / 2).getTime(),
                y: 4500,
            },
        ];
        this.changeDataDirection = function () {
            _this.testData = _this.testData.map(function (_a) {
                var x = _a.x, y = _a.y;
                return {
                    x: x,
                    y: -y,
                };
            });
        };
    }
    ScWebglBarChartStartFromZero.prototype.render = function () {
        var _a;
        return (h("div", { class: "synchro-chart-tests" }, h("button", { id: "change-data-direction", onClick: this.changeDataDirection }, "Change Data Direction"), h("br", null), h("br", null), h("div", { id: "chart-container", style: { width: '500px', height: '500px' } }, h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'purple',
                    name: 'test stream',
                    data: [],
                    aggregates: (_a = {},
                        _a[MINUTE_IN_MS] = this.testData,
                        _a),
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null))));
    };
    return ScWebglBarChartStartFromZero;
}());
export { ScWebglBarChartStartFromZero as sc_webgl_bar_chart_start_from_zero };
