// @synchro-charts/core/dist/esm-es5/sc-webgl-bar-chart-positive-negative.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
// viewport boundaries
var Y_MIN = -100;
var Y_MAX = 100;
var X_MIN = new Date(2000, 0, 0, 0, 0);
var X_MAX = new Date(2000, 0, 0, 0, 10);
// test data point dead center of the viewport
var TEST_DATA_POINT = {
    x: new Date(2000, 0, 0, 0, 3).getTime(),
    y: 50,
};
var TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0, 0, 7).getTime(),
    y: -50,
};
var ScWebglBarChartPositiveNegative = /** @class */ (function () {
    function ScWebglBarChartPositiveNegative(hostRef) {
        registerInstance(this, hostRef);
    }
    ScWebglBarChartPositiveNegative.prototype.render = function () {
        var _a;
        return (h("div", { id: "chart-container", style: { width: '500px', height: '500px' } }, h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    aggregates: (_a = {}, _a[MINUTE_IN_MS] = [TEST_DATA_POINT, TEST_DATA_POINT_2], _a),
                    data: [],
                    resolution: MINUTE_IN_MS,
                    color: 'black',
                    name: 'test stream',
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "widget-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return ScWebglBarChartPositiveNegative;
}());
export { ScWebglBarChartPositiveNegative as sc_webgl_bar_chart_positive_negative };
