// @synchro-charts/core/dist/esm-es5/sc-webgl-bar-chart-negative.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
// viewport boundaries
var Y_MIN = -3000;
var Y_MAX = 1000;
var X_MIN = new Date(2000, 0, 0, 0, 0);
var X_MAX = new Date(2000, 0, 0, 0, 10);
// test data point dead center of the viewport
var TEST_DATA_POINT = {
    x: (X_MIN.getTime() + X_MAX.getTime()) / 2,
    y: (Y_MIN - Y_MAX) / 2,
};
var ScWebglBarChartNegative = /** @class */ (function () {
    function ScWebglBarChartNegative(hostRef) {
        registerInstance(this, hostRef);
    }
    ScWebglBarChartNegative.prototype.render = function () {
        var _a;
        return (h("div", { id: "chart-container", style: { width: '500px', height: '500px' } }, h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    aggregates: (_a = {},
                        _a[MINUTE_IN_MS] = [TEST_DATA_POINT],
                        _a),
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "widget-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return ScWebglBarChartNegative;
}());
export { ScWebglBarChartNegative as sc_webgl_bar_chart_negative };
