// @synchro-charts/core/dist/esm-es5/sc-webgl-bar-chart-standard.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
import './utils-96fe4147.js';
import './predicates-67461267.js';
import './number-a7331d73.js';
import './dataFilters-2772c214.js';
import { T as TEST_DATA_POINT_STANDARD, b as Y_MIN, c as Y_MAX, X as X_MIN, d as X_MAX } from './constants-9d36da8a.js';
var ScWebglBarChartStandard = /** @class */ (function () {
    function ScWebglBarChartStandard(hostRef) {
        registerInstance(this, hostRef);
    }
    ScWebglBarChartStandard.prototype.render = function () {
        var _a;
        return (h("div", null, h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [],
                    aggregates: (_a = {},
                        _a[MINUTE_IN_MS] = [TEST_DATA_POINT_STANDARD],
                        _a),
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return ScWebglBarChartStandard;
}());
export { ScWebglBarChartStandard as sc_webgl_bar_chart_standard };
