// @synchro-charts/core/dist/esm-es5/sc-webgl-chart-tooltip-with-multiple-data-streams.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(1998, 0, 0);
var X_MAX = new Date(2000, 0, 1);
var TEST_DATA_POINT = {
    x: new Date((X_MIN.getTime() + X_MAX.getTime()) / 2).getTime(),
    y: 2000,
};
var TEST_DATA_POINT_2 = {
    x: new Date((X_MIN.getTime() + X_MAX.getTime()) / 2).getTime(),
    y: 3000,
};
var ScWebglChartTooltipWithMultipleDataStreams = /** @class */ (function () {
    function ScWebglChartTooltipWithMultipleDataStreams(hostRef) {
        registerInstance(this, hostRef);
    }
    ScWebglChartTooltipWithMultipleDataStreams.prototype.render = function () {
        return (h("div", null, h("sc-line-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_DATA_POINT],
                    resolution: 0,
                    dataType: DataType.NUMBER,
                },
                {
                    id: 'test2',
                    color: 'red',
                    name: 'test stream',
                    data: [TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "widget-id", size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }), h("sc-webgl-context", null)));
    };
    return ScWebglChartTooltipWithMultipleDataStreams;
}());
export { ScWebglChartTooltipWithMultipleDataStreams as sc_webgl_chart_tooltip_with_multiple_data_streams };
