// @synchro-charts/core/dist/esm-es5/sc-webgl-bar-chart-dynamic-data.entry.js
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(2000, 0);
var X_MAX = new Date(2001, 12);
var ScWebglBarChartDynamicData = /** @class */ (function () {
    function ScWebglBarChartDynamicData(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.data = [];
        this.monthIndex = 1;
        this.addDataPoint = function () {
            var dataPoint = {
                x: new Date(2000, _this.monthIndex).getTime(),
                y: 3000,
            };
            _this.data = __spreadArrays([dataPoint], _this.data);
            _this.monthIndex += 2;
        };
        this.removeDataPoint = function () {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            var _a = _this.data, _firstPoint = _a[0], rest = _a.slice(1);
            _this.data = rest;
        };
    }
    ScWebglBarChartDynamicData.prototype.render = function () {
        var _a;
        return (h("div", { class: "synchro-chart-tests" }, h("button", { id: "add-data-point", onClick: this.addDataPoint }, "Add Data Point"), h("button", { id: "remove-data-point", onClick: this.removeDataPoint }, "Remove Data Point"), h("br", null), h("br", null), h("div", { id: "chart-container", style: { height: '500px', width: '500px' } }, h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'red',
                    name: 'test stream',
                    resolution: MONTH_IN_MS,
                    aggregates: (_a = {},
                        _a[MONTH_IN_MS] = this.data,
                        _a),
                    data: [],
                    dataType: DataType.NUMBER,
                },
            ], size: {
                width: 500,
                height: 500,
            }, widgetId: "widget-id", viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            } }), h("sc-webgl-context", null))));
    };
    return ScWebglBarChartDynamicData;
}());
export { ScWebglBarChartDynamicData as sc_webgl_bar_chart_dynamic_data };
