// @synchro-charts/core/dist/esm-es5/single-status.entry.js
import { r as registerInstance, h, g as getElement } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { D as DAY_IN_MS } from './time-b732648c.js';
import './utils-96fe4147.js';
import './predicates-67461267.js';
import './number-a7331d73.js';
import './getDataPoints-6f8b4e89.js';
import './three.module-06da257a.js';
import { w as webGLRenderer } from './webglContext-da311040.js';
import { c as chartScene, H as HEIGHT } from './chartScene-1c4c48c4.js';
import './_commonjsHelpers-1d681858.js';
import './v4-9366210a.js';
import './index-2e17ff6c.js';
import './clipSpaceConversion-8273b9c5.js';
import './getDistanceFromDuration-5c7da5d2.js';
import { C as CHART_SIZE } from './chartSize-6ceb3800.js';
// viewport boundaries
var X_MIN = new Date(2000, 0, 0);
var X_MAX = new Date(2000, 0, 1);
var WIDTH = X_MAX.getTime() - X_MIN.getTime();
var TEST_DATA_POINT_1 = {
    x: X_MIN.getTime() + WIDTH / 3,
    y: 12,
};
var ScSingleStatus = /** @class */ (function () {
    function ScSingleStatus(hostRef) {
        registerInstance(this, hostRef);
    }
    ScSingleStatus.prototype.componentDidLoad = function () {
        var _a;
        var container = this.el.querySelector('#test-container');
        var scene = chartScene({
            alarms: { expires: DAY_IN_MS },
            viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: 0,
                yMax: HEIGHT,
            },
            dataStreams: [
                {
                    id: 'test-stream',
                    name: 'test-stream-name',
                    color: 'black',
                    resolution: DAY_IN_MS,
                    aggregates: (_a = {},
                        _a[DAY_IN_MS] = [TEST_DATA_POINT_1],
                        _a),
                    data: [],
                    dataType: DataType.NUMBER,
                },
            ],
            container: container,
            chartSize: CHART_SIZE,
            minBufferSize: 100,
            bufferFactor: 2,
            thresholdOptions: {
                showColor: false,
            },
            thresholds: [],
        });
        webGLRenderer.addChartScene({ manager: scene });
        var rect = container.getBoundingClientRect();
        webGLRenderer.setChartRect(scene.id, Object.assign({ density: 1 }, rect.toJSON()));
    };
    ScSingleStatus.prototype.render = function () {
        return (h("sc-webgl-context", null, h("div", { id: "test-container", style: { width: CHART_SIZE.width + "px", height: CHART_SIZE.height + "px" } })));
    };
    Object.defineProperty(ScSingleStatus.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: false,
        configurable: true
    });
    return ScSingleStatus;
}());
export { ScSingleStatus as single_status };
