// @synchro-charts/core/dist/esm-es5/status-timeline-threshold-coloration-band.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { C as COMPARISON_OPERATOR } from './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
import { b as Y_MIN, c as Y_MAX, X as X_MIN, d as X_MAX, T as TEST_DATA_POINT_STANDARD } from './constants-9d36da8a.js';
var urlParams = new URLSearchParams(window.location.search);
var isDiscreteNumericData = urlParams.get('isDiscreteNumericData');
var isStringData = urlParams.get('isStringData');
var data = TEST_DATA_POINT_STANDARD;
var dataType = isDiscreteNumericData === '1' ? DataType.STRING : DataType.NUMBER;
data.y = 2000;
if (isDiscreteNumericData === '1') {
    data.y = 2000;
}
if (isStringData === '1') {
    data.y = 'test';
}
var StatusTimelineThresholdBand = /** @class */ (function () {
    function StatusTimelineThresholdBand(hostRef) {
        registerInstance(this, hostRef);
    }
    StatusTimelineThresholdBand.prototype.render = function () {
        var _a;
        return (h("div", null, h("sc-status-timeline", { alarms: { expires: MINUTE_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    aggregates: (_a = {}, _a[MINUTE_IN_MS] = [data], _a),
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: dataType,
                },
            ], annotations: {
                y: [
                    {
                        value: isStringData ? 'test' : 2000,
                        label: {
                            text: 'y label',
                            show: true,
                        },
                        showValue: true,
                        color: 'blue',
                        comparisonOperator: COMPARISON_OPERATOR.EQUAL,
                    },
                ],
                thresholdOptions: {
                    showColor: true,
                },
            }, widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return StatusTimelineThresholdBand;
}());
export { StatusTimelineThresholdBand as status_timeline_threshold_coloration_band };
