// @synchro-charts/core/dist/esm-es5/status-timeline-standard.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
import { T as TEST_DATA_POINT_STANDARD, b as Y_MIN, c as Y_MAX, X as X_MIN, d as X_MAX } from './constants-9d36da8a.js';
var StatusTimelineStandard = /** @class */ (function () {
    function StatusTimelineStandard(hostRef) {
        registerInstance(this, hostRef);
    }
    StatusTimelineStandard.prototype.render = function () {
        return (h("div", null, h("sc-status-timeline", { alarms: { expires: MINUTE_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 70 })],
                    resolution: 0,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return StatusTimelineStandard;
}());
export { StatusTimelineStandard as status_timeline_standard };
