// @synchro-charts/core/dist/esm-es5/status-timeline-margin.entry.js
var _a;
import { r as registerInstance, h } from './index-0c943849.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { a as MONTH_IN_MS } from './time-b732648c.js';
// viewport boundaries
var Y_MIN = 0;
var Y_MAX = 5000;
var X_MIN = new Date(1998, 1, 0);
var X_MAX = new Date(1998, 6, 0);
var DATA_STREAM_1 = {
    id: 'test',
    color: 'red',
    name: 'test stream',
    resolution: MONTH_IN_MS,
    aggregates: (_a = {},
        _a[MONTH_IN_MS] = [
            { x: new Date(1998, 3, 0, 0).getTime(), y: 1000 },
            { x: new Date(1998, 4, 0, 0).getTime(), y: 3 },
        ],
        _a),
    data: [],
    dataType: DataType.NUMBER,
};
var DATA_STREAM_2 = {
    id: 'test2',
    color: 'orange',
    name: 'test stream2',
    resolution: MONTH_IN_MS,
    data: [
        { x: new Date(1998, 3, 0, 0).getTime(), y: 2000 },
        { x: new Date(1998, 4, 0, 0).getTime(), y: 4 },
    ],
    dataType: DataType.NUMBER,
};
var StatusTimelineStatusMargin = /** @class */ (function () {
    function StatusTimelineStatusMargin(hostRef) {
        registerInstance(this, hostRef);
        this.data = [];
    }
    StatusTimelineStatusMargin.prototype.render = function () {
        return (h("div", { id: "chart-container", style: { height: '500px', width: '500px', marginTop: '20px' } }, h("sc-status-timeline", { alarms: { expires: MONTH_IN_MS }, widgetId: "widget-id", dataStreams: [DATA_STREAM_1, DATA_STREAM_2], viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return StatusTimelineStatusMargin;
}());
export { StatusTimelineStatusMargin as status_timeline_margin };
