// @synchro-charts/core/dist/esm-es5/status-timeline-multiple-data-streams.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
import './utils-96fe4147.js';
import './predicates-67461267.js';
import './number-a7331d73.js';
import './dataFilters-2772c214.js';
import { T as TEST_DATA_POINT_STANDARD, b as Y_MIN, c as Y_MAX, X as X_MIN, d as X_MAX } from './constants-9d36da8a.js';
var StatusTimelineMultipleDataStreams = /** @class */ (function () {
    function StatusTimelineMultipleDataStreams(hostRef) {
        registerInstance(this, hostRef);
    }
    StatusTimelineMultipleDataStreams.prototype.render = function () {
        var _a, _b, _c;
        return (h("div", null, h("sc-status-timeline", { alarms: { expires: MINUTE_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream 1',
                    aggregates: (_a = {}, _a[MINUTE_IN_MS] = [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 70 })], _a),
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
                {
                    id: 'test2',
                    color: 'blue',
                    name: 'test stream 2',
                    aggregates: (_b = {}, _b[MINUTE_IN_MS] = [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 170 })], _b),
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
                {
                    id: 'test3',
                    color: 'red',
                    name: 'test stream 3',
                    aggregates: (_c = {}, _c[MINUTE_IN_MS] = [Object.assign(Object.assign({}, TEST_DATA_POINT_STANDARD), { y: 60 })], _c),
                    data: [],
                    resolution: MINUTE_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return StatusTimelineMultipleDataStreams;
}());
export { StatusTimelineMultipleDataStreams as status_timeline_multiple_data_streams };
