// @synchro-charts/core/dist/esm-es5/status-timeline-raw-data.entry.js
import { r as registerInstance, h } from './index-0c943849.js';
import { C as COMPARISON_OPERATOR } from './constants-35570903.js';
import { D as DataType } from './dataConstants-856cb1e9.js';
import './index-b3f1c3a1.js';
import { M as MINUTE_IN_MS } from './time-b732648c.js';
// Ten Minute Duration Viewport
var Y_MAX = 100;
var Y_MIN = 0;
var X_MIN = new Date(2000, 0, 0);
var X_MAX = new Date(2000, 0, 0, 0, 10);
var STATUS_ZERO = 'STATUS_ZERO';
var STATUS_ONE = 'STATUS_ONE';
var STATUS_TWO = 'STATUS_TWO';
var STATUS_THREE = 'STATUS_THREE';
var DATA_POINTS = [
    { x: X_MIN.getTime(), y: STATUS_ZERO },
    { x: X_MIN.getTime() + 5 * MINUTE_IN_MS, y: STATUS_ONE },
    { x: X_MIN.getTime() + 6 * MINUTE_IN_MS, y: STATUS_TWO },
    { x: X_MIN.getTime() + 6.5 * MINUTE_IN_MS, y: STATUS_THREE },
];
var annotations = {
    y: [
        {
            value: STATUS_ZERO,
            color: 'red',
            comparisonOperator: COMPARISON_OPERATOR.EQUAL,
        },
        {
            value: STATUS_ONE,
            color: 'blue',
            comparisonOperator: COMPARISON_OPERATOR.EQUAL,
        },
        {
            value: STATUS_TWO,
            color: 'green',
            comparisonOperator: COMPARISON_OPERATOR.EQUAL,
        },
    ],
};
var StatusTimelineRawData = /** @class */ (function () {
    function StatusTimelineRawData(hostRef) {
        registerInstance(this, hostRef);
    }
    StatusTimelineRawData.prototype.render = function () {
        return (h("div", null, h("sc-status-timeline", { alarms: { expires: MINUTE_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    data: DATA_POINTS,
                    color: 'black',
                    name: 'test stream 1',
                    resolution: 0,
                    dataType: DataType.STRING,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, annotations: annotations, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), h("sc-webgl-context", null)));
    };
    return StatusTimelineRawData;
}());
export { StatusTimelineRawData as status_timeline_raw_data };
