// @synchro-charts/core/dist/collection/testing/test-routes/charts/sc-webgl-scatter-chart/sc-scatter-chart-unsupported-data-types.js
import { Component, h } from '@stencil/core';
import { MONTH_IN_MS } from '../../../../utils/time';
import { DataType } from '../../../../utils/dataConstants';
import { Y_VALUE_STRING } from '../constants';
// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2001, 0, 1);
const TEST_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: Y_VALUE_STRING,
};
const TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: Y_VALUE_STRING,
};
export class ScScatterChartUnsupportedDataTypes {
    render() {
        return (h("div", null,
            h("sc-scatter-chart", { widgetId: "widget-id", dataStreams: [
                    {
                        id: 'test',
                        name: 'test stream',
                        color: 'black',
                        aggregates: { [MONTH_IN_MS]: [TEST_DATA_POINT, TEST_DATA_POINT_2] },
                        data: [],
                        resolution: MONTH_IN_MS,
                        dataType: DataType.STRING,
                    },
                ], annotations: {}, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }),
            h("sc-webgl-context", null)));
    }
    static get is() { return "sc-scatter-chart-unsupported-data-types"; }
}
