// @synchro-charts/core/dist/collection/testing/test-routes/sc-size-provider/sc-size-provider-standard.js
import { Component, h, State } from '@stencil/core';
import { DEFAULT_SIZE, SHIFT_X_DIFF } from './constants';
export class ScSizeProviderStandard {
    constructor() {
        this.marginLeft = 0;
        this.onShiftRight = () => {
            this.marginLeft += SHIFT_X_DIFF;
        };
    }
    render() {
        return (h("div", { class: "synchro-chart-tests", style: { width: '2000px', height: '2000px' } },
            h("div", null,
                h("button", { id: "shift-right", onClick: this.onShiftRight }, "Shift Right")),
            h("div", { id: "container", style: {
                    marginLeft: `${this.marginLeft}px`,
                    height: `${DEFAULT_SIZE.height}px`,
                    width: `${DEFAULT_SIZE.width}px`,
                } },
                h("sc-size-provider", { renderFunc: size => h("sc-box", { someObject: size, size: size }) }))));
    }
    static get is() { return "sc-size-provider-standard"; }
    static get states() { return {
        "marginLeft": {}
    }; }
}
