// @awsui/components-react/internal/generated/styles/tokens.js
export var colorAmazonOrange = "var(--color-amazon-orange-4dw41l, #ff9900)";
export var colorAwsSquidInk = "var(--color-aws-squid-ink-lgrnz0, #232f3e)";
export var colorBlack = "var(--color-black-mqvi5g, #000000)";
export var colorBlue100 = "var(--color-blue-100-l6py5i, #f1faff)";
export var colorBlue200 = "var(--color-blue-200-cckjki, #d1f1ff)";
export var colorBlue300 = "var(--color-blue-300-4hpa7r, #99cbe4)";
export var colorBlue400 = "var(--color-blue-400-30i7wf, #44b9d6)";
export var colorBlue500 = "var(--color-blue-500-n2cmt2, #00a1c9)";
export var colorBlue600 = "var(--color-blue-600-g04mdy, #0073bb)";
export var colorBlue700 = "var(--color-blue-700-gv0cdv, #0a4a74)";
export var colorBlue800 = "var(--color-blue-800-3q5reu, #002b66)";
export var colorBlue900 = "var(--color-blue-900-217vcq, #12293b)";
export var colorBlueOpaque = "var(--color-blue-opaque-xtaw90, rgba(51, 136, 221, 0.5))";
export var colorGreen100 = "var(--color-green-100-ft8yl5, #f2f8f0)";
export var colorGreen500 = "var(--color-green-500-rnyhey, #6aaf35)";
export var colorGreen600 = "var(--color-green-600-np6fsm, #1d8102)";
export var colorGreen700 = "var(--color-green-700-b8wf79, #00471e)";
export var colorGreen900 = "var(--color-green-900-7iyupb, #172211)";
export var colorGrey100 = "var(--color-grey-100-c9tcrt, #fafafa)";
export var colorGrey125 = "var(--color-grey-125-tdsxpg, #f6f6f9)";
export var colorGrey150 = "var(--color-grey-150-pkzcfi, #f2f3f3)";
export var colorGrey200 = "var(--color-grey-200-vj72qz, #eaeded)";
export var colorGrey300 = "var(--color-grey-300-4ddb26, #d5dbdb)";
export var colorGrey350 = "var(--color-grey-350-9r2qxk, #c6c6cd)";
export var colorGrey400 = "var(--color-grey-400-3g8wdk, #aab7b8)";
export var colorGrey450 = "var(--color-grey-450-uq3ea9, #95a5a6)";
export var colorGrey500 = "var(--color-grey-500-6yba5o, #879596)";
export var colorGrey550 = "var(--color-grey-550-c5n46a, #687078)";
export var colorGrey600 = "var(--color-grey-600-w2xlqv, #545b64)";
export var colorGrey650 = "var(--color-grey-650-ja4cpk, #414750)";
export var colorGrey700 = "var(--color-grey-700-5jq36n, #2a2e33)";
export var colorGrey750 = "var(--color-grey-750-29qz9f, #21252c)";
export var colorGrey800 = "var(--color-grey-800-96i17i, #1a2029)";
export var colorGrey850 = "var(--color-grey-850-m35fdb, #131920)";
export var colorGrey900 = "var(--color-grey-900-aijj2u, #16191f)";
export var colorGrey950 = "var(--color-grey-950-wez2yb, #06080a)";
export var colorOrange100 = "var(--color-orange-100-p1tn77, #fef6f0)";
export var colorOrange500 = "var(--color-orange-500-8cknkw, #ec7211)";
export var colorOrange600 = "var(--color-orange-600-dmjf5g, #eb5f07)";
export var colorOrange700 = "var(--color-orange-700-zj5hfw, #dd6b10)";
export var colorPurple400 = "var(--color-purple-400-ugpymi, #bf80ff)";
export var colorPurple600 = "var(--color-purple-600-dxe4k6, #7300e5)";
export var colorRed100 = "var(--color-red-100-i3807x, #fdf3f1)";
export var colorRed500 = "var(--color-red-500-u42cd5, #ff5d64)";
export var colorRed600 = "var(--color-red-600-a4aapn, #d13212)";
export var colorRed700 = "var(--color-red-700-jnnvvp, #700000)";
export var colorRed900 = "var(--color-red-900-m7w46f, #270a11)";
export var colorTransparent = "var(--color-transparent-c8buz2, transparent)";
export var colorWhite = "var(--color-white-6tfg1z, #ffffff)";
export var colorYellow100 = "var(--color-yellow-100-aqm8wx, #fffef0)";
export var colorYellow600 = "var(--color-yellow-600-z95g8o, #ffe347)";
export var colorYellow700 = "var(--color-yellow-700-076etr, #fbd332)";
export var colorYellow800 = "var(--color-yellow-800-9aihos, #906806)";
export var colorYellow900 = "var(--color-yellow-900-jqrp3x, #191100)";
export var colorChartsRed300 = "var(--color-charts-red-300-23c2oa, #ea7158)";
export var colorChartsRed400 = "var(--color-charts-red-400-539ac2, #dc5032)";
export var colorChartsRed500 = "var(--color-charts-red-500-llkmm5, #d13313)";
export var colorChartsRed600 = "var(--color-charts-red-600-5wdull, #ba2e0f)";
export var colorChartsRed700 = "var(--color-charts-red-700-a6apbj, #a82a0c)";
export var colorChartsRed800 = "var(--color-charts-red-800-yx4mar, #972709)";
export var colorChartsRed900 = "var(--color-charts-red-900-rpnpq8, #892407)";
export var colorChartsRed1000 = "var(--color-charts-red-1000-qk5pc5, #7d2105)";
export var colorChartsRed1100 = "var(--color-charts-red-1100-pt1j9r, #721e03)";
export var colorChartsRed1200 = "var(--color-charts-red-1200-d626tx, #671c00)";
export var colorChartsOrange300 = "var(--color-charts-orange-300-885arv, #e07941)";
export var colorChartsOrange400 = "var(--color-charts-orange-400-8kv4m2, #cc5f21)";
export var colorChartsOrange500 = "var(--color-charts-orange-500-z22zi7, #bc4d01)";
export var colorChartsOrange600 = "var(--color-charts-orange-600-4myt4p, #a84401)";
export var colorChartsOrange700 = "var(--color-charts-orange-700-gaxkk6, #983c02)";
export var colorChartsOrange800 = "var(--color-charts-orange-800-io45li, #8a3603)";
export var colorChartsOrange900 = "var(--color-charts-orange-900-rz3zc9, #7e3103)";
export var colorChartsOrange1000 = "var(--color-charts-orange-1000-0qpjgf, #732c02)";
export var colorChartsOrange1100 = "var(--color-charts-orange-1100-grw837, #692801)";
export var colorChartsOrange1200 = "var(--color-charts-orange-1200-5nnge9, #602400)";
export var colorChartsYellow300 = "var(--color-charts-yellow-300-6c0xtx, #b2911c)";
export var colorChartsYellow400 = "var(--color-charts-yellow-400-u252ao, #9c7b0b)";
export var colorChartsYellow500 = "var(--color-charts-yellow-500-v2phgx, #8a6b05)";
export var colorChartsYellow600 = "var(--color-charts-yellow-600-4pv3hy, #7b5f04)";
export var colorChartsYellow700 = "var(--color-charts-yellow-700-3emu4t, #6f5504)";
export var colorChartsYellow800 = "var(--color-charts-yellow-800-edh0l0, #654d03)";
export var colorChartsYellow900 = "var(--color-charts-yellow-900-1zoghi, #5d4503)";
export var colorChartsYellow1000 = "var(--color-charts-yellow-1000-4ypqyp, #553f03)";
export var colorChartsYellow1100 = "var(--color-charts-yellow-1100-2bbrrm, #4d3901)";
export var colorChartsYellow1200 = "var(--color-charts-yellow-1200-pml6h8, #483300)";
export var colorChartsGreen300 = "var(--color-charts-green-300-xdy00p, #67a353)";
export var colorChartsGreen400 = "var(--color-charts-green-400-pyju52, #41902c)";
export var colorChartsGreen500 = "var(--color-charts-green-500-x1jom3, #1f8104)";
export var colorChartsGreen600 = "var(--color-charts-green-600-ax7yum, #1a7302)";
export var colorChartsGreen700 = "var(--color-charts-green-700-x1ubuk, #176702)";
export var colorChartsGreen800 = "var(--color-charts-green-800-erepp2, #145d02)";
export var colorChartsGreen900 = "var(--color-charts-green-900-jgvtf4, #125502)";
export var colorChartsGreen1000 = "var(--color-charts-green-1000-ftsnde, #104d01)";
export var colorChartsGreen1100 = "var(--color-charts-green-1100-8is059, #0f4601)";
export var colorChartsGreen1200 = "var(--color-charts-green-1200-pyehtq, #0d4000)";
export var colorChartsTeal300 = "var(--color-charts-teal-300-ze50a3, #2ea597)";
export var colorChartsTeal400 = "var(--color-charts-teal-400-eho3qj, #1c8e81)";
export var colorChartsTeal500 = "var(--color-charts-teal-500-ysfodg, #0d7d70)";
export var colorChartsTeal600 = "var(--color-charts-teal-600-k0lr0d, #096f64)";
export var colorChartsTeal700 = "var(--color-charts-teal-700-pradbd, #06645a)";
export var colorChartsTeal800 = "var(--color-charts-teal-800-l98pd3, #045b52)";
export var colorChartsTeal900 = "var(--color-charts-teal-900-2n2uqa, #03524a)";
export var colorChartsTeal1000 = "var(--color-charts-teal-1000-sqzv15, #014b44)";
export var colorChartsTeal1100 = "var(--color-charts-teal-1100-ugmadm, #01443e)";
export var colorChartsTeal1200 = "var(--color-charts-teal-1200-u1l5vv, #003e38)";
export var colorChartsBlue1300 = "var(--color-charts-blue-1-300-bhq5nc, #529ccb)";
export var colorChartsBlue1400 = "var(--color-charts-blue-1-400-6shq6z, #3184c2)";
export var colorChartsBlue1500 = "var(--color-charts-blue-1-500-u9fbb4, #0273bb)";
export var colorChartsBlue1600 = "var(--color-charts-blue-1-600-6cm0lb, #0166ab)";
export var colorChartsBlue1700 = "var(--color-charts-blue-1-700-pur7e5, #015b9d)";
export var colorChartsBlue1800 = "var(--color-charts-blue-1-800-d90xfc, #015292)";
export var colorChartsBlue1900 = "var(--color-charts-blue-1-900-ugbzx8, #014a87)";
export var colorChartsBlue11000 = "var(--color-charts-blue-1-1000-wn8i30, #01437d)";
export var colorChartsBlue11100 = "var(--color-charts-blue-1-1100-zh7t8t, #003c75)";
export var colorChartsBlue11200 = "var(--color-charts-blue-1-1200-7spkcj, #00366d)";
export var colorChartsBlue2300 = "var(--color-charts-blue-2-300-25paew, #688ae8)";
export var colorChartsBlue2400 = "var(--color-charts-blue-2-400-ppuj2w, #5978e3)";
export var colorChartsBlue2500 = "var(--color-charts-blue-2-500-30q04l, #4066df)";
export var colorChartsBlue2600 = "var(--color-charts-blue-2-600-bqdh5s, #3759ce)";
export var colorChartsBlue2700 = "var(--color-charts-blue-2-700-useohu, #314fbf)";
export var colorChartsBlue2800 = "var(--color-charts-blue-2-800-vbehwc, #2c46b1)";
export var colorChartsBlue2900 = "var(--color-charts-blue-2-900-rjxjr8, #273ea5)";
export var colorChartsBlue21000 = "var(--color-charts-blue-2-1000-fo6yfn, #23379b)";
export var colorChartsBlue21100 = "var(--color-charts-blue-2-1100-r56l0z, #1f3191)";
export var colorChartsBlue21200 = "var(--color-charts-blue-2-1200-mrmzh5, #1b2b88)";
export var colorChartsPurple300 = "var(--color-charts-purple-300-fks7nd, #a783e1)";
export var colorChartsPurple400 = "var(--color-charts-purple-400-g5p5yw, #9469d6)";
export var colorChartsPurple500 = "var(--color-charts-purple-500-teqjg9, #8456ce)";
export var colorChartsPurple600 = "var(--color-charts-purple-600-nf5r9n, #7749bf)";
export var colorChartsPurple700 = "var(--color-charts-purple-700-15kpqx, #6b40b2)";
export var colorChartsPurple800 = "var(--color-charts-purple-800-fskedv, #6237a7)";
export var colorChartsPurple900 = "var(--color-charts-purple-900-9ytuxe, #59309d)";
export var colorChartsPurple1000 = "var(--color-charts-purple-1000-yxj5hz, #512994)";
export var colorChartsPurple1100 = "var(--color-charts-purple-1100-meod30, #4a238b)";
export var colorChartsPurple1200 = "var(--color-charts-purple-1200-d8mpzv, #431d84)";
export var colorChartsPink300 = "var(--color-charts-pink-300-9pujzo, #da7596)";
export var colorChartsPink400 = "var(--color-charts-pink-400-s9k4qh, #ce567c)";
export var colorChartsPink500 = "var(--color-charts-pink-500-pic0fb, #c33d69)";
export var colorChartsPink600 = "var(--color-charts-pink-600-is3nog, #b1325c)";
export var colorChartsPink700 = "var(--color-charts-pink-700-wmp2jn, #a32952)";
export var colorChartsPink800 = "var(--color-charts-pink-800-u9xp2q, #962249)";
export var colorChartsPink900 = "var(--color-charts-pink-900-171x5m, #8b1b42)";
export var colorChartsPink1000 = "var(--color-charts-pink-1000-cm89wx, #81143b)";
export var colorChartsPink1100 = "var(--color-charts-pink-1100-0bdx0v, #780d35)";
export var colorChartsPink1200 = "var(--color-charts-pink-1200-af67f8, #6f062f)";
export var colorChartsStatusCritical = "var(--color-charts-status-critical-81ggtj, #7d2105)";
export var colorChartsStatusHigh = "var(--color-charts-status-high-qe7s3z, #ba2e0f)";
export var colorChartsStatusMedium = "var(--color-charts-status-medium-a45eb3, #cc5f21)";
export var colorChartsStatusLow = "var(--color-charts-status-low-vohc9l, #b2911c)";
export var colorChartsStatusPositive = "var(--color-charts-status-positive-bswwz8, #67a353)";
export var colorChartsStatusInfo = "var(--color-charts-status-info-97ztp9, #3184c2)";
export var colorChartsStatusNeutral = "var(--color-charts-status-neutral-ik95f5, #879596)";
export var colorChartsThresholdNegative = "var(--color-charts-threshold-negative-adabc7, #d13212)";
export var colorChartsThresholdPositive = "var(--color-charts-threshold-positive-rtbub9, #1d8102)";
export var colorChartsThresholdInfo = "var(--color-charts-threshold-info-3592ya, #0073bb)";
export var colorChartsThresholdNeutral = "var(--color-charts-threshold-neutral-h58y4n, #687078)";
export var colorChartsLineGrid = "var(--color-charts-line-grid-b5jia7, #d5dbdb)";
export var colorChartsLineTick = "var(--color-charts-line-tick-6u2dih, #d5dbdb)";
export var colorChartsLineAxis = "var(--color-charts-line-axis-6cxsbx, #d5dbdb)";
export var colorChartsPaletteCategorical1 = "var(--color-charts-palette-categorical-1-s02edj, #688ae8)";
export var colorChartsPaletteCategorical2 = "var(--color-charts-palette-categorical-2-hd7uud, #c33d69)";
export var colorChartsPaletteCategorical3 = "var(--color-charts-palette-categorical-3-tm6hrf, #2ea597)";
export var colorChartsPaletteCategorical4 = "var(--color-charts-palette-categorical-4-fqnj32, #8456ce)";
export var colorChartsPaletteCategorical5 = "var(--color-charts-palette-categorical-5-h9jmtt, #e07941)";
export var colorChartsPaletteCategorical6 = "var(--color-charts-palette-categorical-6-z7e3fr, #3759ce)";
export var colorChartsPaletteCategorical7 = "var(--color-charts-palette-categorical-7-4w1lrz, #962249)";
export var colorChartsPaletteCategorical8 = "var(--color-charts-palette-categorical-8-xesix1, #096f64)";
export var colorChartsPaletteCategorical9 = "var(--color-charts-palette-categorical-9-93ht36, #6237a7)";
export var colorChartsPaletteCategorical10 = "var(--color-charts-palette-categorical-10-wus8nq, #a84401)";
export var colorChartsPaletteCategorical11 = "var(--color-charts-palette-categorical-11-uh7roa, #273ea5)";
export var colorChartsPaletteCategorical12 = "var(--color-charts-palette-categorical-12-ywqdlz, #780d35)";
export var colorChartsPaletteCategorical13 = "var(--color-charts-palette-categorical-13-0aws3q, #03524a)";
export var colorChartsPaletteCategorical14 = "var(--color-charts-palette-categorical-14-5l0pt5, #4a238b)";
export var colorChartsPaletteCategorical15 = "var(--color-charts-palette-categorical-15-glpjzo, #7e3103)";
export var colorChartsPaletteCategorical16 = "var(--color-charts-palette-categorical-16-73ijrp, #1b2b88)";
export var colorChartsPaletteCategorical17 = "var(--color-charts-palette-categorical-17-ltabw6, #ce567c)";
export var colorChartsPaletteCategorical18 = "var(--color-charts-palette-categorical-18-fog2ah, #003e38)";
export var colorChartsPaletteCategorical19 = "var(--color-charts-palette-categorical-19-ig4cn5, #9469d6)";
export var colorChartsPaletteCategorical20 = "var(--color-charts-palette-categorical-20-ta856w, #602400)";
export var colorChartsPaletteCategorical21 = "var(--color-charts-palette-categorical-21-ta0h08, #4066df)";
export var colorChartsPaletteCategorical22 = "var(--color-charts-palette-categorical-22-lj462v, #a32952)";
export var colorChartsPaletteCategorical23 = "var(--color-charts-palette-categorical-23-35jyqu, #0d7d70)";
export var colorChartsPaletteCategorical24 = "var(--color-charts-palette-categorical-24-eem6bw, #6b40b2)";
export var colorChartsPaletteCategorical25 = "var(--color-charts-palette-categorical-25-w302xt, #bc4d01)";
export var colorChartsPaletteCategorical26 = "var(--color-charts-palette-categorical-26-6ughlw, #2c46b1)";
export var colorChartsPaletteCategorical27 = "var(--color-charts-palette-categorical-27-57l9xa, #81143b)";
export var colorChartsPaletteCategorical28 = "var(--color-charts-palette-categorical-28-4q0mir, #045b52)";
export var colorChartsPaletteCategorical29 = "var(--color-charts-palette-categorical-29-cnn7dg, #512994)";
export var colorChartsPaletteCategorical30 = "var(--color-charts-palette-categorical-30-eix60t, #8a3603)";
export var colorChartsPaletteCategorical31 = "var(--color-charts-palette-categorical-31-jthnv1, #1f3191)";
export var colorChartsPaletteCategorical32 = "var(--color-charts-palette-categorical-32-h40wuu, #da7596)";
export var colorChartsPaletteCategorical33 = "var(--color-charts-palette-categorical-33-1iyt0i, #01443e)";
export var colorChartsPaletteCategorical34 = "var(--color-charts-palette-categorical-34-eut0yx, #a783e1)";
export var colorChartsPaletteCategorical35 = "var(--color-charts-palette-categorical-35-utj4hw, #692801)";
export var colorChartsPaletteCategorical36 = "var(--color-charts-palette-categorical-36-mjb3b0, #5978e3)";
export var colorChartsPaletteCategorical37 = "var(--color-charts-palette-categorical-37-6rz96d, #b1325c)";
export var colorChartsPaletteCategorical38 = "var(--color-charts-palette-categorical-38-s8j0ph, #1c8e81)";
export var colorChartsPaletteCategorical39 = "var(--color-charts-palette-categorical-39-5l5hgh, #7749bf)";
export var colorChartsPaletteCategorical40 = "var(--color-charts-palette-categorical-40-2i7lp3, #cc5f21)";
export var colorChartsPaletteCategorical41 = "var(--color-charts-palette-categorical-41-lozcy4, #314fbf)";
export var colorChartsPaletteCategorical42 = "var(--color-charts-palette-categorical-42-qzrtzy, #8b1b42)";
export var colorChartsPaletteCategorical43 = "var(--color-charts-palette-categorical-43-sq9ksf, #06645a)";
export var colorChartsPaletteCategorical44 = "var(--color-charts-palette-categorical-44-2mlngi, #59309d)";
export var colorChartsPaletteCategorical45 = "var(--color-charts-palette-categorical-45-irkhaj, #983c02)";
export var colorChartsPaletteCategorical46 = "var(--color-charts-palette-categorical-46-r0hukp, #23379b)";
export var colorChartsPaletteCategorical47 = "var(--color-charts-palette-categorical-47-l2wh9o, #6f062f)";
export var colorChartsPaletteCategorical48 = "var(--color-charts-palette-categorical-48-k8gno4, #014b44)";
export var colorChartsPaletteCategorical49 = "var(--color-charts-palette-categorical-49-e507mj, #431d84)";
export var colorChartsPaletteCategorical50 = "var(--color-charts-palette-categorical-50-hfws7f, #732c02)";
export var colorSeverityDarkRed = "var(--color-severity-dark-red-4k1f2e, #870303)";
export var colorSeverityRed = "var(--color-severity-red-6e8u8t, #ce3311)";
export var colorSeverityOrange = "var(--color-severity-orange-oir9u5, #f89256)";
export var colorSeverityYellow = "var(--color-severity-yellow-dr51ej, #f2cd54)";
export var colorSeverityGrey = "var(--color-severity-grey-nis4uy, #687078)";
export var colorBackgroundNotificationSeverityCritical = "var(--color-background-notification-severity-critical-ib32sw, #870303)";
export var colorBackgroundNotificationSeverityHigh = "var(--color-background-notification-severity-high-gv1pxl, #ce3311)";
export var colorBackgroundNotificationSeverityMedium = "var(--color-background-notification-severity-medium-hghstn, #f89256)";
export var colorBackgroundNotificationSeverityLow = "var(--color-background-notification-severity-low-asztnl, #f2cd54)";
export var colorBackgroundNotificationSeverityNeutral = "var(--color-background-notification-severity-neutral-1tkwxg, #687078)";
export var colorTextNotificationSeverityCritical = "var(--color-text-notification-severity-critical-eowkv6, #fafafa)";
export var colorTextNotificationSeverityHigh = "var(--color-text-notification-severity-high-oant7z, #fafafa)";
export var colorTextNotificationSeverityMedium = "var(--color-text-notification-severity-medium-hglkx1, #000000)";
export var colorTextNotificationSeverityLow = "var(--color-text-notification-severity-low-nq2syi, #16191f)";
export var colorTextNotificationSeverityNeutral = "var(--color-text-notification-severity-neutral-1948do, #fafafa)";
export var colorGreyOpaque10 = "var(--color-grey-opaque-10-jfbu9n, rgba(0, 0, 0, 0.1))";
export var colorGreyOpaque25 = "var(--color-grey-opaque-25-30w3ix, rgba(255, 255, 255, 0.25))";
export var colorGreyOpaque40 = "var(--color-grey-opaque-40-q0psxn, rgba(0, 0, 0, 0.4))";
export var colorGreyOpaque50 = "var(--color-grey-opaque-50-l76obh, rgba(0, 0, 0, 0.5))";
export var colorGreyOpaque70 = "var(--color-grey-opaque-70-h6950x, rgba(255, 255, 255, 0.7))";
export var colorGreyOpaque80 = "var(--color-grey-opaque-80-rkmedf, rgba(22, 25, 31, 0.8))";
export var colorGreyOpaque90 = "var(--color-grey-opaque-90-tgknfm, rgba(242, 243, 243, 0.9))";
export var colorGreyTransparent = "var(--color-grey-transparent-8l9x7z, rgba(0, 28, 36, 0.3))";
export var colorGreyTransparentHeavy = "var(--color-grey-transparent-heavy-31x0o5, rgba(0, 28, 36, 0.5))";
export var colorGreyTransparentLight = "var(--color-grey-transparent-light-gyg0yk, rgba(0, 28, 36, 0.15))";
export var colorBackgroundBadgeIcon = "var(--color-background-badge-icon-f6ehyg, #d13212)";
export var colorBackgroundButtonLinkActive = "var(--color-background-button-link-active-nmhuw7, #eaeded)";
export var colorBackgroundButtonLinkHover = "var(--color-background-button-link-hover-p90906, #fafafa)";
export var colorBackgroundButtonNormalActive = "var(--color-background-button-normal-active-wuxvsv, #eaeded)";
export var colorBackgroundButtonNormalDefault = "var(--color-background-button-normal-default-5fdgc5, #ffffff)";
export var colorBackgroundButtonNormalDisabled = "var(--color-background-button-normal-disabled-pzswm7, #ffffff)";
export var colorBackgroundButtonNormalHover = "var(--color-background-button-normal-hover-odla4l, #fafafa)";
export var colorBackgroundToggleButtonNormalPressed = "var(--color-background-toggle-button-normal-pressed-dc1nqz, #eaeded)";
export var colorBackgroundButtonPrimaryActive = "var(--color-background-button-primary-active-mcw2z2, #12293b)";
export var colorBackgroundButtonPrimaryDefault = "var(--color-background-button-primary-default-sdrd7j, #0073bb)";
export var colorBackgroundButtonPrimaryDisabled = "var(--color-background-button-primary-disabled-57e0jj, #ffffff)";
export var colorBackgroundButtonPrimaryHover = "var(--color-background-button-primary-hover-yt5487, #0a4a74)";
export var colorBackgroundCalendarCurrentDate = "var(--color-background-calendar-current-date-i2bg1g, #f2f3f3)";
export var colorBackgroundCellShaded = "var(--color-background-cell-shaded-47masj, #f2f3f3)";
export var colorBackgroundCodeEditorGutterActiveLineDefault = "var(--color-background-code-editor-gutter-active-line-default-5c0bff, #687078)";
export var colorBackgroundCodeEditorGutterActiveLineError = "var(--color-background-code-editor-gutter-active-line-error-g995ge, #d13212)";
export var colorBackgroundCodeEditorGutterDefault = "var(--color-background-code-editor-gutter-default-4z602i, #f2f3f3)";
export var colorBackgroundCodeEditorLoading = "var(--color-background-code-editor-loading-5uj70h, #f2f3f3)";
export var colorBackgroundCodeEditorPaneItemHover = "var(--color-background-code-editor-pane-item-hover-1iy38c, #eaeded)";
export var colorBackgroundCodeEditorStatusBar = "var(--color-background-code-editor-status-bar-0czmp3, #f2f3f3)";
export var colorBackgroundContainerContent = "var(--color-background-container-content-cpvvzs, #ffffff)";
export var colorBackgroundContainerHeader = "var(--color-background-container-header-1edgmi, #fafafa)";
export var colorBackgroundControlChecked = "var(--color-background-control-checked-6j23t6, #0073bb)";
export var colorBackgroundControlDefault = "var(--color-background-control-default-6lirk9, #ffffff)";
export var colorBackgroundControlDisabled = "var(--color-background-control-disabled-h93c4j, #d5dbdb)";
export var colorBackgroundDropdownItemDefault = "var(--color-background-dropdown-item-default-q1rj8l, #ffffff)";
export var colorBackgroundDropdownItemDimmed = "var(--color-background-dropdown-item-dimmed-oa2pjk, transparent)";
export var colorBackgroundDropdownItemFilterMatch = "var(--color-background-dropdown-item-filter-match-u1g6p2, #f1faff)";
export var colorBackgroundDropdownItemHover = "var(--color-background-dropdown-item-hover-6zmazg, #f2f3f3)";
export var colorBackgroundDropdownItemSelected = "var(--color-background-dropdown-item-selected-6qo3t1, #f1faff)";
export var colorBackgroundHomeHeader = "var(--color-background-home-header-q7il70, #000000)";
export var colorBackgroundInputDefault = "var(--color-background-input-default-xowsm5, #ffffff)";
export var colorBackgroundInputDisabled = "var(--color-background-input-disabled-fib0qr, #eaeded)";
export var colorBackgroundItemSelected = "var(--color-background-item-selected-bc67ge, #f1faff)";
export var colorBackgroundLayoutMain = "var(--color-background-layout-main-f3jtyd, #f2f3f3)";
export var colorBackgroundLayoutMobilePanel = "var(--color-background-layout-mobile-panel-vzg4qz, #ffffff)";
export var colorBackgroundLayoutPanelContent = "var(--color-background-layout-panel-content-jymwa8, #ffffff)";
export var colorBackgroundLayoutPanelHover = "var(--color-background-layout-panel-hover-ogjd65, #eaeded)";
export var colorBackgroundLayoutToggleActive = "var(--color-background-layout-toggle-active-9rnqxf, #2a2e33)";
export var colorBackgroundLayoutToggleDefault = "var(--color-background-layout-toggle-default-b3e5ga, transparent)";
export var colorBackgroundLayoutToggleHover = "var(--color-background-layout-toggle-hover-zeuoxl, #545b64)";
export var colorBackgroundLayoutToggleSelectedActive = "var(--color-background-layout-toggle-selected-active-67ljrt, #0073bb)";
export var colorBackgroundLayoutToggleSelectedDefault = "var(--color-background-layout-toggle-selected-default-tq3fmx, #0073bb)";
export var colorBackgroundLayoutToggleSelectedHover = "var(--color-background-layout-toggle-selected-hover-1zcs0k, #0a4a74)";
export var colorBackgroundModalOverlay = "var(--color-background-modal-overlay-zp5lne, rgba(242, 243, 243, 0.9))";
export var colorBackgroundNotificationBlue = "var(--color-background-notification-blue-jz7lex, #0073bb)";
export var colorBackgroundNotificationGreen = "var(--color-background-notification-green-0jtiqi, #1d8102)";
export var colorBackgroundNotificationGrey = "var(--color-background-notification-grey-5thu1e, #545b64)";
export var colorBackgroundNotificationRed = "var(--color-background-notification-red-05rvmd, #d13212)";
export var colorBackgroundNotificationYellow = "var(--color-background-notification-yellow-vv98ba, #ffe347)";
export var colorBackgroundNotificationStackBar = "var(--color-background-notification-stack-bar-u5f8lr, #2a2e33)";
export var colorBackgroundNotificationStackBarActive = "var(--color-background-notification-stack-bar-active-c74y1k, #2a2e33)";
export var colorBackgroundNotificationStackBarHover = "var(--color-background-notification-stack-bar-hover-mjfc85, #545b64)";
export var colorBackgroundPopover = "var(--color-background-popover-q6ltpd, #ffffff)";
export var colorBackgroundProgressBarContentDefault = "var(--color-background-progress-bar-content-default-ix5bce, #0073bb)";
export var colorBackgroundProgressBarContentInFlash = "var(--color-background-progress-bar-content-in-flash-u94w2k, rgba(255, 255, 255, 0.7))";
export var colorBackgroundProgressBarLayoutDefault = "var(--color-background-progress-bar-layout-default-49vehf, #eaeded)";
export var colorBackgroundProgressBarLayoutInFlash = "var(--color-background-progress-bar-layout-in-flash-37yw4h, rgba(255, 255, 255, 0.25))";
export var colorBackgroundSegmentActive = "var(--color-background-segment-active-73phfs, #0073bb)";
export var colorBackgroundSegmentDefault = "var(--color-background-segment-default-akl4mj, #ffffff)";
export var colorBackgroundSegmentDisabled = "var(--color-background-segment-disabled-4h6zm5, #ffffff)";
export var colorBackgroundSegmentHover = "var(--color-background-segment-hover-mpe3ui, #ffffff)";
export var colorBackgroundSliderHandleDefault = "var(--color-background-slider-handle-default-vg2dgl, #0073bb)";
export var colorBackgroundSliderHandleActive = "var(--color-background-slider-handle-active-nu61m7, #0a4a74)";
export var colorBackgroundSliderTrack = "var(--color-background-slider-track-d0xs3i, #879596)";
export var colorBackgroundSliderHandleRing = "var(--color-background-slider-handle-ring-8gbqdz, #ffffff)";
export var colorBackgroundSliderErrorPressed = "var(--color-background-slider-error-pressed-0sk3qh, #700000)";
export var colorBackgroundSliderWarningPressed = "var(--color-background-slider-warning-pressed-w5zjl3, #906806)";
export var colorBackgroundStatusError = "var(--color-background-status-error-mqjczn, #fdf3f1)";
export var colorBackgroundStatusInfo = "var(--color-background-status-info-07a262, #f1faff)";
export var colorBackgroundStatusSuccess = "var(--color-background-status-success-to45h8, #f2f8f0)";
export var colorBackgroundStatusWarning = "var(--color-background-status-warning-8czyg1, #fffef0)";
export var colorBackgroundTableHeader = "var(--color-background-table-header-rhxv7u, #fafafa)";
export var colorBackgroundTilesDisabled = "var(--color-background-tiles-disabled-94qpgn, #eaeded)";
export var colorBackgroundToggleCheckedDisabled = "var(--color-background-toggle-checked-disabled-jkq3ea, #99cbe4)";
export var colorBackgroundToggleDefault = "var(--color-background-toggle-default-64bz87, #545b64)";
export var colorBackgroundAvatarGenAi = "var(--color-background-avatar-gen-ai-eq7zlc, radial-gradient(circle farthest-corner at top left,rgba(0, 150, 250, 1) -25%,rgba(0, 150, 250, 0) 55%),radial-gradient(circle farthest-corner at top right, rgba(216, 178, 255, 1) -10%, rgba(115, 0, 229, 1) 50%))";
export var colorBackgroundAvatarDefault = "var(--color-background-avatar-default-xypvc3, #545b64)";
export var colorTextAvatar = "var(--color-text-avatar-y5n3t1, #ffffff)";
export var colorBackgroundLoadingBarGenAi = "var(--color-background-loading-bar-gen-ai-nbwy81, linear-gradient(90deg, #99f7ff 0%, #0096fa 10%, #bf80ff 24%, #7300e5 50%, #bf80ff 76%, #0096fa 90%, #99f7ff 100%))";
export var colorBackgroundChatBubbleOutgoing = "var(--color-background-chat-bubble-outgoing-7qngnp, transparent)";
export var colorBackgroundChatBubbleIncoming = "var(--color-background-chat-bubble-incoming-04glkh, #f2f3f3)";
export var colorTextChatBubbleOutgoing = "var(--color-text-chat-bubble-outgoing-79xrhd, #16191f)";
export var colorTextChatBubbleIncoming = "var(--color-text-chat-bubble-incoming-72165z, #16191f)";
export var colorBorderButtonNormalActive = "var(--color-border-button-normal-active-vkwe9f, #545b64)";
export var colorBorderButtonNormalDefault = "var(--color-border-button-normal-default-9jkxwh, #545b64)";
export var colorBorderToggleButtonNormalPressed = "var(--color-border-toggle-button-normal-pressed-c5n6wh, #545b64)";
export var colorBorderButtonNormalDisabled = "var(--color-border-button-normal-disabled-orwwg4, #d5dbdb)";
export var colorTextButtonNormalDisabled = "var(--color-text-button-normal-disabled-u9dsmh, #879596)";
export var colorBorderButtonNormalHover = "var(--color-border-button-normal-hover-ap4j3a, #16191f)";
export var colorBorderButtonPrimaryDisabled = "var(--color-border-button-primary-disabled-2hg9e9, #d5dbdb)";
export var colorTextButtonPrimaryDisabled = "var(--color-text-button-primary-disabled-a4619v, #879596)";
export var colorBorderCalendarGrid = "var(--color-border-calendar-grid-77wqzj, #eaeded)";
export var colorBorderCalendarGridSelectedFocusRing = "var(--color-border-calendar-grid-selected-focus-ring-nbq93t, #0073bb)";
export var colorBorderCodeEditorAceActiveLineLightTheme = "var(--color-border-code-editor-ace-active-line-light-theme-s3ap93, #d5dbdb)";
export var colorBorderCodeEditorAceActiveLineDarkTheme = "var(--color-border-code-editor-ace-active-line-dark-theme-nqjhd4, #687078)";
export var colorBorderCodeEditorDefault = "var(--color-border-code-editor-default-n0j971, #d5dbdb)";
export var colorBorderCodeEditorPaneItemHover = "var(--color-border-code-editor-pane-item-hover-lhf6r1, #687078)";
export var colorBorderContainerDivider = "var(--color-border-container-divider-1j5zof, #eaeded)";
export var colorBorderContainerTop = "var(--color-border-container-top-wiud2r, #eaeded)";
export var colorBorderControlChecked = "var(--color-border-control-checked-lshsoq, #0073bb)";
export var colorBorderControlDefault = "var(--color-border-control-default-nj5wiz, #687078)";
export var colorBorderControlDisabled = "var(--color-border-control-disabled-hrmqh9, #d5dbdb)";
export var colorBorderDividerActive = "var(--color-border-divider-active-vyvjsd, #687078)";
export var colorBorderDividerDefault = "var(--color-border-divider-default-g2zqci, #eaeded)";
export var colorBorderDividerPanelBottom = "var(--color-border-divider-panel-bottom-qg365q, rgba(0, 28, 36, 0.15))";
export var colorBorderDividerPanelSide = "var(--color-border-divider-panel-side-r7izb1, transparent)";
export var colorBorderDividerSecondary = "var(--color-border-divider-secondary-qt0aq9, #eaeded)";
export var colorBorderDropdownContainer = "var(--color-border-dropdown-container-xmqvfn, transparent)";
export var colorBorderDropdownGroup = "var(--color-border-dropdown-group-tg9vgt, #eaeded)";
export var colorBorderDropdownItemDefault = "var(--color-border-dropdown-item-default-tizn4l, #eaeded)";
export var colorBorderDropdownItemHover = "var(--color-border-dropdown-item-hover-8pzdhc, #879596)";
export var colorBorderDropdownItemDimmedHover = "var(--color-border-dropdown-item-dimmed-hover-7hrdqp, #879596)";
export var colorBorderDropdownItemSelected = "var(--color-border-dropdown-item-selected-3abw87, #eaeded)";
export var colorBorderDropdownItemTop = "var(--color-border-dropdown-item-top-chy7j8, #eaeded)";
export var colorBorderEditableCellHover = "var(--color-border-editable-cell-hover-01ffb1, #879596)";
export var colorBorderInputDefault = "var(--color-border-input-default-ijkrzo, #687078)";
export var colorBorderInputDisabled = "var(--color-border-input-disabled-i75jkh, #eaeded)";
export var colorBorderInputFocused = "var(--color-border-input-focused-5893wz, #0073bb)";
export var colorBorderItemFocused = "var(--color-border-item-focused-j88ehv, #0073bb)";
export var colorBorderDropdownItemFocused = "var(--color-border-dropdown-item-focused-hebyze, #0073bb)";
export var colorBorderItemPlaceholder = "var(--color-border-item-placeholder-1r9l90, transparent)";
export var colorBorderItemSelected = "var(--color-border-item-selected-xu3hwl, #0073bb)";
export var colorBorderLayout = "var(--color-border-layout-jd9fka, #d5dbdb)";
export var colorBorderNotificationStackBar = "var(--color-border-notification-stack-bar-0ixmq0, #2a2e33)";
export var colorBorderPanelHeader = "var(--color-border-panel-header-x89ddl, #eaeded)";
export var colorBorderPopover = "var(--color-border-popover-si3ei7, #d5dbdb)";
export var colorBorderSegmentActive = "var(--color-border-segment-active-mv1bbg, #16191f)";
export var colorBorderSegmentDefault = "var(--color-border-segment-default-4q3z1n, #687078)";
export var colorBorderSegmentDisabled = "var(--color-border-segment-disabled-vpnkqa, #d5dbdb)";
export var colorBorderSegmentHover = "var(--color-border-segment-hover-x2c7ym, #16191f)";
export var colorBorderStatusError = "var(--color-border-status-error-rw7xpm, #d13212)";
export var colorBorderStatusInfo = "var(--color-border-status-info-ytrrza, #0073bb)";
export var colorBorderStatusSuccess = "var(--color-border-status-success-re24ca, #1d8102)";
export var colorBorderStatusWarning = "var(--color-border-status-warning-absgf7, #906806)";
export var colorBorderDividerInteractiveDefault = "var(--color-border-divider-interactive-default-ojrs8s, #687078)";
export var colorBorderTabsDivider = "var(--color-border-tabs-divider-b29g6f, #aab7b8)";
export var colorBorderTabsShadow = "var(--color-border-tabs-shadow-60ydf7, #eaeded)";
export var colorBorderTabsUnderline = "var(--color-border-tabs-underline-fhgpyx, #16191f)";
export var colorBorderTilesDisabled = "var(--color-border-tiles-disabled-vfug4m, transparent)";
export var colorBorderTutorial = "var(--color-border-tutorial-ohwjx4, #aab7b8)";
export var colorForegroundControlDefault = "var(--color-foreground-control-default-2lrfrr, #ffffff)";
export var colorForegroundControlDisabled = "var(--color-foreground-control-disabled-dksqnz, #ffffff)";
export var colorForegroundControlReadOnly = "var(--color-foreground-control-read-only-4cwci1, #687078)";
export var colorShadowDefault = "var(--color-shadow-default-4lqu2z, rgba(0, 28, 36, 0.5))";
export var colorShadowMedium = "var(--color-shadow-medium-38rqv1, rgba(0, 28, 36, 0.3))";
export var colorShadowSide = "var(--color-shadow-side-vytace, rgba(0, 28, 36, 0.15))";
export var colorStrokeChartLine = "var(--color-stroke-chart-line-rgou38, #879596)";
export var colorStrokeCodeEditorResizeHandler = "var(--color-stroke-code-editor-resize-handler-6f09e4, #687078)";
export var colorStrokeCodeEditorGutterActiveLineDefault = "var(--color-stroke-code-editor-gutter-active-line-default-c0vpqr, #d5dbdb)";
export var colorStrokeCodeEditorGutterActiveLineHover = "var(--color-stroke-code-editor-gutter-active-line-hover-nz19fz, #fafafa)";
export var colorTextAccent = "var(--color-text-accent-27s5l3, #0073bb)";
export var colorTextBodyDefault = "var(--color-text-body-default-zs35hz, #16191f)";
export var colorTextBodySecondary = "var(--color-text-body-secondary-kc95tq, #545b64)";
export var colorTextBreadcrumbCurrent = "var(--color-text-breadcrumb-current-kt41ir, #687078)";
export var colorTextBreadcrumbIcon = "var(--color-text-breadcrumb-icon-8qiz1d, #687078)";
export var colorTextButtonInlineIconDefault = "var(--color-text-button-inline-icon-default-aqdqv7, #545b64)";
export var colorTextButtonInlineIconDisabled = "var(--color-text-button-inline-icon-disabled-o6s17n, #aab7b8)";
export var colorTextButtonInlineIconHover = "var(--color-text-button-inline-icon-hover-vsgk24, #16191f)";
export var colorTextButtonNormalActive = "var(--color-text-button-normal-active-nhgenz, #16191f)";
export var colorTextToggleButtonNormalPressed = "var(--color-text-toggle-button-normal-pressed-mmq27n, #16191f)";
export var colorTextButtonNormalDefault = "var(--color-text-button-normal-default-mc7t60, #545b64)";
export var colorTextButtonNormalHover = "var(--color-text-button-normal-hover-bki2l2, #16191f)";
export var colorTextLinkButtonNormalDefault = "var(--color-text-link-button-normal-default-1tq8hi, #545b64)";
export var colorTextLinkButtonNormalHover = "var(--color-text-link-button-normal-hover-ucswxh, #16191f)";
export var colorTextLinkButtonNormalActive = "var(--color-text-link-button-normal-active-qqlz4b, #16191f)";
export var colorTextButtonPrimaryActive = "var(--color-text-button-primary-active-0gc6a4, #ffffff)";
export var colorTextButtonPrimaryDefault = "var(--color-text-button-primary-default-v7x7uq, #ffffff)";
export var colorTextButtonPrimaryHover = "var(--color-text-button-primary-hover-imb9dy, #ffffff)";
export var colorTextCalendarDateHover = "var(--color-text-calendar-date-hover-q9r0yp, #16191f)";
export var colorTextCalendarDateSelected = "var(--color-text-calendar-date-selected-50yedd, #16191f)";
export var colorTextCalendarMonth = "var(--color-text-calendar-month-h5j04j, #545b64)";
export var colorTextCodeEditorGutterActiveLine = "var(--color-text-code-editor-gutter-active-line-6qr5dy, #ffffff)";
export var colorTextCodeEditorGutterDefault = "var(--color-text-code-editor-gutter-default-6diu38, #16191f)";
export var colorTextCodeEditorStatusBarDisabled = "var(--color-text-code-editor-status-bar-disabled-96znq5, #879596)";
export var colorTextCodeEditorTabButtonError = "var(--color-text-code-editor-tab-button-error-s2rugm, #ffffff)";
export var colorTextColumnHeader = "var(--color-text-column-header-xxfilq, #545b64)";
export var colorTextColumnSortingIcon = "var(--color-text-column-sorting-icon-xqw2vv, #687078)";
export var colorTextControlDisabled = "var(--color-text-control-disabled-ztkhgz, #aab7b8)";
export var colorTextCounter = "var(--color-text-counter-a9cq2m, #687078)";
export var colorTextDisabled = "var(--color-text-disabled-e7l57l, #aab7b8)";
export var colorTextDisabledInlineEdit = "var(--color-text-disabled-inline-edit-tp0ypb, #545b64)";
export var colorTextDropdownFooter = "var(--color-text-dropdown-footer-z0err6, #687078)";
export var colorTextDropdownGroupLabel = "var(--color-text-dropdown-group-label-s1hwg5, #545b64)";
export var colorTextDropdownItemDefault = "var(--color-text-dropdown-item-default-ti6pjc, #16191f)";
export var colorTextDropdownItemDimmed = "var(--color-text-dropdown-item-dimmed-6g8wds, #aab7b8)";
export var colorTextDropdownItemDisabled = "var(--color-text-dropdown-item-disabled-j9pzby, #aab7b8)";
export var colorTextDropdownItemFilterMatch = "var(--color-text-dropdown-item-filter-match-62fflr, #0073bb)";
export var colorTextDropdownItemHighlighted = "var(--color-text-dropdown-item-highlighted-0ns2tb, #16191f)";
export var colorTextDropdownItemSecondary = "var(--color-text-dropdown-item-secondary-rkiy3t, #687078)";
export var colorTextDropdownItemSecondaryHover = "var(--color-text-dropdown-item-secondary-hover-zyoem0, #687078)";
export var colorTextEmpty = "var(--color-text-empty-mn4nrj, #687078)";
export var colorTextExpandableSectionDefault = "var(--color-text-expandable-section-default-oowjk4, #545b64)";
export var colorTextExpandableSectionHover = "var(--color-text-expandable-section-hover-ws4pqf, #16191f)";
export var colorTextExpandableSectionNavigationIconDefault = "var(--color-text-expandable-section-navigation-icon-default-a0g6p0, #879596)";
export var colorTextFormDefault = "var(--color-text-form-default-b48udf, #16191f)";
export var colorTextFormLabel = "var(--color-text-form-label-uk0nvv, #16191f)";
export var colorTextFormSecondary = "var(--color-text-form-secondary-6vvbul, #687078)";
export var colorTextGroupLabel = "var(--color-text-group-label-37uws0, #545b64)";
export var colorTextLabelGenAi = "var(--color-text-label-gen-ai-6dxzrs, #7300e5)";
export var colorTextHeadingDefault = "var(--color-text-heading-default-wliqyu, #16191f)";
export var colorTextHeadingSecondary = "var(--color-text-heading-secondary-urdy5f, #545b64)";
export var colorTextHomeHeaderDefault = "var(--color-text-home-header-default-gz3zbq, #ffffff)";
export var colorTextHomeHeaderSecondary = "var(--color-text-home-header-secondary-v8lrx6, #d5dbdb)";
export var colorTextIconCaret = "var(--color-text-icon-caret-4trikz, #879596)";
export var colorTextIconSubtle = "var(--color-text-icon-subtle-x6qw4s, #687078)";
export var colorTextInputDisabled = "var(--color-text-input-disabled-scisz5, #879596)";
export var colorTextInputPlaceholder = "var(--color-text-input-placeholder-bclxfj, #687078)";
export var colorTextInputPlaceholderDisabled = "var(--color-text-input-placeholder-disabled-2odnh9, #687078)";
export var colorTextInteractiveActive = "var(--color-text-interactive-active-ifmuec, #16191f)";
export var colorTextInteractiveDefault = "var(--color-text-interactive-default-yjja72, #545b64)";
export var colorTextInteractiveDisabled = "var(--color-text-interactive-disabled-jwju02, #aab7b8)";
export var colorTextInteractiveHover = "var(--color-text-interactive-hover-fgfl9g, #16191f)";
export var colorTextToggleButtonIconPressed = "var(--color-text-toggle-button-icon-pressed-31hcbc, #16191f)";
export var colorTextInteractiveInvertedDefault = "var(--color-text-interactive-inverted-default-wi2ail, #d5dbdb)";
export var colorTextInteractiveInvertedHover = "var(--color-text-interactive-inverted-hover-hf0cki, #fafafa)";
export var colorTextInverted = "var(--color-text-inverted-brc9ln, #ffffff)";
export var colorTextLabel = "var(--color-text-label-du7fzg, #545b64)";
export var colorTextLayoutToggle = "var(--color-text-layout-toggle-sucjzk, #545b64)";
export var colorTextLayoutToggleActive = "var(--color-text-layout-toggle-active-g2y9le, #ffffff)";
export var colorTextLayoutToggleHover = "var(--color-text-layout-toggle-hover-1lm7x2, #0073bb)";
export var colorTextLayoutToggleSelected = "var(--color-text-layout-toggle-selected-9msxpj, #ffffff)";
export var colorTextLinkDefault = "var(--color-text-link-default-owstl4, #0073bb)";
export var colorTextLinkHover = "var(--color-text-link-hover-aohp90, #0a4a74)";
export var colorTextLinkInvertedHover = "var(--color-text-link-inverted-hover-5p8h7v, #fafafa)";
export var colorTextLinkButtonUnderline = "var(--color-text-link-button-underline-tca6v7, currentColor)";
export var colorTextLinkButtonUnderlineHover = "var(--color-text-link-button-underline-hover-trjqy2, currentColor)";
export var colorTextNotificationDefault = "var(--color-text-notification-default-zu8c4i, #fafafa)";
export var colorTextNotificationStackBar = "var(--color-text-notification-stack-bar-c0qu9n, #ffffff)";
export var colorTextNotificationYellow = "var(--color-text-notification-yellow-ut1o5a, #16191f)";
export var colorTextPaginationPageNumberActiveDisabled = "var(--color-text-pagination-page-number-active-disabled-alq2uc, #545b64)";
export var colorTextPaginationPageNumberDefault = "var(--color-text-pagination-page-number-default-rngnmy, #545b64)";
export var colorTextSegmentActive = "var(--color-text-segment-active-4d27ka, #ffffff)";
export var colorTextSegmentDefault = "var(--color-text-segment-default-48d69d, #545b64)";
export var colorTextSegmentHover = "var(--color-text-segment-hover-u9d76i, #0073bb)";
export var colorTextSmall = "var(--color-text-small-w05wee, #687078)";
export var colorTextStatusError = "var(--color-text-status-error-oa993i, #d13212)";
export var colorTextStatusInactive = "var(--color-text-status-inactive-97nxyu, #687078)";
export var colorTextStatusInfo = "var(--color-text-status-info-vkd30u, #0073bb)";
export var colorTextStatusSuccess = "var(--color-text-status-success-z431uf, #1d8102)";
export var colorTextStatusWarning = "var(--color-text-status-warning-a2jajm, #906806)";
export var colorTextTopNavigationTitle = "var(--color-text-top-navigation-title-fzc955, #16191f)";
export var colorBoardPlaceholderActive = "var(--color-board-placeholder-active-ua4ik1, #d5dbdb)";
export var colorBoardPlaceholderHover = "var(--color-board-placeholder-hover-jc1c35, #99cbe4)";
export var colorDragPlaceholderActive = "var(--color-drag-placeholder-active-smas3a, #d5dbdb)";
export var colorDragPlaceholderHover = "var(--color-drag-placeholder-hover-0mqnqu, #99cbe4)";
export var colorDropzoneBackgroundDefault = "var(--color-dropzone-background-default-cpzr1q, #ffffff)";
export var colorDropzoneBackgroundHover = "var(--color-dropzone-background-hover-lyldwy, #f1faff)";
export var colorDropzoneTextDefault = "var(--color-dropzone-text-default-yjkxlf, #545b64)";
export var colorDropzoneTextHover = "var(--color-dropzone-text-hover-nz9px6, #545b64)";
export var colorDropzoneBorderDefault = "var(--color-dropzone-border-default-y9gvh9, #879596)";
export var colorDropzoneBorderHover = "var(--color-dropzone-border-hover-sjjerq, #002b66)";
export var fontBoxValueLargeWeight = "var(--font-box-value-large-weight-irki1m, 300)";
export var fontButtonLetterSpacing = "var(--font-button-letter-spacing-6ag9q5, 0.25px)";
export var fontChartDetailSize = "var(--font-chart-detail-size-3pgg4u, 14px)";
export var fontDisplayLabelWeight = "var(--font-display-label-weight-q56x6f, 400)";
export var fontExpandableHeadingSize = "var(--font-expandable-heading-size-gqddjg, 14px)";
export var fontFamilyBase = "var(--font-family-base-5z8fw5, 'Noto Sans', 'Helvetica Neue', Roboto, Arial, sans-serif)";
export var fontFamilyMonospace = "var(--font-family-monospace-froq9j, Monaco, Menlo, Consolas, 'Courier Prime', Courier, 'Courier New', monospace)";
export var fontHeaderH2DescriptionLineHeight = "var(--font-header-h2-description-line-height-bhmh72, 16px)";
export var fontHeaderH2DescriptionSize = "var(--font-header-h2-description-size-jhlfgf, 12px)";
export var fontLinkButtonLetterSpacing = "var(--font-link-button-letter-spacing-3t8f8m, normal)";
export var fontLinkButtonWeight = "var(--font-link-button-weight-vgr3fq, 400)";
export var fontPanelHeaderLineHeight = "var(--font-panel-header-line-height-pqpwyv, 22px)";
export var fontPanelHeaderSize = "var(--font-panel-header-size-ey4ywd, 18px)";
export var fontSizeBodyM = "var(--font-size-body-m-kbh530, 14px)";
export var fontSizeBodyS = "var(--font-size-body-s-6sh5cc, 12px)";
export var fontSizeDisplayL = "var(--font-size-display-l-zttzwo, 44px)";
export var fontSizeHeadingXl = "var(--font-size-heading-xl-e8bqij, 28px)";
export var fontSizeHeadingL = "var(--font-size-heading-l-04luh4, 18px)";
export var fontSizeHeadingM = "var(--font-size-heading-m-9xjzrj, 18px)";
export var fontSizeHeadingS = "var(--font-size-heading-s-d8ts7u, 16px)";
export var fontSizeHeadingXs = "var(--font-size-heading-xs-pgspul, 16px)";
export var fontSmoothingMozOsx = "var(--font-smoothing-moz-osx-7ax6sl, auto)";
export var fontSmoothingWebkit = "var(--font-smoothing-webkit-gsroen, auto)";
export var fontTabsDisabledWeight = "var(--font-tabs-disabled-weight-2xstmx, 400)";
export var fontTabsLineHeight = "var(--font-tabs-line-height-2g3saf, 22px)";
export var fontTabsSize = "var(--font-tabs-size-74jl7j, 14px)";
export var fontWayfindingLinkActiveWeight = "var(--font-wayfinding-link-active-weight-frdnvo, 700)";
export var fontWeightButton = "var(--font-weight-button-rb1vba, 700)";
export var fontWeightHeadingXl = "var(--font-weight-heading-xl-nre0k2, 400)";
export var fontWeightHeadingL = "var(--font-weight-heading-l-0lcea3, 700)";
export var fontWeightHeadingM = "var(--font-weight-heading-m-0c5l08, 400)";
export var fontWeightHeadingS = "var(--font-weight-heading-s-6rdhmg, 700)";
export var fontWeightHeadingXs = "var(--font-weight-heading-xs-xuskfi, 400)";
export var fontWeightHeavy = "var(--font-weight-heavy-gd0ixk, 700)";
export var letterSpacingBodyS = "var(--letter-spacing-body-s-4f9u0r, normal)";
export var letterSpacingDisplayL = "var(--letter-spacing-display-l-btj7nl, normal)";
export var letterSpacingHeadingXl = "var(--letter-spacing-heading-xl-cbi7sv, normal)";
export var letterSpacingHeadingL = "var(--letter-spacing-heading-l-zb25sz, normal)";
export var letterSpacingHeadingM = "var(--letter-spacing-heading-m-m4ll99, normal)";
export var letterSpacingHeadingS = "var(--letter-spacing-heading-s-i1mjjy, normal)";
export var lineHeightBodyM = "var(--line-height-body-m-0awzkk, 22px)";
export var lineHeightBodyS = "var(--line-height-body-s-usc3fr, 16px)";
export var lineHeightDisplayL = "var(--line-height-display-l-0oa5q5, 56px)";
export var lineHeightHeadingXl = "var(--line-height-heading-xl-gnbljj, 36px)";
export var lineHeightHeadingL = "var(--line-height-heading-l-4wuokq, 22px)";
export var lineHeightHeadingM = "var(--line-height-heading-m-gv8yla, 22px)";
export var lineHeightHeadingS = "var(--line-height-heading-s-i8coai, 20px)";
export var lineHeightHeadingXs = "var(--line-height-heading-xs-8t7s1x, 20px)";
export var borderActiveWidth = "var(--border-active-width-1q70ry, 2px)";
export var borderCodeEditorStatusDividerWidth = "var(--border-code-editor-status-divider-width-z6f06q, 0px)";
export var borderContainerStickyWidth = "var(--border-container-sticky-width-gj4eew, 1px)";
export var borderContainerTopWidth = "var(--border-container-top-width-khbhr4, 1px)";
export var borderControlFocusRingShadowSpread = "var(--border-control-focus-ring-shadow-spread-o8rs1p, 1px)";
export var borderControlInvalidFocusRingShadowSpread = "var(--border-control-invalid-focus-ring-shadow-spread-quihle, 1px)";
export var borderDividerListWidth = "var(--border-divider-list-width-um3zli, 1px)";
export var borderDividerSectionWidth = "var(--border-divider-section-width-4f9905, 1px)";
export var borderDropdownVirtualOffsetWidth = "var(--border-dropdown-virtual-offset-width-4cjhpv, 0px)";
export var borderInvalidWidth = "var(--border-invalid-width-jub7d4, 4px)";
export var borderItemWidth = "var(--border-item-width-ip3vxj, 1px)";
export var borderLineChartDashArray = "var(--border-line-chart-dash-array-ry3ko4, 3 5)";
export var borderLineChartLineJoin = "var(--border-line-chart-line-join-27xl3m, miter)";
export var borderLineChartWidth = "var(--border-line-chart-width-qfyiov, 2px)";
export var borderPanelHeaderWidth = "var(--border-panel-header-width-7a2phv, 0px)";
export var borderPanelTopWidth = "var(--border-panel-top-width-6k0jao, 0px)";
export var borderRadiusAlert = "var(--border-radius-alert-8j75to, 2px)";
export var borderRadiusBadge = "var(--border-radius-badge-5bxe5h, 16px)";
export var borderRadiusButton = "var(--border-radius-button-f5klif, 2px)";
export var borderRadiusCalendarDayFocusRing = "var(--border-radius-calendar-day-focus-ring-i451al, 2px)";
export var borderRadiusCodeEditor = "var(--border-radius-code-editor-9j71jc, 0px)";
export var borderRadiusContainer = "var(--border-radius-container-r46evd, 0px)";
export var borderRadiusControlCircularFocusRing = "var(--border-radius-control-circular-focus-ring-adbgw5, 50%)";
export var borderRadiusControlDefaultFocusRing = "var(--border-radius-control-default-focus-ring-eq2wmk, 2px)";
export var borderRadiusDropdown = "var(--border-radius-dropdown-awjhsp, 0px)";
export var borderRadiusDropzone = "var(--border-radius-dropzone-x9mbj3, 0px)";
export var borderRadiusFlashbar = "var(--border-radius-flashbar-e6itil, 0px)";
export var borderRadiusItem = "var(--border-radius-item-ilq21a, 0px)";
export var borderRadiusInput = "var(--border-radius-input-gycuv6, 2px)";
export var borderRadiusPopover = "var(--border-radius-popover-ewhy15, 2px)";
export var borderRadiusTabsFocusRing = "var(--border-radius-tabs-focus-ring-yky44v, 0px)";
export var borderRadiusTiles = "var(--border-radius-tiles-c0zt4m, 2px)";
export var borderRadiusToken = "var(--border-radius-token-8fyrg1, 2px)";
export var borderRadiusChatBubble = "var(--border-radius-chat-bubble-u484sq, 2px)";
export var borderRadiusTutorialPanelItem = "var(--border-radius-tutorial-panel-item-k7sfhh, 2px)";
export var borderTableStickyWidth = "var(--border-table-sticky-width-vd8s3w, 0px)";
export var borderLinkFocusRingOutline = "var(--border-link-focus-ring-outline-o0qm2f, 5px auto Highlight)";
export var borderLinkFocusRingShadowSpread = "var(--border-link-focus-ring-shadow-spread-ziqmfh, 0px)";
export var borderWidthAlert = "var(--border-width-alert-pz0bqs, 1px)";
export var borderWidthField = "var(--border-width-field-404eki, 1px)";
export var borderWidthPopover = "var(--border-width-popover-clocg5, 1px)";
export var motionDurationExtraFast = "var(--motion-duration-extra-fast-ieas1v, 45ms)";
export var motionDurationExtraSlow = "var(--motion-duration-extra-slow-q9xfa7, 270ms)";
export var motionDurationFast = "var(--motion-duration-fast-3oez8u, 90ms)";
export var motionDurationModerate = "var(--motion-duration-moderate-tnjqaa, 135ms)";
export var motionDurationRefreshOnlyAmbient = "var(--motion-duration-refresh-only-ambient-i5xgpx, 0ms)";
export var motionDurationRefreshOnlyFast = "var(--motion-duration-refresh-only-fast-2lwlmd, 0ms)";
export var motionDurationRefreshOnlyMedium = "var(--motion-duration-refresh-only-medium-l0r8lz, 0ms)";
export var motionDurationRefreshOnlySlow = "var(--motion-duration-refresh-only-slow-c396qe, 0ms)";
export var motionDurationAvatarGenAiGradient = "var(--motion-duration-avatar-gen-ai-gradient-3fx7ww, 3600ms)";
export var motionDurationAvatarLoadingDots = "var(--motion-duration-avatar-loading-dots-ii4xzu, 1200ms)";
export var motionDurationRotate180 = "var(--motion-duration-rotate-180-m9g4a9, 135ms)";
export var motionDurationRotate90 = "var(--motion-duration-rotate-90-jfxxiy, 135ms)";
export var motionDurationShowPaced = "var(--motion-duration-show-paced-uryptc, 180ms)";
export var motionDurationShowQuick = "var(--motion-duration-show-quick-fkbsov, 135ms)";
export var motionDurationSlow = "var(--motion-duration-slow-qk0n8k, 180ms)";
export var motionDurationTransitionQuick = "var(--motion-duration-transition-quick-0viajb, 90ms)";
export var motionDurationTransitionShowPaced = "var(--motion-duration-transition-show-paced-4yudkh, 180ms)";
export var motionDurationTransitionShowQuick = "var(--motion-duration-transition-show-quick-tx10w8, 90ms)";
export var motionEasingEaseOutQuart = "var(--motion-easing-ease-out-quart-aqx2l0, cubic-bezier(0.165, 0.84, 0.44, 1))";
export var motionEasingRefreshOnlyA = "var(--motion-easing-refresh-only-a-6ywz7k, cubic-bezier(0, 0, 0, 1))";
export var motionEasingRefreshOnlyB = "var(--motion-easing-refresh-only-b-uzywy7, cubic-bezier(1, 0, 0.83, 1))";
export var motionEasingRefreshOnlyC = "var(--motion-easing-refresh-only-c-1md9b3, cubic-bezier(0.84, 0, 0.16, 1))";
export var motionEasingRefreshOnlyD = "var(--motion-easing-refresh-only-d-v0itw5, cubic-bezier(0.33, 0, 0.67, 1))";
export var motionEasingAvatarGenAiGradient = "var(--motion-easing-avatar-gen-ai-gradient-o8yfn2, cubic-bezier(0.7, 0, 0.3, 1))";
export var motionEasingRotate180 = "var(--motion-easing-rotate-180-h4cyjx, cubic-bezier(0.165, 0.84, 0.44, 1))";
export var motionEasingRotate90 = "var(--motion-easing-rotate-90-ax5lt7, cubic-bezier(0.165, 0.84, 0.44, 1))";
export var motionEasingShowPaced = "var(--motion-easing-show-paced-tbgeqx, ease-out)";
export var motionEasingShowQuick = "var(--motion-easing-show-quick-17va6f, ease-out)";
export var motionEasingTransitionQuick = "var(--motion-easing-transition-quick-191occ, linear)";
export var motionEasingTransitionShowPaced = "var(--motion-easing-transition-show-paced-47k9w2, ease-out)";
export var motionEasingTransitionShowQuick = "var(--motion-easing-transition-show-quick-0cfmgj, linear)";
export var motionEasingResponsive = "var(--motion-easing-responsive-kehcbg, ease-out)";
export var motionEasingSticky = "var(--motion-easing-sticky-h2dhfe, ease-out)";
export var motionEasingExpressive = "var(--motion-easing-expressive-ahoycd, ease-out)";
export var motionDurationResponsive = "var(--motion-duration-responsive-o555ts, 135ms)";
export var motionDurationExpressive = "var(--motion-duration-expressive-jidfsq, 180ms)";
export var motionDurationComplex = "var(--motion-duration-complex-gcxcdu, 270ms)";
export var motionKeyframesFadeIn = "var(--motion-keyframes-fade-in-2qjwil, awsui-fade-in-35003c)";
export var motionKeyframesFadeOut = "var(--motion-keyframes-fade-out-ddkkh3, awsui-fade-out-35003c)";
export var motionKeyframesStatusIconError = "var(--motion-keyframes-status-icon-error-6klkxp, awsui-none-35003c)";
export var motionKeyframesScalePopup = "var(--motion-keyframes-scale-popup-qv5vbi, awsui-none-35003c)";
export var sizeCalendarGridWidth = "var(--size-calendar-grid-width-awqqfd, 234px)";
export var sizeControl = "var(--size-control-wa1dpj, 14px)";
export var sizeIconBig = "var(--size-icon-big-0l8skx, 32px)";
export var sizeIconLarge = "var(--size-icon-large-jawmww, 48px)";
export var sizeIconMedium = "var(--size-icon-medium-l5o7fk, 16px)";
export var sizeIconNormal = "var(--size-icon-normal-jq6jat, 16px)";
export var sizeTableSelectionHorizontal = "var(--size-table-selection-horizontal-ztdx5v, 54px)";
export var sizeVerticalInput = "var(--size-vertical-input-pta06y, 32px)";
export var sizeVerticalPanelIconOffset = "var(--size-vertical-panel-icon-offset-zd2rli, 15px)";
export var spaceAlertActionLeft = "var(--space-alert-action-left-qthmsu, 20px)";
export var spaceAlertHorizontal = "var(--space-alert-horizontal-zomgmv, 20px)";
export var spaceAlertMessageRight = "var(--space-alert-message-right-5j7pp5, 0px)";
export var spaceAlertVertical = "var(--space-alert-vertical-avo898, 12px)";
export var spaceButtonFocusOutlineGutter = "var(--space-button-focus-outline-gutter-u9axxl, 3px)";
export var spaceButtonHorizontal = "var(--space-button-horizontal-gnfbt4, 20px)";
export var spaceButtonIconFocusOutlineGutterVertical = "var(--space-button-icon-focus-outline-gutter-vertical-qj7cs9, 3px)";
export var spaceButtonIconOnlyHorizontal = "var(--space-button-icon-only-horizontal-reffsr, 16px)";
export var spaceButtonInlineIconFocusOutlineGutter = "var(--space-button-inline-icon-focus-outline-gutter-hidang, 3px)";
export var spaceButtonModalDismissVertical = "var(--space-button-modal-dismiss-vertical-e96ta7, 4px)";
export var spaceCalendarGridFocusOutlineGutter = "var(--space-calendar-grid-focus-outline-gutter-t1skwe, 0px)";
export var spaceCalendarGridSelectedFocusOutlineGutter = "var(--space-calendar-grid-selected-focus-outline-gutter-9bnvp5, 2px)";
export var spaceCalendarGridGutter = "var(--space-calendar-grid-gutter-6q2zbo, 0px)";
export var spaceCardHorizontal = "var(--space-card-horizontal-5nogdm, 20px)";
export var spaceCardVertical = "var(--space-card-vertical-uxiz64, 20px)";
export var spaceCodeEditorStatusFocusOutlineGutter = "var(--space-code-editor-status-focus-outline-gutter-dg55qa, 3px)";
export var spaceContainerContentTop = "var(--space-container-content-top-8wf5md, 16px)";
export var spaceContainerHeaderTop = "var(--space-container-header-top-c6bw9y, 12px)";
export var spaceContainerHeaderBottom = "var(--space-container-header-bottom-83qu8n, 12px)";
export var spaceContainerHorizontal = "var(--space-container-horizontal-jxdgil, 20px)";
export var spaceContentHeaderPaddingBottom = "var(--space-content-header-padding-bottom-3kdnde, 16px)";
export var spaceDarkHeaderOverlapDistance = "var(--space-dark-header-overlap-distance-judax7, 0px)";
export var spaceExpandableSectionIconOffsetTop = "var(--space-expandable-section-icon-offset-top-hnx3nn, 8px)";
export var spaceFieldHorizontal = "var(--space-field-horizontal-zt16s8, 8px)";
export var spaceFieldIconOffset = "var(--space-field-icon-offset-3lrw7i, 32px)";
export var spaceFilteringTokenDismissButtonFocusOutlineGutter = "var(--space-filtering-token-dismiss-button-focus-outline-gutter-zp8mvv, 0px)";
export var spaceFilteringTokenOperationSelectFocusOutlineGutter = "var(--space-filtering-token-operation-select-focus-outline-gutter-ywqus3, 0px)";
export var spaceFlashbarActionLeft = "var(--space-flashbar-action-left-ikwdyl, 16px)";
export var spaceFlashbarDismissRight = "var(--space-flashbar-dismiss-right-b4ikbe, 4px)";
export var spaceFlashbarHorizontal = "var(--space-flashbar-horizontal-k8g0hg, 12px)";
export var spaceGridGutter = "var(--space-grid-gutter-s46zha, 20px)";
export var spaceKeyValueGap = "var(--space-key-value-gap-wtha4n, 2px)";
export var spaceLayoutContentBottom = "var(--space-layout-content-bottom-hy0m3u, 20px)";
export var spaceLayoutContentHorizontal = "var(--space-layout-content-horizontal-yghacl, 40px)";
export var spaceLayoutToggleDiameter = "var(--space-layout-toggle-diameter-c697hn, 36px)";
export var spaceLayoutTogglePadding = "var(--space-layout-toggle-padding-ae8rw5, 12px)";
export var spaceModalContentBottom = "var(--space-modal-content-bottom-7sx2ce, 20px)";
export var spaceModalHorizontal = "var(--space-modal-horizontal-rzqrfo, 20px)";
export var spaceOptionIconBigTop = "var(--space-option-icon-big-top-5d52ea, 2px)";
export var spacePanelContentBottom = "var(--space-panel-content-bottom-s2l12q, 40px)";
export var spacePanelContentTop = "var(--space-panel-content-top-kw8n3a, 20px)";
export var spacePanelDividerMarginHorizontal = "var(--space-panel-divider-margin-horizontal-ouvw5o, 12px)";
export var spacePanelHeaderVertical = "var(--space-panel-header-vertical-zbqp6n, 20px)";
export var spacePanelNavLeft = "var(--space-panel-nav-left-fseqea, 32px)";
export var spacePanelSideLeft = "var(--space-panel-side-left-2amwxl, 32px)";
export var spacePanelSideRight = "var(--space-panel-side-right-lgtbro, 32px)";
export var spacePanelSplitTop = "var(--space-panel-split-top-i0cspk, 0px)";
export var spacePanelSplitBottom = "var(--space-panel-split-bottom-rtddar, 20px)";
export var spaceSegmentedControlFocusOutlineGutter = "var(--space-segmented-control-focus-outline-gutter-ptaps9, 3px)";
export var spaceTabsContentTop = "var(--space-tabs-content-top-u7ubtj, 16px)";
export var spaceTabsFocusOutlineGutter = "var(--space-tabs-focus-outline-gutter-7s1ypx, 0px)";
export var spaceTableContentBottom = "var(--space-table-content-bottom-7je75o, 0px)";
export var spaceTableEmbeddedHeaderTop = "var(--space-table-embedded-header-top-rsxnfq, 12px)";
export var spaceTableFooterHorizontal = "var(--space-table-footer-horizontal-he98bg, 20px)";
export var spaceTableHeaderFocusOutlineGutter = "var(--space-table-header-focus-outline-gutter-rt15yg, 0px)";
export var spaceTableHeaderHorizontal = "var(--space-table-header-horizontal-t0q2vn, 20px)";
export var spaceTableHeaderToolsBottom = "var(--space-table-header-tools-bottom-7eth1u, 4px)";
export var spaceTableHeaderToolsFullPageBottom = "var(--space-table-header-tools-full-page-bottom-4uhg4l, 4px)";
export var spaceTableHorizontal = "var(--space-table-horizontal-7d96o1, 0px)";
export var spaceTileGutter = "var(--space-tile-gutter-z0463f, 24px)";
export var spaceScaled2xNone = "var(--space-scaled-2x-none-p5mfw4, 0px)";
export var spaceScaled2xXxxs = "var(--space-scaled-2x-xxxs-w836iq, 2px)";
export var spaceScaled2xXxs = "var(--space-scaled-2x-xxs-0k2v6n, 4px)";
export var spaceScaled2xXs = "var(--space-scaled-2x-xs-g2id1i, 8px)";
export var spaceScaled2xS = "var(--space-scaled-2x-s-znnwwz, 12px)";
export var spaceScaled2xM = "var(--space-scaled-2x-m-lwig8x, 16px)";
export var spaceScaled2xL = "var(--space-scaled-2x-l-qun137, 20px)";
export var spaceScaled2xXl = "var(--space-scaled-2x-xl-a8w4ze, 24px)";
export var spaceScaled2xXxl = "var(--space-scaled-2x-xxl-4k813d, 32px)";
export var spaceScaled2xXxxl = "var(--space-scaled-2x-xxxl-qc7pg7, 40px)";
export var spaceScaledNone = "var(--space-scaled-none-dlx64l, 0px)";
export var spaceScaledXxxs = "var(--space-scaled-xxxs-jdgipl, 2px)";
export var spaceScaledXxs = "var(--space-scaled-xxs-t2t62i, 4px)";
export var spaceScaledXs = "var(--space-scaled-xs-wbfgrv, 8px)";
export var spaceScaledS = "var(--space-scaled-s-913kwi, 12px)";
export var spaceScaledM = "var(--space-scaled-m-gxhdpl, 16px)";
export var spaceScaledL = "var(--space-scaled-l-2rs0gk, 20px)";
export var spaceScaledXl = "var(--space-scaled-xl-jawob6, 24px)";
export var spaceScaledXxl = "var(--space-scaled-xxl-fzmynl, 32px)";
export var spaceScaledXxxl = "var(--space-scaled-xxxl-crpygs, 40px)";
export var spaceStaticXxxs = "var(--space-static-xxxs-b6tgp7, 2px)";
export var spaceStaticXxs = "var(--space-static-xxs-jqh73n, 4px)";
export var spaceStaticXs = "var(--space-static-xs-zgeph4, 8px)";
export var spaceStaticS = "var(--space-static-s-q0yjjb, 12px)";
export var spaceStaticM = "var(--space-static-m-1plzkw, 16px)";
export var spaceStaticL = "var(--space-static-l-hozrvc, 20px)";
export var spaceStaticXl = "var(--space-static-xl-nmb7ge, 24px)";
export var spaceStaticXxl = "var(--space-static-xxl-qhb0le, 32px)";
export var spaceStaticXxxl = "var(--space-static-xxxl-srpnkv, 40px)";
export var spaceNone = "var(--space-none-w09y0m, 0px)";
export var spaceXxxs = "var(--space-xxxs-j6dpcy, 2px)";
export var spaceXxs = "var(--space-xxs-ja5cp8, 4px)";
export var spaceXs = "var(--space-xs-edba2s, 8px)";
export var spaceS = "var(--space-s-hdd878, 12px)";
export var spaceM = "var(--space-m-h2th94, 16px)";
export var spaceL = "var(--space-l-f4l5gr, 20px)";
export var spaceXl = "var(--space-xl-gsucfg, 24px)";
export var spaceXxl = "var(--space-xxl-sgg5p8, 32px)";
export var spaceXxxl = "var(--space-xxxl-65sweb, 40px)";
export var shadowContainer = "var(--shadow-container-j1l5h7, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowContainerStacked = "var(--shadow-container-stacked-z0y6h2, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowContainerActive = "var(--shadow-container-active-vb9kmq, 0px 4px 8px rgba(0, 28, 36, 0.45))";
export var shadowDropdown = "var(--shadow-dropdown-yqlbut, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowDropup = "var(--shadow-dropup-xkpv46, 0 -1px 1px 0 rgba(0, 28, 36, 0.3), 1px -1px 1px 0 rgba(0, 28, 36, 0.15), -1px -1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowFlashCollapsed = "var(--shadow-flash-collapsed-rsyg0z, 0px 2px 2px rgba(0, 0, 0, 0.15))";
export var shadowFlashSticky = "var(--shadow-flash-sticky-jpgjq1, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowModal = "var(--shadow-modal-knebe7, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowPanel = "var(--shadow-panel-vjg8pn, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowPanelToggle = "var(--shadow-panel-toggle-jqgpi8, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowPopover = "var(--shadow-popover-fogsol, 0px 1px 4px -2px rgba(0, 28, 36, 0.5))";
export var shadowSplitBottom = "var(--shadow-split-bottom-bsye2w, 0 -2px 1px -1px rgba(0, 28, 36, 0.15), 0 -1px 1px -1px rgba(0, 28, 36, 0.3))";
export var shadowSplitSide = "var(--shadow-split-side-e4njpi, 0 1px 1px 0 rgba(0, 28, 36, 0.3), 1px 1px 1px 0 rgba(0, 28, 36, 0.15), -1px 1px 1px 0 rgba(0, 28, 36, 0.15))";
export var shadowSticky = "var(--shadow-sticky-fsddv7, 0px 1px 4px -2px rgba(0, 28, 36, 0.5))";
export var shadowStickyEmbedded = "var(--shadow-sticky-embedded-i1iccv, 0px 1px 4px -2px rgba(0, 28, 36, 0.5))";
export var shadowStickyColumnFirst = "var(--shadow-sticky-column-first-u88e78, 4px 0 8px 0 rgba(0, 28, 36, 0.1))";
export var shadowStickyColumnLast = "var(--shadow-sticky-column-last-avmjwq, -4px 0 8px 0 rgba(0, 28, 36, 0.1))";
