// @synchro-charts/core/dist/collection/testing/test-routes/charts/shaders/status-timeline/single-colored-status.js
import { Component, Element, h } from '@stencil/core';
import { webGLRenderer } from '../../../../../components/sc-webgl-context/webglContext';
import { chartScene } from '../../../../../components/charts/sc-status-timeline/chartScene';
import { CHART_SIZE } from '../chartSize';
import { DAY_IN_MS } from '../../../../../utils/time';
import { HEIGHT } from '../../../../../components/charts/sc-status-timeline/constants';
import { DataType } from '../../../../../utils/dataConstants';
// viewport boundaries
const X_MIN = new Date(2000, 0, 0);
const X_MAX = new Date(2000, 0, 1);
const WIDTH = X_MAX.getTime() - X_MIN.getTime();
const TEST_DATA_POINT_1 = {
    x: X_MIN.getTime() + WIDTH / 3,
    y: 123,
};
export class SingleColoredStatus {
    componentDidLoad() {
        const container = this.el.querySelector('#test-container');
        const scene = chartScene({
            alarms: { expires: DAY_IN_MS },
            viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: 0,
                yMax: HEIGHT,
            },
            dataStreams: [
                {
                    id: 'test-stream',
                    name: 'test-stream-name',
                    color: 'red',
                    aggregates: {
                        [DAY_IN_MS]: [TEST_DATA_POINT_1],
                    },
                    data: [],
                    resolution: DAY_IN_MS,
                    dataType: DataType.NUMBER,
                },
            ],
            container,
            chartSize: CHART_SIZE,
            minBufferSize: 100,
            bufferFactor: 2,
            thresholdOptions: {
                showColor: false,
            },
            thresholds: [],
        });
        webGLRenderer.addChartScene({ manager: scene });
        const rect = container.getBoundingClientRect();
        webGLRenderer.setChartRect(scene.id, Object.assign({ density: 1 }, rect.toJSON()));
    }
    render() {
        return (h("sc-webgl-context", null,
            h("div", { id: "test-container", style: { width: `${CHART_SIZE.width}px`, height: `${CHART_SIZE.height}px` } })));
    }
    static get is() { return "single-colored-status"; }
    static get elementRef() { return "el"; }
}
