// @iot-app-kit/core/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPlugin = exports.registerPlugin = exports.DAY_IN_MS = exports.HOUR_IN_MS = exports.MINUTE_IN_MS = exports.SECOND_IN_MS = exports.NANO_SECOND_IN_MS = exports.STATUS_ICON_TYPE = exports.COMPARISON_OPERATOR = exports.STREAM_TYPE = exports.DATA_TYPE = exports.TimeSeriesDataModule = exports.isEdgeModeEnabled = exports.viewportStartDate = exports.viewportEndDate = exports.isDurationViewport = exports.isHistoricalViewport = exports.viewportManager = exports.parseDuration = exports.toSiteWiseAssetProperty = exports.pointBisector = exports.getDataBeforeDate = exports.getVisibleData = exports.round = exports.combineProviders = void 0;
__exportStar(require("./data-module/data-cache/requestTypes"), exports);
__exportStar(require("./data-module/types"), exports);
__exportStar(require("./common/types"), exports);
var combineProviders_1 = require("./common/combineProviders");
Object.defineProperty(exports, "combineProviders", { enumerable: true, get: function () { return combineProviders_1.combineProviders; } });
var number_1 = require("./common/number");
Object.defineProperty(exports, "round", { enumerable: true, get: function () { return number_1.round; } });
// Data utilities
var dataFilters_1 = require("./common/dataFilters");
Object.defineProperty(exports, "getVisibleData", { enumerable: true, get: function () { return dataFilters_1.getVisibleData; } });
Object.defineProperty(exports, "getDataBeforeDate", { enumerable: true, get: function () { return dataFilters_1.getDataBeforeDate; } });
Object.defineProperty(exports, "pointBisector", { enumerable: true, get: function () { return dataFilters_1.pointBisector; } });
var dataStreamId_1 = require("./common/dataStreamId");
Object.defineProperty(exports, "toSiteWiseAssetProperty", { enumerable: true, get: function () { return dataStreamId_1.toSiteWiseAssetProperty; } });
// Viewport utilities
var time_1 = require("./common/time");
Object.defineProperty(exports, "parseDuration", { enumerable: true, get: function () { return time_1.parseDuration; } });
var viewportManager_1 = require("./viewportManager/viewportManager");
Object.defineProperty(exports, "viewportManager", { enumerable: true, get: function () { return viewportManager_1.viewportManager; } });
var predicates_1 = require("./common/predicates");
Object.defineProperty(exports, "isHistoricalViewport", { enumerable: true, get: function () { return predicates_1.isHistoricalViewport; } });
Object.defineProperty(exports, "isDurationViewport", { enumerable: true, get: function () { return predicates_1.isDurationViewport; } });
var viewport_1 = require("./common/viewport");
Object.defineProperty(exports, "viewportEndDate", { enumerable: true, get: function () { return viewport_1.viewportEndDate; } });
Object.defineProperty(exports, "viewportStartDate", { enumerable: true, get: function () { return viewport_1.viewportStartDate; } });
// Edge utilities
var types_1 = require("./edge/types");
Object.defineProperty(exports, "isEdgeModeEnabled", { enumerable: true, get: function () { return types_1.isEdgeModeEnabled; } });
// Exposed but for internal usage only. Liable to change.
var TimeSeriesDataModule_1 = require("./data-module/TimeSeriesDataModule");
Object.defineProperty(exports, "TimeSeriesDataModule", { enumerable: true, get: function () { return TimeSeriesDataModule_1.TimeSeriesDataModule; } });
var constants_1 = require("./common/constants");
Object.defineProperty(exports, "DATA_TYPE", { enumerable: true, get: function () { return constants_1.DATA_TYPE; } });
Object.defineProperty(exports, "STREAM_TYPE", { enumerable: true, get: function () { return constants_1.STREAM_TYPE; } });
Object.defineProperty(exports, "COMPARISON_OPERATOR", { enumerable: true, get: function () { return constants_1.COMPARISON_OPERATOR; } });
Object.defineProperty(exports, "STATUS_ICON_TYPE", { enumerable: true, get: function () { return constants_1.STATUS_ICON_TYPE; } });
exports.NANO_SECOND_IN_MS = 1 / 1000000;
exports.SECOND_IN_MS = 1000;
exports.MINUTE_IN_MS = 60 * exports.SECOND_IN_MS;
exports.HOUR_IN_MS = 60 * exports.MINUTE_IN_MS;
exports.DAY_IN_MS = 24 * exports.HOUR_IN_MS;
var pluginsRegistry_1 = require("./plugins/pluginsRegistry");
Object.defineProperty(exports, "registerPlugin", { enumerable: true, get: function () { return pluginsRegistry_1.registerPlugin; } });
Object.defineProperty(exports, "getPlugin", { enumerable: true, get: function () { return pluginsRegistry_1.getPlugin; } });
