// @iot-app-kit/core/dist/cjs/mockWidgetProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STRING_STREAM_1 = exports.ALARM_THRESHOLD = exports.ALARM_STREAM = exports.BEFORE_VIEWPORT_DATE = exports.WITHIN_VIEWPORT_DATE = exports.OK = exports.ALARM = exports.DATA_STREAM_2 = exports.DATA_STREAM = exports.NUMBER_STREAM_1 = void 0;
const constants_1 = require("./common/constants");
const time_1 = require("./common/time");
const VIEW_PORT = {
    start: new Date(2000, 0, 0, 0),
    end: new Date(2001, 0, 0, 0),
    yMin: 0,
    yMax: 100,
};
exports.NUMBER_STREAM_1 = {
    id: 'number-some-id',
    color: 'cyan',
    name: 'number-some-name',
    dataType: constants_1.DATA_TYPE.NUMBER,
    resolution: 0,
    data: [
        {
            x: new Date(2000, 0, 0, 0, 0).getTime(),
            y: 100,
        },
    ],
};
exports.DATA_STREAM = {
    id: 'some-asset-id---some-property-id',
    resolution: 0,
    detailedName: 'data-stream-name/detailed-name',
    name: 'data-stream-name',
    color: 'black',
    dataType: constants_1.DATA_TYPE.NUMBER,
    data: [],
};
exports.DATA_STREAM_2 = {
    id: 'some-asset-id-2---some-property-id-2',
    name: 'data-stream-name-2',
    color: 'black',
    resolution: 0,
    dataType: constants_1.DATA_TYPE.NUMBER,
    data: [],
};
exports.ALARM = 'alarm';
exports.OK = 'ok';
exports.WITHIN_VIEWPORT_DATE = new Date(2000, 0, 1);
exports.BEFORE_VIEWPORT_DATE = new Date(VIEW_PORT.start.getTime() - time_1.DAY_IN_MS);
exports.ALARM_STREAM = {
    id: 'alarm-stream',
    dataType: constants_1.DATA_TYPE.STRING,
    name: 'alarm stream',
    color: 'red',
    streamType: constants_1.STREAM_TYPE.ALARM,
    resolution: 0,
    data: [
        {
            x: exports.WITHIN_VIEWPORT_DATE.getTime(),
            y: exports.ALARM,
        },
    ],
};
exports.ALARM_THRESHOLD = {
    value: exports.ALARM,
    color: 'orange',
    comparisonOperator: constants_1.COMPARISON_OPERATOR.EQ,
    icon: constants_1.STATUS_ICON_TYPE.active,
};
exports.STRING_STREAM_1 = {
    id: 'some-string-info',
    dataType: constants_1.DATA_TYPE.STRING,
    color: 'red',
    name: 'some-name',
    resolution: 0,
    data: [
        {
            x: new Date(2000, 0, 0, 0, 0).getTime(),
            y: 'ALARM',
        },
    ],
};
