// @synchro-charts/core/dist/collection/testing/test-routes/charts/shaders/sc-bar-chart/sc-single-colored-bar.js
import { Component, Element, h } from '@stencil/core';
import { webGLRenderer } from '../../../../../components/sc-webgl-context/webglContext';
import { chartScene } from '../../../../../components/charts/sc-bar-chart/chartScene';
import { CHART_SIZE } from '../chartSize';
import { DAY_IN_MS } from '../../../../../utils/time';
import { DataType } from '../../../../../utils/dataConstants';
// viewport boundaries
const X_MIN = new Date(2000, 0, 0);
const X_MAX = new Date(2000, 0, 1);
const Y_MIN = 0;
const Y_MAX = 100;
const WIDTH = X_MAX.getTime() - X_MIN.getTime();
const HEIGHT = Y_MAX - Y_MIN;
const TEST_DATA_POINT_1 = {
    x: new Date(X_MIN.getTime() + WIDTH / 3).getTime(),
    y: Y_MIN + HEIGHT / 2,
};
/**
 * Testing route for the webGL rendering without being fully coupled to the chart.
 *
 * Tests that a single point renders as a colored bar correctly.
 */
export class ScSingleColoredBar {
    componentDidLoad() {
        const container = this.el.querySelector('#test-container');
        const scene = chartScene({
            viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: Y_MIN,
                yMax: Y_MAX,
            },
            dataStreams: [
                {
                    id: 'test-stream',
                    name: 'test-stream-name',
                    color: 'red',
                    resolution: DAY_IN_MS,
                    data: [],
                    aggregates: {
                        [DAY_IN_MS]: [TEST_DATA_POINT_1],
                    },
                    dataType: DataType.NUMBER,
                },
            ],
            container,
            chartSize: CHART_SIZE,
            minBufferSize: 100,
            bufferFactor: 2,
            thresholdOptions: {
                showColor: false,
            },
            thresholds: [],
        });
        webGLRenderer.addChartScene({ manager: scene });
        const rect = container.getBoundingClientRect();
        webGLRenderer.setChartRect(scene.id, Object.assign({ density: 1 }, rect.toJSON()));
    }
    render() {
        return (h("sc-webgl-context", null,
            h("div", { id: "test-container", style: { width: `${CHART_SIZE.width}px`, height: `${CHART_SIZE.height}px` } })));
    }
    static get is() { return "sc-single-colored-bar"; }
    static get elementRef() { return "el"; }
}
