// @awsui/design-tokens/index.js
export var colorChartsRed300 = "var(--color-charts-red-300-23c2oa, #ea7158)";
export var colorChartsRed400 = "var(--color-charts-red-400-539ac2, #dc5032)";
export var colorChartsRed500 = "var(--color-charts-red-500-llkmm5, #d13313)";
export var colorChartsRed600 = "var(--color-charts-red-600-5wdull, #ba2e0f)";
export var colorChartsRed700 = "var(--color-charts-red-700-a6apbj, #a82a0c)";
export var colorChartsRed800 = "var(--color-charts-red-800-yx4mar, #972709)";
export var colorChartsRed900 = "var(--color-charts-red-900-rpnpq8, #892407)";
export var colorChartsRed1000 = "var(--color-charts-red-1000-qk5pc5, #7d2105)";
export var colorChartsRed1100 = "var(--color-charts-red-1100-pt1j9r, #721e03)";
export var colorChartsRed1200 = "var(--color-charts-red-1200-d626tx, #671c00)";
export var colorChartsOrange300 = "var(--color-charts-orange-300-885arv, #e07941)";
export var colorChartsOrange400 = "var(--color-charts-orange-400-8kv4m2, #cc5f21)";
export var colorChartsOrange500 = "var(--color-charts-orange-500-z22zi7, #bc4d01)";
export var colorChartsOrange600 = "var(--color-charts-orange-600-4myt4p, #a84401)";
export var colorChartsOrange700 = "var(--color-charts-orange-700-gaxkk6, #983c02)";
export var colorChartsOrange800 = "var(--color-charts-orange-800-io45li, #8a3603)";
export var colorChartsOrange900 = "var(--color-charts-orange-900-rz3zc9, #7e3103)";
export var colorChartsOrange1000 = "var(--color-charts-orange-1000-0qpjgf, #732c02)";
export var colorChartsOrange1100 = "var(--color-charts-orange-1100-grw837, #692801)";
export var colorChartsOrange1200 = "var(--color-charts-orange-1200-5nnge9, #602400)";
export var colorChartsYellow300 = "var(--color-charts-yellow-300-6c0xtx, #b2911c)";
export var colorChartsYellow400 = "var(--color-charts-yellow-400-u252ao, #9c7b0b)";
export var colorChartsYellow500 = "var(--color-charts-yellow-500-v2phgx, #8a6b05)";
export var colorChartsYellow600 = "var(--color-charts-yellow-600-4pv3hy, #7b5f04)";
export var colorChartsYellow700 = "var(--color-charts-yellow-700-3emu4t, #6f5504)";
export var colorChartsYellow800 = "var(--color-charts-yellow-800-edh0l0, #654d03)";
export var colorChartsYellow900 = "var(--color-charts-yellow-900-1zoghi, #5d4503)";
export var colorChartsYellow1000 = "var(--color-charts-yellow-1000-4ypqyp, #553f03)";
export var colorChartsYellow1100 = "var(--color-charts-yellow-1100-2bbrrm, #4d3901)";
export var colorChartsYellow1200 = "var(--color-charts-yellow-1200-pml6h8, #483300)";
export var colorChartsGreen300 = "var(--color-charts-green-300-xdy00p, #67a353)";
export var colorChartsGreen400 = "var(--color-charts-green-400-pyju52, #41902c)";
export var colorChartsGreen500 = "var(--color-charts-green-500-x1jom3, #1f8104)";
export var colorChartsGreen600 = "var(--color-charts-green-600-ax7yum, #1a7302)";
export var colorChartsGreen700 = "var(--color-charts-green-700-x1ubuk, #176702)";
export var colorChartsGreen800 = "var(--color-charts-green-800-erepp2, #145d02)";
export var colorChartsGreen900 = "var(--color-charts-green-900-jgvtf4, #125502)";
export var colorChartsGreen1000 = "var(--color-charts-green-1000-ftsnde, #104d01)";
export var colorChartsGreen1100 = "var(--color-charts-green-1100-8is059, #0f4601)";
export var colorChartsGreen1200 = "var(--color-charts-green-1200-pyehtq, #0d4000)";
export var colorChartsTeal300 = "var(--color-charts-teal-300-ze50a3, #2ea597)";
export var colorChartsTeal400 = "var(--color-charts-teal-400-eho3qj, #1c8e81)";
export var colorChartsTeal500 = "var(--color-charts-teal-500-ysfodg, #0d7d70)";
export var colorChartsTeal600 = "var(--color-charts-teal-600-k0lr0d, #096f64)";
export var colorChartsTeal700 = "var(--color-charts-teal-700-pradbd, #06645a)";
export var colorChartsTeal800 = "var(--color-charts-teal-800-l98pd3, #045b52)";
export var colorChartsTeal900 = "var(--color-charts-teal-900-2n2uqa, #03524a)";
export var colorChartsTeal1000 = "var(--color-charts-teal-1000-sqzv15, #014b44)";
export var colorChartsTeal1100 = "var(--color-charts-teal-1100-ugmadm, #01443e)";
export var colorChartsTeal1200 = "var(--color-charts-teal-1200-u1l5vv, #003e38)";
export var colorChartsBlue1300 = "var(--color-charts-blue-1-300-bhq5nc, #529ccb)";
export var colorChartsBlue1400 = "var(--color-charts-blue-1-400-6shq6z, #3184c2)";
export var colorChartsBlue1500 = "var(--color-charts-blue-1-500-u9fbb4, #0273bb)";
export var colorChartsBlue1600 = "var(--color-charts-blue-1-600-6cm0lb, #0166ab)";
export var colorChartsBlue1700 = "var(--color-charts-blue-1-700-pur7e5, #015b9d)";
export var colorChartsBlue1800 = "var(--color-charts-blue-1-800-d90xfc, #015292)";
export var colorChartsBlue1900 = "var(--color-charts-blue-1-900-ugbzx8, #014a87)";
export var colorChartsBlue11000 = "var(--color-charts-blue-1-1000-wn8i30, #01437d)";
export var colorChartsBlue11100 = "var(--color-charts-blue-1-1100-zh7t8t, #003c75)";
export var colorChartsBlue11200 = "var(--color-charts-blue-1-1200-7spkcj, #00366d)";
export var colorChartsBlue2300 = "var(--color-charts-blue-2-300-25paew, #688ae8)";
export var colorChartsBlue2400 = "var(--color-charts-blue-2-400-ppuj2w, #5978e3)";
export var colorChartsBlue2500 = "var(--color-charts-blue-2-500-30q04l, #4066df)";
export var colorChartsBlue2600 = "var(--color-charts-blue-2-600-bqdh5s, #3759ce)";
export var colorChartsBlue2700 = "var(--color-charts-blue-2-700-useohu, #314fbf)";
export var colorChartsBlue2800 = "var(--color-charts-blue-2-800-vbehwc, #2c46b1)";
export var colorChartsBlue2900 = "var(--color-charts-blue-2-900-rjxjr8, #273ea5)";
export var colorChartsBlue21000 = "var(--color-charts-blue-2-1000-fo6yfn, #23379b)";
export var colorChartsBlue21100 = "var(--color-charts-blue-2-1100-r56l0z, #1f3191)";
export var colorChartsBlue21200 = "var(--color-charts-blue-2-1200-mrmzh5, #1b2b88)";
export var colorChartsPurple300 = "var(--color-charts-purple-300-fks7nd, #a783e1)";
export var colorChartsPurple400 = "var(--color-charts-purple-400-g5p5yw, #9469d6)";
export var colorChartsPurple500 = "var(--color-charts-purple-500-teqjg9, #8456ce)";
export var colorChartsPurple600 = "var(--color-charts-purple-600-nf5r9n, #7749bf)";
export var colorChartsPurple700 = "var(--color-charts-purple-700-15kpqx, #6b40b2)";
export var colorChartsPurple800 = "var(--color-charts-purple-800-fskedv, #6237a7)";
export var colorChartsPurple900 = "var(--color-charts-purple-900-9ytuxe, #59309d)";
export var colorChartsPurple1000 = "var(--color-charts-purple-1000-yxj5hz, #512994)";
export var colorChartsPurple1100 = "var(--color-charts-purple-1100-meod30, #4a238b)";
export var colorChartsPurple1200 = "var(--color-charts-purple-1200-d8mpzv, #431d84)";
export var colorChartsPink300 = "var(--color-charts-pink-300-9pujzo, #da7596)";
export var colorChartsPink400 = "var(--color-charts-pink-400-s9k4qh, #ce567c)";
export var colorChartsPink500 = "var(--color-charts-pink-500-pic0fb, #c33d69)";
export var colorChartsPink600 = "var(--color-charts-pink-600-is3nog, #b1325c)";
export var colorChartsPink700 = "var(--color-charts-pink-700-wmp2jn, #a32952)";
export var colorChartsPink800 = "var(--color-charts-pink-800-u9xp2q, #962249)";
export var colorChartsPink900 = "var(--color-charts-pink-900-171x5m, #8b1b42)";
export var colorChartsPink1000 = "var(--color-charts-pink-1000-cm89wx, #81143b)";
export var colorChartsPink1100 = "var(--color-charts-pink-1100-0bdx0v, #780d35)";
export var colorChartsPink1200 = "var(--color-charts-pink-1200-af67f8, #6f062f)";
export var colorChartsStatusCritical = "var(--color-charts-status-critical-81ggtj, #7d2105)";
export var colorChartsStatusHigh = "var(--color-charts-status-high-qe7s3z, #ba2e0f)";
export var colorChartsStatusMedium = "var(--color-charts-status-medium-a45eb3, #cc5f21)";
export var colorChartsStatusLow = "var(--color-charts-status-low-vohc9l, #b2911c)";
export var colorChartsStatusPositive = "var(--color-charts-status-positive-bswwz8, #67a353)";
export var colorChartsStatusInfo = "var(--color-charts-status-info-97ztp9, #3184c2)";
export var colorChartsStatusNeutral = "var(--color-charts-status-neutral-ik95f5, #879596)";
export var colorChartsThresholdNegative = "var(--color-charts-threshold-negative-adabc7, #d13212)";
export var colorChartsThresholdPositive = "var(--color-charts-threshold-positive-rtbub9, #1d8102)";
export var colorChartsThresholdInfo = "var(--color-charts-threshold-info-3592ya, #0073bb)";
export var colorChartsThresholdNeutral = "var(--color-charts-threshold-neutral-h58y4n, #687078)";
export var colorChartsLineGrid = "var(--color-charts-line-grid-b5jia7, #d5dbdb)";
export var colorChartsLineTick = "var(--color-charts-line-tick-6u2dih, #d5dbdb)";
export var colorChartsLineAxis = "var(--color-charts-line-axis-6cxsbx, #d5dbdb)";
export var colorChartsPaletteCategorical1 = "var(--color-charts-palette-categorical-1-s02edj, #688ae8)";
export var colorChartsPaletteCategorical2 = "var(--color-charts-palette-categorical-2-hd7uud, #c33d69)";
export var colorChartsPaletteCategorical3 = "var(--color-charts-palette-categorical-3-tm6hrf, #2ea597)";
export var colorChartsPaletteCategorical4 = "var(--color-charts-palette-categorical-4-fqnj32, #8456ce)";
export var colorChartsPaletteCategorical5 = "var(--color-charts-palette-categorical-5-h9jmtt, #e07941)";
export var colorChartsPaletteCategorical6 = "var(--color-charts-palette-categorical-6-z7e3fr, #3759ce)";
export var colorChartsPaletteCategorical7 = "var(--color-charts-palette-categorical-7-4w1lrz, #962249)";
export var colorChartsPaletteCategorical8 = "var(--color-charts-palette-categorical-8-xesix1, #096f64)";
export var colorChartsPaletteCategorical9 = "var(--color-charts-palette-categorical-9-93ht36, #6237a7)";
export var colorChartsPaletteCategorical10 = "var(--color-charts-palette-categorical-10-wus8nq, #a84401)";
export var colorChartsPaletteCategorical11 = "var(--color-charts-palette-categorical-11-uh7roa, #273ea5)";
export var colorChartsPaletteCategorical12 = "var(--color-charts-palette-categorical-12-ywqdlz, #780d35)";
export var colorChartsPaletteCategorical13 = "var(--color-charts-palette-categorical-13-0aws3q, #03524a)";
export var colorChartsPaletteCategorical14 = "var(--color-charts-palette-categorical-14-5l0pt5, #4a238b)";
export var colorChartsPaletteCategorical15 = "var(--color-charts-palette-categorical-15-glpjzo, #7e3103)";
export var colorChartsPaletteCategorical16 = "var(--color-charts-palette-categorical-16-73ijrp, #1b2b88)";
export var colorChartsPaletteCategorical17 = "var(--color-charts-palette-categorical-17-ltabw6, #ce567c)";
export var colorChartsPaletteCategorical18 = "var(--color-charts-palette-categorical-18-fog2ah, #003e38)";
export var colorChartsPaletteCategorical19 = "var(--color-charts-palette-categorical-19-ig4cn5, #9469d6)";
export var colorChartsPaletteCategorical20 = "var(--color-charts-palette-categorical-20-ta856w, #602400)";
export var colorChartsPaletteCategorical21 = "var(--color-charts-palette-categorical-21-ta0h08, #4066df)";
export var colorChartsPaletteCategorical22 = "var(--color-charts-palette-categorical-22-lj462v, #a32952)";
export var colorChartsPaletteCategorical23 = "var(--color-charts-palette-categorical-23-35jyqu, #0d7d70)";
export var colorChartsPaletteCategorical24 = "var(--color-charts-palette-categorical-24-eem6bw, #6b40b2)";
export var colorChartsPaletteCategorical25 = "var(--color-charts-palette-categorical-25-w302xt, #bc4d01)";
export var colorChartsPaletteCategorical26 = "var(--color-charts-palette-categorical-26-6ughlw, #2c46b1)";
export var colorChartsPaletteCategorical27 = "var(--color-charts-palette-categorical-27-57l9xa, #81143b)";
export var colorChartsPaletteCategorical28 = "var(--color-charts-palette-categorical-28-4q0mir, #045b52)";
export var colorChartsPaletteCategorical29 = "var(--color-charts-palette-categorical-29-cnn7dg, #512994)";
export var colorChartsPaletteCategorical30 = "var(--color-charts-palette-categorical-30-eix60t, #8a3603)";
export var colorChartsPaletteCategorical31 = "var(--color-charts-palette-categorical-31-jthnv1, #1f3191)";
export var colorChartsPaletteCategorical32 = "var(--color-charts-palette-categorical-32-h40wuu, #da7596)";
export var colorChartsPaletteCategorical33 = "var(--color-charts-palette-categorical-33-1iyt0i, #01443e)";
export var colorChartsPaletteCategorical34 = "var(--color-charts-palette-categorical-34-eut0yx, #a783e1)";
export var colorChartsPaletteCategorical35 = "var(--color-charts-palette-categorical-35-utj4hw, #692801)";
export var colorChartsPaletteCategorical36 = "var(--color-charts-palette-categorical-36-mjb3b0, #5978e3)";
export var colorChartsPaletteCategorical37 = "var(--color-charts-palette-categorical-37-6rz96d, #b1325c)";
export var colorChartsPaletteCategorical38 = "var(--color-charts-palette-categorical-38-s8j0ph, #1c8e81)";
export var colorChartsPaletteCategorical39 = "var(--color-charts-palette-categorical-39-5l5hgh, #7749bf)";
export var colorChartsPaletteCategorical40 = "var(--color-charts-palette-categorical-40-2i7lp3, #cc5f21)";
export var colorChartsPaletteCategorical41 = "var(--color-charts-palette-categorical-41-lozcy4, #314fbf)";
export var colorChartsPaletteCategorical42 = "var(--color-charts-palette-categorical-42-qzrtzy, #8b1b42)";
export var colorChartsPaletteCategorical43 = "var(--color-charts-palette-categorical-43-sq9ksf, #06645a)";
export var colorChartsPaletteCategorical44 = "var(--color-charts-palette-categorical-44-2mlngi, #59309d)";
export var colorChartsPaletteCategorical45 = "var(--color-charts-palette-categorical-45-irkhaj, #983c02)";
export var colorChartsPaletteCategorical46 = "var(--color-charts-palette-categorical-46-r0hukp, #23379b)";
export var colorChartsPaletteCategorical47 = "var(--color-charts-palette-categorical-47-l2wh9o, #6f062f)";
export var colorChartsPaletteCategorical48 = "var(--color-charts-palette-categorical-48-k8gno4, #014b44)";
export var colorChartsPaletteCategorical49 = "var(--color-charts-palette-categorical-49-e507mj, #431d84)";
export var colorChartsPaletteCategorical50 = "var(--color-charts-palette-categorical-50-hfws7f, #732c02)";
export var colorBackgroundNotificationSeverityCritical = "var(--color-background-notification-severity-critical-ib32sw, #870303)";
export var colorBackgroundNotificationSeverityHigh = "var(--color-background-notification-severity-high-gv1pxl, #ce3311)";
export var colorBackgroundNotificationSeverityMedium = "var(--color-background-notification-severity-medium-hghstn, #f89256)";
export var colorBackgroundNotificationSeverityLow = "var(--color-background-notification-severity-low-asztnl, #f2cd54)";
export var colorBackgroundNotificationSeverityNeutral = "var(--color-background-notification-severity-neutral-1tkwxg, #687078)";
export var colorTextNotificationSeverityCritical = "var(--color-text-notification-severity-critical-eowkv6, #fafafa)";
export var colorTextNotificationSeverityHigh = "var(--color-text-notification-severity-high-oant7z, #fafafa)";
export var colorTextNotificationSeverityMedium = "var(--color-text-notification-severity-medium-hglkx1, #000000)";
export var colorTextNotificationSeverityLow = "var(--color-text-notification-severity-low-nq2syi, #16191f)";
export var colorTextNotificationSeverityNeutral = "var(--color-text-notification-severity-neutral-1948do, #fafafa)";
export var colorBackgroundButtonNormalActive = "var(--color-background-button-normal-active-wuxvsv, #eaeded)";
export var colorBackgroundButtonNormalDefault = "var(--color-background-button-normal-default-5fdgc5, #ffffff)";
export var colorBackgroundButtonNormalDisabled = "var(--color-background-button-normal-disabled-pzswm7, #ffffff)";
export var colorBackgroundButtonNormalHover = "var(--color-background-button-normal-hover-odla4l, #fafafa)";
export var colorBackgroundToggleButtonNormalPressed = "var(--color-background-toggle-button-normal-pressed-dc1nqz, #eaeded)";
export var colorBackgroundButtonPrimaryActive = "var(--color-background-button-primary-active-mcw2z2, #12293b)";
export var colorBackgroundButtonPrimaryDefault = "var(--color-background-button-primary-default-sdrd7j, #0073bb)";
export var colorBackgroundButtonPrimaryDisabled = "var(--color-background-button-primary-disabled-57e0jj, #ffffff)";
export var colorBackgroundButtonPrimaryHover = "var(--color-background-button-primary-hover-yt5487, #0a4a74)";
export var colorBackgroundCellShaded = "var(--color-background-cell-shaded-47masj, #f2f3f3)";
export var colorBackgroundContainerContent = "var(--color-background-container-content-cpvvzs, #ffffff)";
export var colorBackgroundContainerHeader = "var(--color-background-container-header-1edgmi, #fafafa)";
export var colorBackgroundControlChecked = "var(--color-background-control-checked-6j23t6, #0073bb)";
export var colorBackgroundControlDefault = "var(--color-background-control-default-6lirk9, #ffffff)";
export var colorBackgroundControlDisabled = "var(--color-background-control-disabled-h93c4j, #d5dbdb)";
export var colorBackgroundDropdownItemDefault = "var(--color-background-dropdown-item-default-q1rj8l, #ffffff)";
export var colorBackgroundDropdownItemFilterMatch = "var(--color-background-dropdown-item-filter-match-u1g6p2, #f1faff)";
export var colorBackgroundDropdownItemHover = "var(--color-background-dropdown-item-hover-6zmazg, #f2f3f3)";
export var colorBackgroundHomeHeader = "var(--color-background-home-header-q7il70, #000000)";
export var colorBackgroundInputDefault = "var(--color-background-input-default-xowsm5, #ffffff)";
export var colorBackgroundInputDisabled = "var(--color-background-input-disabled-fib0qr, #eaeded)";
export var colorBackgroundItemSelected = "var(--color-background-item-selected-bc67ge, #f1faff)";
export var colorBackgroundLayoutMain = "var(--color-background-layout-main-f3jtyd, #f2f3f3)";
export var colorBackgroundLayoutToggleActive = "var(--color-background-layout-toggle-active-9rnqxf, #2a2e33)";
export var colorBackgroundLayoutToggleDefault = "var(--color-background-layout-toggle-default-b3e5ga, transparent)";
export var colorBackgroundLayoutToggleHover = "var(--color-background-layout-toggle-hover-zeuoxl, #545b64)";
export var colorBackgroundLayoutToggleSelectedActive = "var(--color-background-layout-toggle-selected-active-67ljrt, #0073bb)";
export var colorBackgroundLayoutToggleSelectedDefault = "var(--color-background-layout-toggle-selected-default-tq3fmx, #0073bb)";
export var colorBackgroundLayoutToggleSelectedHover = "var(--color-background-layout-toggle-selected-hover-1zcs0k, #0a4a74)";
export var colorBackgroundNotificationBlue = "var(--color-background-notification-blue-jz7lex, #0073bb)";
export var colorBackgroundNotificationGreen = "var(--color-background-notification-green-0jtiqi, #1d8102)";
export var colorBackgroundNotificationGrey = "var(--color-background-notification-grey-5thu1e, #545b64)";
export var colorBackgroundNotificationRed = "var(--color-background-notification-red-05rvmd, #d13212)";
export var colorBackgroundNotificationYellow = "var(--color-background-notification-yellow-vv98ba, #ffe347)";
export var colorBackgroundPopover = "var(--color-background-popover-q6ltpd, #ffffff)";
export var colorBackgroundSegmentActive = "var(--color-background-segment-active-73phfs, #0073bb)";
export var colorBackgroundSegmentDefault = "var(--color-background-segment-default-akl4mj, #ffffff)";
export var colorBackgroundSegmentDisabled = "var(--color-background-segment-disabled-4h6zm5, #ffffff)";
export var colorBackgroundSegmentHover = "var(--color-background-segment-hover-mpe3ui, #ffffff)";
export var colorBackgroundStatusError = "var(--color-background-status-error-mqjczn, #fdf3f1)";
export var colorBackgroundStatusInfo = "var(--color-background-status-info-07a262, #f1faff)";
export var colorBackgroundStatusSuccess = "var(--color-background-status-success-to45h8, #f2f8f0)";
export var colorBackgroundStatusWarning = "var(--color-background-status-warning-8czyg1, #fffef0)";
export var colorBackgroundToggleCheckedDisabled = "var(--color-background-toggle-checked-disabled-jkq3ea, #99cbe4)";
export var colorBackgroundAvatarGenAi = "var(--color-background-avatar-gen-ai-eq7zlc, radial-gradient(circle farthest-corner at top left,rgba(0, 150, 250, 1) -25%,rgba(0, 150, 250, 0) 55%),radial-gradient(circle farthest-corner at top right, rgba(216, 178, 255, 1) -10%, rgba(115, 0, 229, 1) 50%))";
export var colorBackgroundAvatarDefault = "var(--color-background-avatar-default-xypvc3, #545b64)";
export var colorTextAvatar = "var(--color-text-avatar-y5n3t1, #ffffff)";
export var colorBackgroundLoadingBarGenAi = "var(--color-background-loading-bar-gen-ai-nbwy81, linear-gradient(90deg, #99f7ff 0%, #0096fa 10%, #bf80ff 24%, #7300e5 50%, #bf80ff 76%, #0096fa 90%, #99f7ff 100%))";
export var colorBackgroundChatBubbleOutgoing = "var(--color-background-chat-bubble-outgoing-7qngnp, transparent)";
export var colorBackgroundChatBubbleIncoming = "var(--color-background-chat-bubble-incoming-04glkh, #f2f3f3)";
export var colorTextChatBubbleOutgoing = "var(--color-text-chat-bubble-outgoing-79xrhd, #16191f)";
export var colorTextChatBubbleIncoming = "var(--color-text-chat-bubble-incoming-72165z, #16191f)";
export var colorBorderButtonNormalActive = "var(--color-border-button-normal-active-vkwe9f, #545b64)";
export var colorBorderButtonNormalDefault = "var(--color-border-button-normal-default-9jkxwh, #545b64)";
export var colorBorderToggleButtonNormalPressed = "var(--color-border-toggle-button-normal-pressed-c5n6wh, #545b64)";
export var colorBorderButtonNormalDisabled = "var(--color-border-button-normal-disabled-orwwg4, #d5dbdb)";
export var colorTextButtonNormalDisabled = "var(--color-text-button-normal-disabled-u9dsmh, #879596)";
export var colorBorderButtonNormalHover = "var(--color-border-button-normal-hover-ap4j3a, #16191f)";
export var colorBorderButtonPrimaryDisabled = "var(--color-border-button-primary-disabled-2hg9e9, #d5dbdb)";
export var colorTextButtonPrimaryDisabled = "var(--color-text-button-primary-disabled-a4619v, #879596)";
export var colorBorderContainerTop = "var(--color-border-container-top-wiud2r, #eaeded)";
export var colorBorderControlDefault = "var(--color-border-control-default-nj5wiz, #687078)";
export var colorBorderDividerDefault = "var(--color-border-divider-default-g2zqci, #eaeded)";
export var colorBorderDividerSecondary = "var(--color-border-divider-secondary-qt0aq9, #eaeded)";
export var colorBorderDropdownItemHover = "var(--color-border-dropdown-item-hover-8pzdhc, #879596)";
export var colorBorderInputDefault = "var(--color-border-input-default-ijkrzo, #687078)";
export var colorBorderInputFocused = "var(--color-border-input-focused-5893wz, #0073bb)";
export var colorBorderItemFocused = "var(--color-border-item-focused-j88ehv, #0073bb)";
export var colorBorderDropdownItemFocused = "var(--color-border-dropdown-item-focused-hebyze, #0073bb)";
export var colorBorderItemSelected = "var(--color-border-item-selected-xu3hwl, #0073bb)";
export var colorBorderSegmentActive = "var(--color-border-segment-active-mv1bbg, #16191f)";
export var colorBorderSegmentDefault = "var(--color-border-segment-default-4q3z1n, #687078)";
export var colorBorderSegmentDisabled = "var(--color-border-segment-disabled-vpnkqa, #d5dbdb)";
export var colorBorderSegmentHover = "var(--color-border-segment-hover-x2c7ym, #16191f)";
export var colorBorderStatusError = "var(--color-border-status-error-rw7xpm, #d13212)";
export var colorBorderStatusInfo = "var(--color-border-status-info-ytrrza, #0073bb)";
export var colorBorderStatusSuccess = "var(--color-border-status-success-re24ca, #1d8102)";
export var colorBorderStatusWarning = "var(--color-border-status-warning-absgf7, #906806)";
export var colorForegroundControlDefault = "var(--color-foreground-control-default-2lrfrr, #ffffff)";
export var colorForegroundControlDisabled = "var(--color-foreground-control-disabled-dksqnz, #ffffff)";
export var colorForegroundControlReadOnly = "var(--color-foreground-control-read-only-4cwci1, #687078)";
export var colorTextAccent = "var(--color-text-accent-27s5l3, #0073bb)";
export var colorTextBodyDefault = "var(--color-text-body-default-zs35hz, #16191f)";
export var colorTextBodySecondary = "var(--color-text-body-secondary-kc95tq, #545b64)";
export var colorTextBreadcrumbCurrent = "var(--color-text-breadcrumb-current-kt41ir, #687078)";
export var colorTextBreadcrumbIcon = "var(--color-text-breadcrumb-icon-8qiz1d, #687078)";
export var colorTextButtonNormalActive = "var(--color-text-button-normal-active-nhgenz, #16191f)";
export var colorTextToggleButtonNormalPressed = "var(--color-text-toggle-button-normal-pressed-mmq27n, #16191f)";
export var colorTextButtonNormalDefault = "var(--color-text-button-normal-default-mc7t60, #545b64)";
export var colorTextButtonNormalHover = "var(--color-text-button-normal-hover-bki2l2, #16191f)";
export var colorTextButtonPrimaryActive = "var(--color-text-button-primary-active-0gc6a4, #ffffff)";
export var colorTextButtonPrimaryDefault = "var(--color-text-button-primary-default-v7x7uq, #ffffff)";
export var colorTextButtonPrimaryHover = "var(--color-text-button-primary-hover-imb9dy, #ffffff)";
export var colorTextCounter = "var(--color-text-counter-a9cq2m, #687078)";
export var colorTextDropdownItemDefault = "var(--color-text-dropdown-item-default-ti6pjc, #16191f)";
export var colorTextDropdownItemDisabled = "var(--color-text-dropdown-item-disabled-j9pzby, #aab7b8)";
export var colorTextDropdownItemFilterMatch = "var(--color-text-dropdown-item-filter-match-62fflr, #0073bb)";
export var colorTextDropdownItemHighlighted = "var(--color-text-dropdown-item-highlighted-0ns2tb, #16191f)";
export var colorTextEmpty = "var(--color-text-empty-mn4nrj, #687078)";
export var colorTextFormDefault = "var(--color-text-form-default-b48udf, #16191f)";
export var colorTextFormSecondary = "var(--color-text-form-secondary-6vvbul, #687078)";
export var colorTextGroupLabel = "var(--color-text-group-label-37uws0, #545b64)";
export var colorTextLabelGenAi = "var(--color-text-label-gen-ai-6dxzrs, #7300e5)";
export var colorTextHeadingDefault = "var(--color-text-heading-default-wliqyu, #16191f)";
export var colorTextHeadingSecondary = "var(--color-text-heading-secondary-urdy5f, #545b64)";
export var colorTextHomeHeaderDefault = "var(--color-text-home-header-default-gz3zbq, #ffffff)";
export var colorTextHomeHeaderSecondary = "var(--color-text-home-header-secondary-v8lrx6, #d5dbdb)";
export var colorTextInputDisabled = "var(--color-text-input-disabled-scisz5, #879596)";
export var colorTextInputPlaceholder = "var(--color-text-input-placeholder-bclxfj, #687078)";
export var colorTextInteractiveActive = "var(--color-text-interactive-active-ifmuec, #16191f)";
export var colorTextInteractiveDefault = "var(--color-text-interactive-default-yjja72, #545b64)";
export var colorTextInteractiveDisabled = "var(--color-text-interactive-disabled-jwju02, #aab7b8)";
export var colorTextInteractiveHover = "var(--color-text-interactive-hover-fgfl9g, #16191f)";
export var colorTextToggleButtonIconPressed = "var(--color-text-toggle-button-icon-pressed-31hcbc, #16191f)";
export var colorTextInteractiveInvertedDefault = "var(--color-text-interactive-inverted-default-wi2ail, #d5dbdb)";
export var colorTextInteractiveInvertedHover = "var(--color-text-interactive-inverted-hover-hf0cki, #fafafa)";
export var colorTextLabel = "var(--color-text-label-du7fzg, #545b64)";
export var colorTextLayoutToggle = "var(--color-text-layout-toggle-sucjzk, #545b64)";
export var colorTextLayoutToggleActive = "var(--color-text-layout-toggle-active-g2y9le, #ffffff)";
export var colorTextLayoutToggleHover = "var(--color-text-layout-toggle-hover-1lm7x2, #0073bb)";
export var colorTextLayoutToggleSelected = "var(--color-text-layout-toggle-selected-9msxpj, #ffffff)";
export var colorTextLinkDefault = "var(--color-text-link-default-owstl4, #0073bb)";
export var colorTextLinkHover = "var(--color-text-link-hover-aohp90, #0a4a74)";
export var colorTextNotificationDefault = "var(--color-text-notification-default-zu8c4i, #fafafa)";
export var colorTextSegmentActive = "var(--color-text-segment-active-4d27ka, #ffffff)";
export var colorTextSegmentDefault = "var(--color-text-segment-default-48d69d, #545b64)";
export var colorTextSegmentHover = "var(--color-text-segment-hover-u9d76i, #0073bb)";
export var colorTextStatusError = "var(--color-text-status-error-oa993i, #d13212)";
export var colorTextStatusInactive = "var(--color-text-status-inactive-97nxyu, #687078)";
export var colorTextStatusInfo = "var(--color-text-status-info-vkd30u, #0073bb)";
export var colorTextStatusSuccess = "var(--color-text-status-success-z431uf, #1d8102)";
export var colorTextStatusWarning = "var(--color-text-status-warning-a2jajm, #906806)";
export var colorTextTopNavigationTitle = "var(--color-text-top-navigation-title-fzc955, #16191f)";
export var colorBoardPlaceholderActive = "var(--color-board-placeholder-active-ua4ik1, #d5dbdb)";
export var colorBoardPlaceholderHover = "var(--color-board-placeholder-hover-jc1c35, #99cbe4)";
export var colorDragPlaceholderActive = "var(--color-drag-placeholder-active-smas3a, #d5dbdb)";
export var colorDragPlaceholderHover = "var(--color-drag-placeholder-hover-0mqnqu, #99cbe4)";
export var fontFamilyBase = "var(--font-family-base-5z8fw5, 'Noto Sans', 'Helvetica Neue', Roboto, Arial, sans-serif)";
export var fontFamilyMonospace = "var(--font-family-monospace-froq9j, Monaco, Menlo, Consolas, 'Courier Prime', Courier, 'Courier New', monospace)";
export var fontSizeBodyM = "var(--font-size-body-m-kbh530, 14px)";
export var fontSizeBodyS = "var(--font-size-body-s-6sh5cc, 12px)";
export var fontSizeDisplayL = "var(--font-size-display-l-zttzwo, 44px)";
export var fontSizeHeadingXl = "var(--font-size-heading-xl-e8bqij, 28px)";
export var fontSizeHeadingL = "var(--font-size-heading-l-04luh4, 18px)";
export var fontSizeHeadingM = "var(--font-size-heading-m-9xjzrj, 18px)";
export var fontSizeHeadingS = "var(--font-size-heading-s-d8ts7u, 16px)";
export var fontSizeHeadingXs = "var(--font-size-heading-xs-pgspul, 16px)";
export var fontWeightButton = "var(--font-weight-button-rb1vba, 700)";
export var fontWeightHeadingXl = "var(--font-weight-heading-xl-nre0k2, 400)";
export var fontWeightHeadingL = "var(--font-weight-heading-l-0lcea3, 700)";
export var fontWeightHeadingM = "var(--font-weight-heading-m-0c5l08, 400)";
export var fontWeightHeadingS = "var(--font-weight-heading-s-6rdhmg, 700)";
export var fontWeightHeadingXs = "var(--font-weight-heading-xs-xuskfi, 400)";
export var lineHeightBodyM = "var(--line-height-body-m-0awzkk, 22px)";
export var lineHeightBodyS = "var(--line-height-body-s-usc3fr, 16px)";
export var lineHeightDisplayL = "var(--line-height-display-l-0oa5q5, 56px)";
export var lineHeightHeadingXl = "var(--line-height-heading-xl-gnbljj, 36px)";
export var lineHeightHeadingL = "var(--line-height-heading-l-4wuokq, 22px)";
export var lineHeightHeadingM = "var(--line-height-heading-m-gv8yla, 22px)";
export var lineHeightHeadingS = "var(--line-height-heading-s-i8coai, 20px)";
export var lineHeightHeadingXs = "var(--line-height-heading-xs-8t7s1x, 20px)";
export var borderRadiusAlert = "var(--border-radius-alert-8j75to, 2px)";
export var borderRadiusBadge = "var(--border-radius-badge-5bxe5h, 16px)";
export var borderRadiusButton = "var(--border-radius-button-f5klif, 2px)";
export var borderRadiusCalendarDayFocusRing = "var(--border-radius-calendar-day-focus-ring-i451al, 2px)";
export var borderRadiusContainer = "var(--border-radius-container-r46evd, 0px)";
export var borderRadiusControlCircularFocusRing = "var(--border-radius-control-circular-focus-ring-adbgw5, 50%)";
export var borderRadiusControlDefaultFocusRing = "var(--border-radius-control-default-focus-ring-eq2wmk, 2px)";
export var borderRadiusDropdown = "var(--border-radius-dropdown-awjhsp, 0px)";
export var borderRadiusFlashbar = "var(--border-radius-flashbar-e6itil, 0px)";
export var borderRadiusItem = "var(--border-radius-item-ilq21a, 0px)";
export var borderRadiusInput = "var(--border-radius-input-gycuv6, 2px)";
export var borderRadiusPopover = "var(--border-radius-popover-ewhy15, 2px)";
export var borderRadiusTabsFocusRing = "var(--border-radius-tabs-focus-ring-yky44v, 0px)";
export var borderRadiusTiles = "var(--border-radius-tiles-c0zt4m, 2px)";
export var borderRadiusToken = "var(--border-radius-token-8fyrg1, 2px)";
export var borderRadiusChatBubble = "var(--border-radius-chat-bubble-u484sq, 2px)";
export var borderRadiusTutorialPanelItem = "var(--border-radius-tutorial-panel-item-k7sfhh, 2px)";
export var borderWidthAlert = "var(--border-width-alert-pz0bqs, 1px)";
export var borderWidthField = "var(--border-width-field-404eki, 1px)";
export var borderWidthPopover = "var(--border-width-popover-clocg5, 1px)";
export var motionDurationAvatarGenAiGradient = "var(--motion-duration-avatar-gen-ai-gradient-3fx7ww, 3600ms)";
export var motionDurationAvatarLoadingDots = "var(--motion-duration-avatar-loading-dots-ii4xzu, 1200ms)";
export var motionEasingAvatarGenAiGradient = "var(--motion-easing-avatar-gen-ai-gradient-o8yfn2, cubic-bezier(0.7, 0, 0.3, 1))";
export var motionEasingResponsive = "var(--motion-easing-responsive-kehcbg, ease-out)";
export var motionEasingSticky = "var(--motion-easing-sticky-h2dhfe, ease-out)";
export var motionEasingExpressive = "var(--motion-easing-expressive-ahoycd, ease-out)";
export var motionDurationResponsive = "var(--motion-duration-responsive-o555ts, 135ms)";
export var motionDurationExpressive = "var(--motion-duration-expressive-jidfsq, 180ms)";
export var motionDurationComplex = "var(--motion-duration-complex-gcxcdu, 270ms)";
export var motionKeyframesFadeIn = "var(--motion-keyframes-fade-in-2qjwil, awsui-fade-in-35003c)";
export var motionKeyframesFadeOut = "var(--motion-keyframes-fade-out-ddkkh3, awsui-fade-out-35003c)";
export var motionKeyframesStatusIconError = "var(--motion-keyframes-status-icon-error-6klkxp, awsui-none-35003c)";
export var motionKeyframesScalePopup = "var(--motion-keyframes-scale-popup-qv5vbi, awsui-none-35003c)";
export var spaceContainerHorizontal = "var(--space-container-horizontal-jxdgil, 20px)";
export var spaceFieldHorizontal = "var(--space-field-horizontal-zt16s8, 8px)";
export var spaceScaledXxxs = "var(--space-scaled-xxxs-jdgipl, 2px)";
export var spaceScaledXxs = "var(--space-scaled-xxs-t2t62i, 4px)";
export var spaceScaledXs = "var(--space-scaled-xs-wbfgrv, 8px)";
export var spaceScaledS = "var(--space-scaled-s-913kwi, 12px)";
export var spaceScaledM = "var(--space-scaled-m-gxhdpl, 16px)";
export var spaceScaledL = "var(--space-scaled-l-2rs0gk, 20px)";
export var spaceScaledXl = "var(--space-scaled-xl-jawob6, 24px)";
export var spaceScaledXxl = "var(--space-scaled-xxl-fzmynl, 32px)";
export var spaceScaledXxxl = "var(--space-scaled-xxxl-crpygs, 40px)";
export var spaceStaticXxxs = "var(--space-static-xxxs-b6tgp7, 2px)";
export var spaceStaticXxs = "var(--space-static-xxs-jqh73n, 4px)";
export var spaceStaticXs = "var(--space-static-xs-zgeph4, 8px)";
export var spaceStaticS = "var(--space-static-s-q0yjjb, 12px)";
export var spaceStaticM = "var(--space-static-m-1plzkw, 16px)";
export var spaceStaticL = "var(--space-static-l-hozrvc, 20px)";
export var spaceStaticXl = "var(--space-static-xl-nmb7ge, 24px)";
export var spaceStaticXxl = "var(--space-static-xxl-qhb0le, 32px)";
export var spaceStaticXxxl = "var(--space-static-xxxl-srpnkv, 40px)";
export var shadowContainerActive = "var(--shadow-container-active-vb9kmq, 0px 4px 8px rgba(0, 28, 36, 0.45))";
