// @synchro-charts/core/dist/cjs/chartDefaults-b13cbd33.js
'use strict';

const constants = require('./constants-d154fa06.js');

const DEFAULT_CHART_CONFIG = {
    widgetId: 'fake-id',
    viewport: {
        start: new Date(1995, 0, 0, 0),
        end: new Date(2020, 1, 0, 0),
        yMin: 0,
        yMax: 10000,
    },
    // width is width - marginLeft - marginRight
    size: { width: 400 + 50 + 25, height: 350, marginLeft: 50, marginRight: 50, marginTop: 24, marginBottom: 30 },
    movement: {
        enableXScroll: true,
        enableYScroll: false,
        zoomMax: Infinity,
        zoomMin: 0.00001,
    },
    layout: {
        xGridVisible: false,
        yGridVisible: true,
        xTicksVisible: true,
        yTicksVisible: true,
    },
    scale: {
        xScaleType: constants.ScaleType.TimeSeries,
        yScaleType: constants.ScaleType.Linear,
        xScaleSide: 'bottom',
        yScaleSide: 'left',
    },
    dataStreams: [],
    legend: {
        position: constants.LEGEND_POSITION.BOTTOM,
        width: 170,
    },
};
const DEFAULT_THRESHOLD_OPTIONS = {
    showColor: true,
};
const DEFAULT_THRESHOLD_OPTIONS_OFF = {
    showColor: false,
};

exports.DEFAULT_CHART_CONFIG = DEFAULT_CHART_CONFIG;
exports.DEFAULT_THRESHOLD_OPTIONS = DEFAULT_THRESHOLD_OPTIONS;
exports.DEFAULT_THRESHOLD_OPTIONS_OFF = DEFAULT_THRESHOLD_OPTIONS_OFF;
