// @synchro-charts/core/dist/cjs/constants-d154fa06.js
'use strict';

// NOTE: `enum`s are held separately from the types which are exported
(function (ScaleType) {
    ScaleType["TimeSeries"] = "time-series";
    ScaleType["Log"] = "log";
    ScaleType["Linear"] = "linear";
})(exports.ScaleType || (exports.ScaleType = {}));
(function (LEGEND_POSITION) {
    LEGEND_POSITION["RIGHT"] = "RIGHT";
    LEGEND_POSITION["BOTTOM"] = "BOTTOM";
})(exports.LEGEND_POSITION || (exports.LEGEND_POSITION = {}));
(function (COMPARISON_OPERATOR) {
    COMPARISON_OPERATOR["LESS_THAN"] = "LT";
    COMPARISON_OPERATOR["GREATER_THAN"] = "GT";
    COMPARISON_OPERATOR["LESS_THAN_EQUAL"] = "LTE";
    COMPARISON_OPERATOR["GREATER_THAN_EQUAL"] = "GTE";
    COMPARISON_OPERATOR["EQUAL"] = "EQ";
    COMPARISON_OPERATOR["CONTAINS"] = "CONTAINS";
})(exports.COMPARISON_OPERATOR || (exports.COMPARISON_OPERATOR = {}));
const COMPARATOR_MAP = {
    GTE: '>=',
    GT: '>',
    LTE: '<=',
    LT: '<',
    EQ: '=',
};
(function (StatusIcon) {
    StatusIcon["ERROR"] = "error";
    StatusIcon["ACTIVE"] = "active";
    StatusIcon["NORMAL"] = "normal";
    StatusIcon["ACKNOWLEDGED"] = "acknowledged";
    StatusIcon["SNOOZED"] = "snoozed";
    StatusIcon["DISABLED"] = "disabled";
    StatusIcon["LATCHED"] = "latched";
})(exports.StatusIcon || (exports.StatusIcon = {}));
const STATUS_ICONS = [
    exports.StatusIcon.ERROR,
    exports.StatusIcon.ACTIVE,
    exports.StatusIcon.NORMAL,
    exports.StatusIcon.ACKNOWLEDGED,
    exports.StatusIcon.SNOOZED,
    exports.StatusIcon.DISABLED,
    exports.StatusIcon.LATCHED,
];
(function (DATA_ALIGNMENT) {
    DATA_ALIGNMENT["EITHER"] = "EITHER";
    DATA_ALIGNMENT["RIGHT"] = "RIGHT";
    DATA_ALIGNMENT["LEFT"] = "LEFT";
})(exports.DATA_ALIGNMENT || (exports.DATA_ALIGNMENT = {}));

exports.COMPARATOR_MAP = COMPARATOR_MAP;
exports.STATUS_ICONS = STATUS_ICONS;
