// @synchro-charts/core/dist/cjs/dataTypes-bde2225c.js
'use strict';

const DEFAULT_MESSAGE_OVERRIDES = {
    liveTimeFrameValueLabel: 'Value',
    historicalTimeFrameValueLabel: 'Value',
    noDataStreamsPresentHeader: 'No properties or alarms',
    noDataStreamsPresentSubHeader: "This widget doesn't have any properties or alarms.",
    noDataPresentHeader: 'No data',
    noDataPresentSubHeader: "There's no data to display for this time range.",
    liveModeOnly: 'This visualization displays only live data. Choose a live time frame to display data in this visualization.',
    unsupportedDataTypeHeader: 'Unable to render your data',
    unsupportedDataTypeSubHeader: 'This chart only supports the following DataType(s):',
    supportedTypes: 'Number, String, Boolean',
};
/** SVG Constants */
const STREAM_ICON_STROKE_LINECAP = 'round';
const STREAM_ICON_STROKE_WIDTH = 3;
const STREAM_ICON_PATH_COMMAND = 'M 2 2 H 15';
const TREND_ICON_DASH_ARRAY = '1, 5';

exports.DEFAULT_MESSAGE_OVERRIDES = DEFAULT_MESSAGE_OVERRIDES;
exports.STREAM_ICON_PATH_COMMAND = STREAM_ICON_PATH_COMMAND;
exports.STREAM_ICON_STROKE_LINECAP = STREAM_ICON_STROKE_LINECAP;
exports.STREAM_ICON_STROKE_WIDTH = STREAM_ICON_STROKE_WIDTH;
exports.TREND_ICON_DASH_ARRAY = TREND_ICON_DASH_ARRAY;
