// @synchro-charts/core/dist/cjs/line-chart-unsupported-data-types.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
const constants$1 = require('./constants-aac4f7fd.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2000, 0, 1);
const WIDTH = X_MAX.getTime() - X_MIN.getTime();
const LineChartUnsupportedDataTypes = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [
                        {
                            x: new Date(X_MIN.getTime() + WIDTH / 4).getTime(),
                            y: constants$1.Y_VALUE_STRING,
                        },
                        {
                            x: new Date(X_MIN.getTime() + WIDTH / 4).getTime(),
                            y: constants$1.Y_VALUE_STRING,
                        },
                    ],
                    resolution: 0,
                    dataType: dataConstants.DataType.STRING,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX }, bufferFactor: 1, minBufferSize: 1 }), index.h("sc-webgl-context", null))));
    }
};

exports.line_chart_unsupported_data_types = LineChartUnsupportedDataTypes;
