// @synchro-charts/core/dist/cjs/sc-annotations-draggable-multi.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');

const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2001, 0, 1);
const ANNOTATIONS = {
    y: [
        {
            id: 'annotation-1',
            color: 'red',
            isEditable: true,
            showValue: true,
            value: 1000,
            label: {
                text: 'annotation-1',
                show: true,
            },
        },
        {
            id: 'annotation-2',
            color: 'blue',
            isEditable: true,
            showValue: true,
            value: 0,
            label: {
                text: 'annotation-2',
                show: true,
            },
        },
    ],
};
const SIZE = {
    height: 340,
    width: 340,
};
const ScAnnotationsDraggableMulti = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("div", { style: { width: '500px', height: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [], annotations: ANNOTATIONS, size: SIZE, viewport: { start: X_MIN, end: X_MAX } })), index.h("div", { style: { width: '500px', height: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id-2", dataStreams: [], annotations: ANNOTATIONS, size: SIZE, viewport: { start: X_MIN, end: X_MAX } })), index.h("sc-webgl-context", null)));
    }
};

exports.sc_annotations_draggable_multi = ScAnnotationsDraggableMulti;
