// @synchro-charts/core/dist/cjs/sc-line-chart-stream-data.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2000, 0, 1);
// test data point dead center of the viewport
const TEST_DATA_POINT = {
    x: (X_MIN.getTime() + X_MAX.getTime()) / 2,
    y: (Y_MIN + Y_MAX) / 2,
};
const urlParams = new URLSearchParams(window.location.search);
const dataPerRoundParam = urlParams.get('dataPerRound');
const roundFrequencyParam = urlParams.get('roundFrequency');
const viewportSpeedParam = urlParams.get('viewportSpeed');
const DATA_SIZE_BATCH = dataPerRoundParam ? Number.parseInt(dataPerRoundParam, 10) : 1;
const DATA_FREQUENCY_MS = roundFrequencyParam ? Number.parseInt(roundFrequencyParam, 10) : time.SECOND_IN_MS;
const VIEWPORT_SPEED = viewportSpeedParam ? Number.parseInt(viewportSpeedParam, 10) : 0;
const createData = (point) => new Array(DATA_SIZE_BATCH).fill(0).map((_, i) => ({
    x: point.x,
    y: point.y + i,
}));
const ScLineChartStreamData = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.dataPoints = [TEST_DATA_POINT];
        this.viewport = { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX };
        this.viewportLoop = () => window.requestAnimationFrame(() => {
            this.viewport = Object.assign(Object.assign({}, this.viewport), { start: new Date(this.viewport.start.getTime() + VIEWPORT_SPEED), end: new Date(this.viewport.end.getTime() + VIEWPORT_SPEED) });
            this.viewportShifter = this.viewportLoop();
        });
    }
    componentWillLoad() {
        if (VIEWPORT_SPEED > 0) {
            this.viewportShifter = this.viewportLoop();
        }
        this.dataLoop = window.setInterval(() => {
            this.dataPoints = [
                ...this.dataPoints,
                ...createData({
                    x: (this.dataPoints[this.dataPoints.length - 1].x + X_MAX.getTime()) / 2,
                    y: TEST_DATA_POINT.y,
                }),
            ];
        }, DATA_FREQUENCY_MS);
    }
    disconnectedCallback() {
        clearInterval(this.dataLoop);
        if (this.viewportShifter != null) {
            window.cancelAnimationFrame(this.viewportShifter);
        }
    }
    render() {
        return (index.h("div", { id: "chart-container", style: { border: '1px solid lightgray', height: '500px', width: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: this.dataPoints,
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: this.viewport }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_line_chart_stream_data = ScLineChartStreamData;
