// @synchro-charts/core/dist/cjs/sc-scatter-chart-unsupported-data-types.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');
const constants$1 = require('./constants-aac4f7fd.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2001, 0, 1);
const TEST_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: constants$1.Y_VALUE_STRING,
};
const TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: constants$1.Y_VALUE_STRING,
};
const ScScatterChartUnsupportedDataTypes = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-scatter-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    name: 'test stream',
                    color: 'black',
                    aggregates: { [time.MONTH_IN_MS]: [TEST_DATA_POINT, TEST_DATA_POINT_2] },
                    data: [],
                    resolution: time.MONTH_IN_MS,
                    dataType: dataConstants.DataType.STRING,
                },
            ], annotations: {}, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_scatter_chart_unsupported_data_types = ScScatterChartUnsupportedDataTypes;
