// @synchro-charts/core/dist/cjs/sc-scatter-chart-trend-line-color-configuration.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const constants = require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const X_MIN = new Date(2018, 0, 0);
const X_MAX = new Date(2020, 0, 0);
const Y_MIN = 0;
const Y_MAX = 5000;
// test data point dead center of the viewport
const TEST_DATA_POINTS = [
    {
        x: new Date(2018, 6, 0).getTime(),
        y: 1000,
    },
    {
        x: new Date(2019, 0, 0).getTime(),
        y: 4000,
    },
    {
        x: new Date(2019, 6, 0).getTime(),
        y: 3000,
    },
];
const ScScatterChartTrendLineColorConfiguration = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'red',
                    name: 'test stream',
                    aggregates: { [time.SECOND_IN_MS]: TEST_DATA_POINTS },
                    data: [],
                    resolution: time.SECOND_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: {
                start: X_MIN,
                end: X_MAX,
                yMin: Y_MIN,
                yMax: Y_MAX,
            }, legend: {
                position: constants.LEGEND_POSITION.BOTTOM,
                width: 300,
            }, trends: [
                {
                    dataStreamId: 'test',
                    type: dataConstants.TREND_TYPE.LINEAR,
                    color: '#123abc',
                },
            ] }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_scatter_chart_trend_line_color_configuration = ScScatterChartTrendLineColorConfiguration;
