// @synchro-charts/core/dist/cjs/sc-size-provider-standard.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');

const DEFAULT_SIZE = {
    width: 500,
    height: 600,
};
const SHIFT_X_DIFF = 50;

const ScSizeProviderStandard = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.marginLeft = 0;
        this.onShiftRight = () => {
            this.marginLeft += SHIFT_X_DIFF;
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests", style: { width: '2000px', height: '2000px' } }, index.h("div", null, index.h("button", { id: "shift-right", onClick: this.onShiftRight }, "Shift Right")), index.h("div", { id: "container", style: {
                marginLeft: `${this.marginLeft}px`,
                height: `${DEFAULT_SIZE.height}px`,
                width: `${DEFAULT_SIZE.width}px`,
            } }, index.h("sc-size-provider", { renderFunc: size => index.h("sc-box", { someObject: size, size: size }) }))));
    }
};

exports.sc_size_provider_standard = ScSizeProviderStandard;
