// @synchro-charts/core/dist/cjs/sc-webgl-chart-threshold-coloration-multiple-thresholds.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const constants = require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2001, 0, 1);
const TEST_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: 2000,
};
const TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: 4000,
};
const TEST_2_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: 4000,
};
const TEST_2_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: 2000,
};
const ScWebglChartThresholdColorationMultipleThresholds = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_DATA_POINT, TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
                {
                    id: 'test2',
                    color: 'red',
                    name: 'test stream2',
                    data: [TEST_2_DATA_POINT, TEST_2_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], annotations: {
                y: [
                    {
                        value: 3500,
                        label: {
                            text: 'y label',
                            show: true,
                        },
                        showValue: true,
                        color: 'blue',
                        comparisonOperator: constants.COMPARISON_OPERATOR.GREATER_THAN_EQUAL,
                    },
                    {
                        value: 2500,
                        label: {
                            text: 'y label',
                            show: true,
                        },
                        showValue: true,
                        color: 'purple',
                        comparisonOperator: constants.COMPARISON_OPERATOR.GREATER_THAN_EQUAL,
                        isEditable: true,
                    },
                ],
            }, size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_chart_threshold_coloration_multiple_thresholds = ScWebglChartThresholdColorationMultipleThresholds;
