// @synchro-charts/core/dist/cjs/sc-webgl-bar-chart-unsupported-data-types.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');
const constants$1 = require('./constants-aac4f7fd.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(2000, 0);
const X_MAX = new Date(2001, 3);
const LEFT_X = new Date(2000, 3).getTime();
const MIDDLE_X = new Date(2000, 6).getTime();
const RIGHT_X = new Date(2000, 9).getTime();
const ScWebglBarChartUnsupportedDataTypes = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, index.h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test-name',
                    aggregates: {},
                    data: [
                        {
                            x: LEFT_X,
                            y: constants$1.Y_VALUE_STRING,
                        },
                        {
                            x: MIDDLE_X,
                            y: constants$1.Y_VALUE_STRING,
                        },
                        {
                            x: RIGHT_X,
                            y: constants$1.Y_VALUE_STRING,
                        },
                    ],
                    resolution: time.MONTH_IN_MS,
                    dataType: dataConstants.DataType.STRING,
                },
            ], size: {
                width: 500,
                height: 500,
            }, widgetId: "widget-id", viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            } })), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_bar_chart_unsupported_data_types = ScWebglBarChartUnsupportedDataTypes;
