// @synchro-charts/core/dist/cjs/sc-webgl-line-chart-dynamic-data-streams.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');
const constants$1 = require('./constants-aac4f7fd.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2000, 0, 1);
// test data point dead center of the viewport
const VIEWPORT_WIDTH = X_MAX.getTime() - X_MIN.getTime();
const LEFT_X = new Date(X_MIN.getTime() + VIEWPORT_WIDTH * (1 / 6)).getTime();
const MIDDLE_X = new Date(X_MIN.getTime() + VIEWPORT_WIDTH * (1 / 3)).getTime();
const RIGHT_X = new Date(X_MIN.getTime() + VIEWPORT_WIDTH * (1 / 2)).getTime();
const ScWebglLineChartDynamicDataStreams = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.dataStreams = [];
        this.addStream = () => {
            const leftPoint = {
                x: LEFT_X,
                y: constants$1.Y_VALUE,
            };
            const middlePoint = {
                x: MIDDLE_X,
                y: constants$1.Y_VALUE,
            };
            const rightPoint = {
                x: RIGHT_X,
                y: constants$1.Y_VALUE,
            };
            const streamId = `stream-${this.dataStreams.length + 1}`;
            this.dataStreams = [
                {
                    id: streamId,
                    color: 'black',
                    name: `${streamId}-name`,
                    aggregates: { [time.MINUTE_IN_MS]: [leftPoint, middlePoint, rightPoint] },
                    data: [],
                    resolution: time.MINUTE_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
                ...this.dataStreams,
            ];
        };
        this.removeStream = () => {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const [_firstStream, ...restStreams] = this.dataStreams;
            this.dataStreams = restStreams;
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("button", { id: "add-stream", onClick: this.addStream }, "Add Stream"), index.h("button", { id: "remove-stream", onClick: this.removeStream }, "Remove Stream"), index.h("br", null), index.h("br", null), index.h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: this.dataStreams, size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } })), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_line_chart_dynamic_data_streams = ScWebglLineChartDynamicDataStreams;
