// @synchro-charts/core/dist/cjs/sc-webgl-bar-chart-dynamic-data-streams.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');
const constants$1 = require('./constants-aac4f7fd.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(2000, 0);
const X_MAX = new Date(2001, 3);
const LEFT_X = new Date(2000, 3).getTime();
const MIDDLE_X = new Date(2000, 6).getTime();
const RIGHT_X = new Date(2000, 9).getTime();
const ScWebglBarChartDynamicDataStreams = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.dataStreams = [];
        this.colorIndex = 0;
        this.colors = ['#264653', '#2a9d8f', '#e9c46a', '#f4a261', '#e76f51'];
        this.increaseColorIndex = () => {
            this.colorIndex += 1;
        };
        this.getColor = () => {
            // Modding the  will cycle through the colors array when the color index becomes greater than
            // the colors array length
            return this.colors[this.colorIndex % this.colors.length];
        };
        this.addStream = () => {
            const leftPoint = {
                x: LEFT_X,
                y: constants$1.Y_VALUE,
            };
            const middlePoint = {
                x: MIDDLE_X,
                y: constants$1.Y_VALUE,
            };
            const rightPoint = {
                x: RIGHT_X,
                y: constants$1.Y_VALUE,
            };
            const streamId = `stream-${this.dataStreams.length + 1}`;
            this.dataStreams = [
                ...this.dataStreams,
                {
                    id: streamId,
                    color: this.getColor(),
                    name: `${streamId}-name`,
                    aggregates: {
                        [time.MONTH_IN_MS]: [leftPoint, middlePoint, rightPoint],
                    },
                    data: [],
                    resolution: time.MONTH_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ];
            this.increaseColorIndex();
        };
        this.removeStream = () => {
            this.dataStreams.pop();
            this.dataStreams = [...this.dataStreams];
            this.colorIndex -= 1;
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("button", { id: "add-stream", onClick: this.addStream }, "Add Stream"), index.h("button", { id: "remove-stream", onClick: this.removeStream }, "Remove Stream"), index.h("br", null), index.h("br", null), index.h("div", { id: "chart-container", style: { marginTop: '20px', width: '500px', height: '500px' } }, index.h("sc-bar-chart", { dataStreams: this.dataStreams, size: {
                width: 500,
                height: 500,
            }, widgetId: "widget-id", viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            } })), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_bar_chart_dynamic_data_streams = ScWebglBarChartDynamicDataStreams;
