// @synchro-charts/core/dist/cjs/sc-webgl-context-root.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2001, 0, 1);
const TEST_DATA_POINT = {
    x: new Date(2001, 0, 0).getTime(),
    y: 0,
};
const TEST_DATA_POINT_2 = {
    x: new Date(2001, 0, 0).getTime(),
    y: 4000,
};
const TEST_2DATA_POINT = {
    x: new Date(1998, 0, 0).getTime(),
    y: 0,
};
const TEST_2DATA_POINT_2 = {
    x: new Date(1998, 0, 0).getTime(),
    y: 4000,
};
const ScWebglContextRoot = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("div", { style: { width: '10000px', height: '10000px' } }, index.h("div", { style: { height: '500px', width: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_2DATA_POINT, TEST_2DATA_POINT_2, TEST_DATA_POINT, TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } })), index.h("div", { style: { height: '500px', width: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_2DATA_POINT, TEST_2DATA_POINT_2, TEST_DATA_POINT, TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } })), index.h("div", { style: { height: '500px', width: '500px' } }, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_2DATA_POINT, TEST_2DATA_POINT_2, TEST_DATA_POINT, TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], size: {
                height: 500,
                width: 500,
            }, viewport: { start: X_MIN, end: X_MAX, yMin: Y_MIN, yMax: Y_MAX } })), index.h("sc-webgl-context", null))));
    }
};

exports.sc_webgl_context_root = ScWebglContextRoot;
