// @synchro-charts/core/dist/cjs/sc-webgl-chart-large-viewport.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');

const VIEWPORT = {
    yMin: 0,
    yMax: 5000,
    start: new Date(2000, 0, 0),
    end: new Date(2001, 0, 0),
};
// test data point dead center of the viewport
const TEST_DATA_POINT = {
    x: new Date((VIEWPORT.start.getTime() + VIEWPORT.end.getTime()) / 2).getTime(),
    y: (VIEWPORT.yMin + VIEWPORT.yMax) / 2,
};
const ScWebglChartLargeViewport = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-line-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_DATA_POINT],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], widgetId: "widget-id", size: {
                height: 500,
                width: 500,
            }, viewport: VIEWPORT }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_chart_large_viewport = ScWebglChartLargeViewport;
