// @synchro-charts/core/dist/cjs/sc-webgl-bar-chart-dynamic-buffer.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2000, 0, 1);
/**
 * Testing route for the webGL rendering without being fully coupled to the chart.
 *
 * Used to test the behavior of a bar chart when adding/removing data points
 */
const WIDTH = X_MAX.getTime() - X_MIN.getTime();
const ScWebglBarChartDynamicBuffer = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.data = [];
        this.addDataPoint = () => {
            const dataPoint = {
                // To generate new bar with half of the distance.
                x: new Date(X_MIN.getTime() + WIDTH / (2 + this.data.length)).getTime(),
                y: 2500,
            };
            this.data = [dataPoint, ...this.data];
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("button", { id: "add-data-point", onClick: this.addDataPoint }, "Add Data Point"), index.h("div", { id: "chart-container", style: { height: '500px', width: '500px', marginTop: '20px' } }, index.h("sc-bar-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'red',
                    name: 'test stream',
                    aggregates: {
                        [time.MONTH_IN_MS]: this.data,
                    },
                    data: [],
                    resolution: time.MONTH_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            }, bufferFactor: 1, minBufferSize: 1 }), index.h("sc-webgl-context", null))));
    }
};

exports.sc_webgl_bar_chart_dynamic_buffer = ScWebglBarChartDynamicBuffer;
