// @synchro-charts/core/dist/cjs/sc-webgl-bar-chart-standard.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');
require('./utils-43efeece.js');
require('./predicates-739d43b9.js');
require('./number-46ecc337.js');
require('./dataFilters-b9b845d5.js');
const constants$1 = require('./constants-aac4f7fd.js');

const ScWebglBarChartStandard = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [],
                    aggregates: {
                        [time.MINUTE_IN_MS]: [constants$1.TEST_DATA_POINT_STANDARD],
                    },
                    resolution: time.MINUTE_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: constants$1.Y_MIN, yMax: constants$1.Y_MAX, start: constants$1.X_MIN, end: constants$1.X_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_bar_chart_standard = ScWebglBarChartStandard;
