// @synchro-charts/core/dist/cjs/sc-webgl-bar-chart-start-from-zero.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

const X_MIN = new Date(2000, 0, 0, 0, 0);
const X_MAX = new Date(2000, 0, 0, 0, 10);
const ScWebglBarChartStartFromZero = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.testData = [
            {
                x: new Date((X_MIN.getTime() + X_MAX.getTime()) / 2).getTime(),
                y: 4500,
            },
        ];
        this.changeDataDirection = () => {
            this.testData = this.testData.map(({ x, y }) => {
                return {
                    x,
                    y: -y,
                };
            });
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("button", { id: "change-data-direction", onClick: this.changeDataDirection }, "Change Data Direction"), index.h("br", null), index.h("br", null), index.h("div", { id: "chart-container", style: { width: '500px', height: '500px' } }, index.h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    color: 'purple',
                    name: 'test stream',
                    data: [],
                    aggregates: {
                        [time.MINUTE_IN_MS]: this.testData,
                    },
                    resolution: time.MINUTE_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, viewport: { start: X_MIN, end: X_MAX } }), index.h("sc-webgl-context", null))));
    }
};

exports.sc_webgl_bar_chart_start_from_zero = ScWebglBarChartStartFromZero;
