// @synchro-charts/core/dist/cjs/sc-webgl-bar-chart-positive-negative.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const Y_MIN = -100;
const Y_MAX = 100;
const X_MIN = new Date(2000, 0, 0, 0, 0);
const X_MAX = new Date(2000, 0, 0, 0, 10);
// test data point dead center of the viewport
const TEST_DATA_POINT = {
    x: new Date(2000, 0, 0, 0, 3).getTime(),
    y: 50,
};
const TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0, 0, 7).getTime(),
    y: -50,
};
const ScWebglBarChartPositiveNegative = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", { id: "chart-container", style: { width: '500px', height: '500px' } }, index.h("sc-bar-chart", { dataStreams: [
                {
                    id: 'test',
                    aggregates: { [time.MINUTE_IN_MS]: [TEST_DATA_POINT, TEST_DATA_POINT_2] },
                    data: [],
                    resolution: time.MINUTE_IN_MS,
                    color: 'black',
                    name: 'test stream',
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], widgetId: "widget-id", size: {
                width: 500,
                height: 500,
            }, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_bar_chart_positive_negative = ScWebglBarChartPositiveNegative;
