// @synchro-charts/core/dist/cjs/sc-webgl-chart-annotation-editable.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const constants = require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');

const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2001, 0, 1);
const TEST_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: 2000,
};
const TEST_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: 4000,
};
const TEST_2_DATA_POINT = {
    x: new Date(1999, 0, 0).getTime(),
    y: 4000,
};
const TEST_2_DATA_POINT_2 = {
    x: new Date(2000, 0, 0).getTime(),
    y: 2000,
};
const Y_THRESHOLD = {
    isEditable: true,
    comparisonOperator: constants.COMPARISON_OPERATOR.GREATER_THAN,
    value: 3800,
    label: {
        text: 'here is a y label',
        show: true,
    },
    showValue: true,
    color: 'blue',
    id: 'blue-threshold',
};
const Y_ANNOTATION = {
    isEditable: true,
    value: 3200,
    label: {
        text: 'another y label',
        show: true,
    },
    showValue: true,
    color: 'green',
    id: 'green-anotation',
};
let oldAnnotations;
const ScWebglChartAnnotationRescaling = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.isEditableValue = false;
        this.isShowValue = true;
        this.annotations = {
            x: [
                {
                    value: new Date((X_MAX.getTime() + X_MIN.getTime()) / 2),
                    label: {
                        text: 'here is a x label',
                        show: true,
                    },
                    showValue: true,
                    color: 'purple',
                },
            ],
            y: [
                Object.assign(Object.assign({}, Y_ANNOTATION), { isEditable: !this.isEditableValue, showValue: this.isShowValue }),
                Object.assign(Object.assign({}, Y_THRESHOLD), { isEditable: !this.isEditableValue, showValue: !this.isShowValue }),
                Object.assign(Object.assign({}, Y_ANNOTATION), { isEditable: this.isEditableValue, value: 2300, color: 'red', showValue: this.isShowValue, id: 'red-annotation' }),
            ],
        };
        this.changeValue = () => {
            // this.annotations = oldAnnotations; //passed by reference so this doesn't work since draggable will modify oldAnnotations
            // this mimics the seperate annotation state of the app and the SC layer
            const { y } = this.annotations;
            this.annotations = Object.assign(Object.assign({}, this.annotations), { y: y.map(annotation => {
                    return Object.assign(Object.assign({}, annotation), { value: oldAnnotations.y.find(oldannotation => oldannotation.id === annotation.id).value });
                }) });
        };
        this.onEditableChange = () => {
            const { y } = this.annotations;
            this.annotations = Object.assign(Object.assign({}, this.annotations), { y: y.map(annotation => {
                    return Object.assign(Object.assign({}, annotation), { isEditable: !annotation.isEditable });
                }) });
        };
        this.onShowValueChange = () => {
            const { y } = this.annotations;
            this.annotations = Object.assign(Object.assign({}, this.annotations), { y: y.map(annotation => {
                    return Object.assign(Object.assign({}, annotation), { showValue: !annotation.showValue });
                }) });
        };
    }
    onWidgetUpdated({ detail: configUpdate }) {
        this.annotations = configUpdate.annotations;
        oldAnnotations = this.annotations;
    }
    componentDidLoad() {
        oldAnnotations = this.annotations;
        setInterval(this.changeValue, 1000);
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests", style: { width: '1000px', height: '1000px' } }, index.h("div", null, index.h("button", { id: "change-editable", onClick: this.onEditableChange }, "Change Editable")), index.h("div", null, index.h("button", { id: "change-showvalue", onClick: this.onShowValueChange }, "Change Show Value")), index.h("div", null, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [
                {
                    id: 'test',
                    color: 'black',
                    name: 'test stream',
                    data: [TEST_DATA_POINT, TEST_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
                {
                    id: 'test2',
                    color: 'orange',
                    name: 'test stream2',
                    data: [TEST_2_DATA_POINT, TEST_2_DATA_POINT_2],
                    resolution: 0,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], annotations: this.annotations, viewport: { start: X_MIN, end: X_MAX }, size: {
                height: 1000,
                width: 1000,
            } }), index.h("sc-webgl-context", null))));
    }
};

exports.sc_webgl_chart_annotation_editable = ScWebglChartAnnotationRescaling;
