// @synchro-charts/core/dist/cjs/sc-webgl-chart-no-annotations.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
require('./constants-d154fa06.js');
const constants$1 = require('./constants-aac4f7fd.js');

const ScWebglChartNoAnnotations = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-line-chart", { widgetId: "widget-id", dataStreams: [], annotations: {
                x: [
                    {
                        value: constants$1.X_MIN,
                        label: {
                            text: 'x label',
                            show: true,
                        },
                        showValue: true,
                        color: 'red',
                    },
                ],
                y: [
                    {
                        value: constants$1.Y_MIN,
                        label: {
                            text: 'y label',
                            show: true,
                        },
                        showValue: true,
                        color: 'blue',
                    },
                ],
                show: false,
            }, size: {
                height: 500,
                width: 500,
            }, viewport: { start: constants$1.X_MIN, end: constants$1.X_MAX, yMin: constants$1.Y_MIN, yMax: constants$1.Y_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.sc_webgl_chart_no_annotations = ScWebglChartNoAnnotations;
