// @synchro-charts/core/dist/cjs/status-timeline-margin.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 1, 0);
const X_MAX = new Date(1998, 6, 0);
const DATA_STREAM_1 = {
    id: 'test',
    color: 'red',
    name: 'test stream',
    resolution: time.MONTH_IN_MS,
    aggregates: {
        [time.MONTH_IN_MS]: [
            { x: new Date(1998, 3, 0, 0).getTime(), y: 1000 },
            { x: new Date(1998, 4, 0, 0).getTime(), y: 3 },
        ],
    },
    data: [],
    dataType: dataConstants.DataType.NUMBER,
};
const DATA_STREAM_2 = {
    id: 'test2',
    color: 'orange',
    name: 'test stream2',
    resolution: time.MONTH_IN_MS,
    data: [
        { x: new Date(1998, 3, 0, 0).getTime(), y: 2000 },
        { x: new Date(1998, 4, 0, 0).getTime(), y: 4 },
    ],
    dataType: dataConstants.DataType.NUMBER,
};
const StatusTimelineStatusMargin = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.data = [];
    }
    render() {
        return (index.h("div", { id: "chart-container", style: { height: '500px', width: '500px', marginTop: '20px' } }, index.h("sc-status-timeline", { alarms: { expires: time.MONTH_IN_MS }, widgetId: "widget-id", dataStreams: [DATA_STREAM_1, DATA_STREAM_2], viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.status_timeline_margin = StatusTimelineStatusMargin;
