// @synchro-charts/core/dist/cjs/status-timeline-dynamic-data.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(2000, 0);
const X_MAX = new Date(2001, 12);
const StatusTimelineDynamicData = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.data = [];
        this.monthIndex = 1;
        this.addDataPoint = () => {
            const dataPoint = {
                x: new Date(2000, this.monthIndex).getTime(),
                y: 3000,
            };
            this.data = [dataPoint, ...this.data];
            this.monthIndex += 2;
        };
        this.removeDataPoint = () => {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const [_firstPoint, ...rest] = this.data;
            this.data = rest;
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("button", { id: "add-data-point", onClick: this.addDataPoint }, "Add Data Point"), index.h("button", { id: "remove-data-point", onClick: this.removeDataPoint }, "Remove Data Point"), index.h("br", null), index.h("br", null), index.h("div", { id: "chart-container", style: { height: '500px', width: '500px' } }, index.h("sc-status-timeline", { alarms: { expires: time.MONTH_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    name: 'test stream',
                    color: 'red',
                    aggregates: { [time.MONTH_IN_MS]: this.data },
                    data: [],
                    resolution: time.MONTH_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], size: {
                width: 500,
                height: 500,
            }, widgetId: "widget-id", viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            } }), index.h("sc-webgl-context", null))));
    }
};

exports.status_timeline_dynamic_data = StatusTimelineDynamicData;
