// @synchro-charts/core/dist/cjs/status-timeline-dynamic-buffer.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// viewport boundaries
const Y_MIN = 0;
const Y_MAX = 5000;
const X_MIN = new Date(1998, 0, 0);
const X_MAX = new Date(2000, 0, 1);
const WIDTH = X_MAX.getTime() - X_MIN.getTime();
const StatusTimelineDynamicBuffer = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.data = [];
        this.addDataPoint = () => {
            const dataPoint = {
                // To generate new status with half of the distance.
                x: X_MIN.getTime() + WIDTH / (2 + this.data.length),
                y: 2.5,
            };
            this.data = [dataPoint, ...this.data];
        };
    }
    render() {
        return (index.h("div", { class: "synchro-chart-tests" }, index.h("button", { id: "add-data-point", onClick: this.addDataPoint }, "Add Data Point"), index.h("div", { id: "chart-container", style: { height: '500px', width: '500px', marginTop: '20px' } }, index.h("sc-status-timeline", { widgetId: "widget-id", alarms: { expires: time.MONTH_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    color: 'red',
                    name: 'test stream',
                    aggregates: { [time.MONTH_IN_MS]: this.data },
                    data: [],
                    resolution: time.MONTH_IN_MS,
                    dataType: dataConstants.DataType.NUMBER,
                },
            ], viewport: {
                yMin: Y_MIN,
                yMax: Y_MAX,
                start: X_MIN,
                end: X_MAX,
            }, bufferFactor: 1, minBufferSize: 1 }), index.h("sc-webgl-context", null))));
    }
};

exports.status_timeline_dynamic_buffer = StatusTimelineDynamicBuffer;
