// @synchro-charts/core/dist/cjs/status-timeline-raw-data.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0b3ed36d.js');
const constants = require('./constants-d154fa06.js');
const dataConstants = require('./dataConstants-02642889.js');
require('./index-0fadf3ad.js');
const time = require('./time-3c18516e.js');

// Ten Minute Duration Viewport
const Y_MAX = 100;
const Y_MIN = 0;
const X_MIN = new Date(2000, 0, 0);
const X_MAX = new Date(2000, 0, 0, 0, 10);
const STATUS_ZERO = 'STATUS_ZERO';
const STATUS_ONE = 'STATUS_ONE';
const STATUS_TWO = 'STATUS_TWO';
const STATUS_THREE = 'STATUS_THREE';
const DATA_POINTS = [
    { x: X_MIN.getTime(), y: STATUS_ZERO },
    { x: X_MIN.getTime() + 5 * time.MINUTE_IN_MS, y: STATUS_ONE },
    { x: X_MIN.getTime() + 6 * time.MINUTE_IN_MS, y: STATUS_TWO },
    { x: X_MIN.getTime() + 6.5 * time.MINUTE_IN_MS, y: STATUS_THREE },
];
const annotations = {
    y: [
        {
            value: STATUS_ZERO,
            color: 'red',
            comparisonOperator: constants.COMPARISON_OPERATOR.EQUAL,
        },
        {
            value: STATUS_ONE,
            color: 'blue',
            comparisonOperator: constants.COMPARISON_OPERATOR.EQUAL,
        },
        {
            value: STATUS_TWO,
            color: 'green',
            comparisonOperator: constants.COMPARISON_OPERATOR.EQUAL,
        },
    ],
};
const StatusTimelineRawData = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    render() {
        return (index.h("div", null, index.h("sc-status-timeline", { alarms: { expires: time.MINUTE_IN_MS }, dataStreams: [
                {
                    id: 'test',
                    data: DATA_POINTS,
                    color: 'black',
                    name: 'test stream 1',
                    resolution: 0,
                    dataType: dataConstants.DataType.STRING,
                },
            ], widgetId: "test-id", size: {
                width: 500,
                height: 500,
            }, annotations: annotations, viewport: { yMin: Y_MIN, yMax: Y_MAX, start: X_MIN, end: X_MAX } }), index.h("sc-webgl-context", null)));
    }
};

exports.status_timeline_raw_data = StatusTimelineRawData;
