// @aws-sdk/client-iotsitewise/dist-es/models/models_0.js
import { SENSITIVE_STRING } from "@smithy/smithy-client";
import { IoTSiteWiseServiceException as __BaseException } from "./IoTSiteWiseServiceException";
export class AccessDeniedException extends __BaseException {
    name = "AccessDeniedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
    }
}
export const Permission = {
    ADMINISTRATOR: "ADMINISTRATOR",
    VIEWER: "VIEWER",
};
export const Quality = {
    BAD: "BAD",
    GOOD: "GOOD",
    UNCERTAIN: "UNCERTAIN",
};
export const AggregateType = {
    AVERAGE: "AVERAGE",
    COUNT: "COUNT",
    MAXIMUM: "MAXIMUM",
    MINIMUM: "MINIMUM",
    STANDARD_DEVIATION: "STANDARD_DEVIATION",
    SUM: "SUM",
};
export const PropertyDataType = {
    BOOLEAN: "BOOLEAN",
    DOUBLE: "DOUBLE",
    INTEGER: "INTEGER",
    STRING: "STRING",
    STRUCT: "STRUCT",
};
export const PropertyNotificationState = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const AssetErrorCode = {
    INTERNAL_FAILURE: "INTERNAL_FAILURE",
};
export const ForwardingConfigState = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const ComputeLocation = {
    CLOUD: "CLOUD",
    EDGE: "EDGE",
};
export const AssetModelState = {
    ACTIVE: "ACTIVE",
    CREATING: "CREATING",
    DELETING: "DELETING",
    FAILED: "FAILED",
    PROPAGATING: "PROPAGATING",
    UPDATING: "UPDATING",
};
export const ErrorCode = {
    INTERNAL_FAILURE: "INTERNAL_FAILURE",
    VALIDATION_ERROR: "VALIDATION_ERROR",
};
export const DetailedErrorCode = {
    INCOMPATIBLE_COMPUTE_LOCATION: "INCOMPATIBLE_COMPUTE_LOCATION",
    INCOMPATIBLE_FORWARDING_CONFIGURATION: "INCOMPATIBLE_FORWARDING_CONFIGURATION",
};
export const AssetModelType = {
    ASSET_MODEL: "ASSET_MODEL",
    COMPONENT_MODEL: "COMPONENT_MODEL",
};
export const AssetModelVersionType = {
    ACTIVE: "ACTIVE",
    LATEST: "LATEST",
};
export const AssetRelationshipType = {
    HIERARCHY: "HIERARCHY",
};
export const AssetState = {
    ACTIVE: "ACTIVE",
    CREATING: "CREATING",
    DELETING: "DELETING",
    FAILED: "FAILED",
    UPDATING: "UPDATING",
};
export class ConflictingOperationException extends __BaseException {
    name = "ConflictingOperationException";
    $fault = "client";
    resourceId;
    resourceArn;
    constructor(opts) {
        super({
            name: "ConflictingOperationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConflictingOperationException.prototype);
        this.resourceId = opts.resourceId;
        this.resourceArn = opts.resourceArn;
    }
}
export class InternalFailureException extends __BaseException {
    name = "InternalFailureException";
    $fault = "server";
    constructor(opts) {
        super({
            name: "InternalFailureException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalFailureException.prototype);
    }
}
export class InvalidRequestException extends __BaseException {
    name = "InvalidRequestException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidRequestException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRequestException.prototype);
    }
}
export class LimitExceededException extends __BaseException {
    name = "LimitExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, LimitExceededException.prototype);
    }
}
export class ResourceAlreadyExistsException extends __BaseException {
    name = "ResourceAlreadyExistsException";
    $fault = "client";
    resourceId;
    resourceArn;
    constructor(opts) {
        super({
            name: "ResourceAlreadyExistsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceAlreadyExistsException.prototype);
        this.resourceId = opts.resourceId;
        this.resourceArn = opts.resourceArn;
    }
}
export class ResourceNotFoundException extends __BaseException {
    name = "ResourceNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
export class ThrottlingException extends __BaseException {
    name = "ThrottlingException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
    }
}
export const AuthMode = {
    IAM: "IAM",
    SSO: "SSO",
};
export const TimeOrdering = {
    ASCENDING: "ASCENDING",
    DESCENDING: "DESCENDING",
};
export const BatchGetAssetPropertyAggregatesErrorCode = {
    AccessDeniedException: "AccessDeniedException",
    InvalidRequestException: "InvalidRequestException",
    ResourceNotFoundException: "ResourceNotFoundException",
};
export const BatchEntryCompletionStatus = {
    ERROR: "ERROR",
    SUCCESS: "SUCCESS",
};
export class ServiceUnavailableException extends __BaseException {
    name = "ServiceUnavailableException";
    $fault = "server";
    constructor(opts) {
        super({
            name: "ServiceUnavailableException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceUnavailableException.prototype);
    }
}
export const BatchGetAssetPropertyValueErrorCode = {
    AccessDeniedException: "AccessDeniedException",
    InvalidRequestException: "InvalidRequestException",
    ResourceNotFoundException: "ResourceNotFoundException",
};
export const BatchGetAssetPropertyValueHistoryErrorCode = {
    AccessDeniedException: "AccessDeniedException",
    InvalidRequestException: "InvalidRequestException",
    ResourceNotFoundException: "ResourceNotFoundException",
};
export const BatchPutAssetPropertyValueErrorCode = {
    AccessDeniedException: "AccessDeniedException",
    ConflictingOperationException: "ConflictingOperationException",
    InternalFailureException: "InternalFailureException",
    InvalidRequestException: "InvalidRequestException",
    LimitExceededException: "LimitExceededException",
    ResourceNotFoundException: "ResourceNotFoundException",
    ServiceUnavailableException: "ServiceUnavailableException",
    ThrottlingException: "ThrottlingException",
    TimestampOutOfRangeException: "TimestampOutOfRangeException",
};
export class PreconditionFailedException extends __BaseException {
    name = "PreconditionFailedException";
    $fault = "client";
    resourceId;
    resourceArn;
    constructor(opts) {
        super({
            name: "PreconditionFailedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PreconditionFailedException.prototype);
        this.resourceId = opts.resourceId;
        this.resourceArn = opts.resourceArn;
    }
}
export const ColumnName = {
    ALIAS: "ALIAS",
    ASSET_ID: "ASSET_ID",
    DATA_TYPE: "DATA_TYPE",
    PROPERTY_ID: "PROPERTY_ID",
    QUALITY: "QUALITY",
    TIMESTAMP_NANO_OFFSET: "TIMESTAMP_NANO_OFFSET",
    TIMESTAMP_SECONDS: "TIMESTAMP_SECONDS",
    VALUE: "VALUE",
};
export const JobStatus = {
    CANCELLED: "CANCELLED",
    COMPLETED: "COMPLETED",
    COMPLETED_WITH_FAILURES: "COMPLETED_WITH_FAILURES",
    FAILED: "FAILED",
    PENDING: "PENDING",
    RUNNING: "RUNNING",
};
export const DatasetSourceFormat = {
    KNOWLEDGE_BASE: "KNOWLEDGE_BASE",
};
export const DatasetSourceType = {
    KENDRA: "KENDRA",
};
export const DatasetState = {
    ACTIVE: "ACTIVE",
    CREATING: "CREATING",
    DELETING: "DELETING",
    FAILED: "FAILED",
    UPDATING: "UPDATING",
};
export const ImageFileType = {
    PNG: "PNG",
};
export const PortalType = {
    SITEWISE_PORTAL_V1: "SITEWISE_PORTAL_V1",
    SITEWISE_PORTAL_V2: "SITEWISE_PORTAL_V2",
};
export const MonitorErrorCode = {
    INTERNAL_FAILURE: "INTERNAL_FAILURE",
    LIMIT_EXCEEDED: "LIMIT_EXCEEDED",
    VALIDATION_ERROR: "VALIDATION_ERROR",
};
export const PortalState = {
    ACTIVE: "ACTIVE",
    CREATING: "CREATING",
    DELETING: "DELETING",
    FAILED: "FAILED",
    PENDING: "PENDING",
    UPDATING: "UPDATING",
};
export const ConfigurationState = {
    ACTIVE: "ACTIVE",
    UPDATE_FAILED: "UPDATE_FAILED",
    UPDATE_IN_PROGRESS: "UPDATE_IN_PROGRESS",
};
export const EncryptionType = {
    KMS_BASED_ENCRYPTION: "KMS_BASED_ENCRYPTION",
    SITEWISE_DEFAULT_ENCRYPTION: "SITEWISE_DEFAULT_ENCRYPTION",
};
export const CapabilitySyncStatus = {
    IN_SYNC: "IN_SYNC",
    NOT_APPLICABLE: "NOT_APPLICABLE",
    OUT_OF_SYNC: "OUT_OF_SYNC",
    SYNC_FAILED: "SYNC_FAILED",
    UNKNOWN: "UNKNOWN",
};
export const LoggingLevel = {
    ERROR: "ERROR",
    INFO: "INFO",
    OFF: "OFF",
};
export const DisassociatedDataStorageState = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const StorageType = {
    MULTI_LAYER_STORAGE: "MULTI_LAYER_STORAGE",
    SITEWISE_DEFAULT_STORAGE: "SITEWISE_DEFAULT_STORAGE",
};
export const WarmTierState = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
export const ScalarType = {
    BOOLEAN: "BOOLEAN",
    DOUBLE: "DOUBLE",
    INT: "INT",
    STRING: "STRING",
    TIMESTAMP: "TIMESTAMP",
};
export class QueryTimeoutException extends __BaseException {
    name = "QueryTimeoutException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "QueryTimeoutException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, QueryTimeoutException.prototype);
    }
}
export class ValidationException extends __BaseException {
    name = "ValidationException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ValidationException.prototype);
    }
}
export var ResponseStream;
(function (ResponseStream) {
    ResponseStream.visit = (value, visitor) => {
        if (value.trace !== undefined)
            return visitor.trace(value.trace);
        if (value.output !== undefined)
            return visitor.output(value.output);
        if (value.accessDeniedException !== undefined)
            return visitor.accessDeniedException(value.accessDeniedException);
        if (value.conflictingOperationException !== undefined)
            return visitor.conflictingOperationException(value.conflictingOperationException);
        if (value.internalFailureException !== undefined)
            return visitor.internalFailureException(value.internalFailureException);
        if (value.invalidRequestException !== undefined)
            return visitor.invalidRequestException(value.invalidRequestException);
        if (value.limitExceededException !== undefined)
            return visitor.limitExceededException(value.limitExceededException);
        if (value.resourceNotFoundException !== undefined)
            return visitor.resourceNotFoundException(value.resourceNotFoundException);
        if (value.throttlingException !== undefined)
            return visitor.throttlingException(value.throttlingException);
        return visitor._(value.$unknown[0], value.$unknown[1]);
    };
})(ResponseStream || (ResponseStream = {}));
export const IdentityType = {
    GROUP: "GROUP",
    IAM: "IAM",
    USER: "USER",
};
export const ResourceType = {
    PORTAL: "PORTAL",
    PROJECT: "PROJECT",
};
export const TargetResourceType = {
    ASSET: "ASSET",
};
export const ListAssetModelPropertiesFilter = {
    ALL: "ALL",
    BASE: "BASE",
};
export const ListAssetPropertiesFilter = {
    ALL: "ALL",
    BASE: "BASE",
};
export const TraversalType = {
    PATH_TO_ROOT: "PATH_TO_ROOT",
};
export const ListAssetsFilter = {
    ALL: "ALL",
    TOP_LEVEL: "TOP_LEVEL",
};
export const TraversalDirection = {
    CHILD: "CHILD",
    PARENT: "PARENT",
};
export const CreatePortalRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.portalContactEmail && { portalContactEmail: SENSITIVE_STRING }),
    ...(obj.notificationSenderEmail && { notificationSenderEmail: SENSITIVE_STRING }),
});
export const DescribePortalResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.portalContactEmail && { portalContactEmail: SENSITIVE_STRING }),
    ...(obj.notificationSenderEmail && { notificationSenderEmail: SENSITIVE_STRING }),
});
export const InvokeAssistantRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.message && { message: SENSITIVE_STRING }),
});
export const ResponseStreamFilterSensitiveLog = (obj) => {
    if (obj.trace !== undefined)
        return { trace: obj.trace };
    if (obj.output !== undefined)
        return { output: obj.output };
    if (obj.accessDeniedException !== undefined)
        return { accessDeniedException: obj.accessDeniedException };
    if (obj.conflictingOperationException !== undefined)
        return { conflictingOperationException: obj.conflictingOperationException };
    if (obj.internalFailureException !== undefined)
        return { internalFailureException: obj.internalFailureException };
    if (obj.invalidRequestException !== undefined)
        return { invalidRequestException: obj.invalidRequestException };
    if (obj.limitExceededException !== undefined)
        return { limitExceededException: obj.limitExceededException };
    if (obj.resourceNotFoundException !== undefined)
        return { resourceNotFoundException: obj.resourceNotFoundException };
    if (obj.throttlingException !== undefined)
        return { throttlingException: obj.throttlingException };
    if (obj.$unknown !== undefined)
        return { [obj.$unknown[0]]: "UNKNOWN" };
};
export const InvokeAssistantResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.body && { body: "STREAMING_CONTENT" }),
});
