// @iot-app-kit/components/dist/collection/components/iot-resource-explorer/iot-tree-table.js
import { Component, Host, h, Element, Prop, State } from '@stencil/core';
import { createRoot } from 'react-dom/client';
import { RelatedTable, withUseTreeCollection, } from '@iot-app-kit/related-table';
import { createElement } from 'react';
const RelatedTableWithCollectionHooks = withUseTreeCollection(RelatedTable);
export class IotTreeTable {
  constructor() {
    this.selectedItems = [];
  }
  componentDidLoad() {
    this.root = createRoot(this.host);
    this.componentDidUpdate();
  }
  componentDidUpdate() {
    const attributes = {
      items: this.items,
      columnDefinitions: this.columnDefinitions,
      collectionOptions: this.collectionOptions,
      expanded: this.expanded,
      loading: this.loading,
      loadingText: this.loadingText,
      filterPlaceholder: this.filterPlaceholder,
      selectionType: this.selectionType,
      selectedItems: this.selectedItems,
      empty: this.empty,
      isItemDisabled: this.isItemDisabled,
      wrapLines: this.wrapLines,
      resizableColumns: this.resizableColumns,
      sortingDisabled: this.sortingDisabled,
      ariaLabels: this.ariaLabels,
      expandChildren: this.onExpandChildren,
      onSortingChange: this.onSortingChange,
      onSelectionChange: (event) => {
        this.selectedItems = event.detail.selectedItems;
        if (this.onSelectionChange) {
          this.onSelectionChange(event);
        }
      },
    };
    this.root.render(createElement(RelatedTableWithCollectionHooks, attributes));
  }
  render() {
    return h(Host, null);
  }
  static get is() { return "iot-tree-table"; }
  static get properties() { return {
    "items": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "unknown[]",
        "resolved": "unknown[]",
        "references": {}
      },
      "required": true,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "columnDefinitions": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "TableProps.ColumnDefinition<any>[]",
        "resolved": "ColumnDefinition<any>[]",
        "references": {
          "TableProps": {
            "location": "import",
            "path": "@awsui/components-react/table"
          }
        }
      },
      "required": true,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "collectionOptions": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "UseTreeCollection<unknown>",
        "resolved": "UseTreeCollection<unknown>",
        "references": {
          "UseTreeCollection": {
            "location": "import",
            "path": "@iot-app-kit/related-table"
          }
        }
      },
      "required": true,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "expanded": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "expanded",
      "reflect": false
    },
    "loading": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "loading",
      "reflect": false
    },
    "loadingText": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "loading-text",
      "reflect": false
    },
    "filterPlaceholder": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "filter-placeholder",
      "reflect": false
    },
    "selectionType": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "TableProps.SelectionType",
        "resolved": "\"multi\" | \"single\"",
        "references": {
          "TableProps": {
            "location": "import",
            "path": "@awsui/components-react/table"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "selection-type",
      "reflect": false
    },
    "empty": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "EmptyStateProps",
        "resolved": "EmptyStateProps",
        "references": {
          "EmptyStateProps": {
            "location": "import",
            "path": "@iot-app-kit/related-table"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "isItemDisabled": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "(item: unknown) => boolean",
        "resolved": "(item: unknown) => boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "wrapLines": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "wrap-lines",
      "reflect": false
    },
    "resizableColumns": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "resizable-columns",
      "reflect": false
    },
    "sortingDisabled": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "sorting-disabled",
      "reflect": false
    },
    "ariaLabels": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "TableProps.AriaLabels<unknown>",
        "resolved": "AriaLabels<unknown>",
        "references": {
          "TableProps": {
            "location": "import",
            "path": "@awsui/components-react/table"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "onSelectionChange": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "(\n    event: NonCancelableCustomEvent<TableProps.SelectionChangeDetail<unknown>>\n  ) => void",
        "resolved": "(event: NonCancelableCustomEvent<SelectionChangeDetail<unknown>>) => void",
        "references": {
          "NonCancelableCustomEvent": {
            "location": "import",
            "path": "@awsui/components-react"
          },
          "TableProps": {
            "location": "import",
            "path": "@awsui/components-react/table"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "onExpandChildren": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "(node: ITreeNode<any>) => void",
        "resolved": "(node: any) => void",
        "references": {
          "ITreeNode": {
            "location": "import",
            "path": "@iot-app-kit/related-table"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    },
    "onSortingChange": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "(\n    event: NonCancelableCustomEvent<TableProps.SortingState<unknown>>\n  ) => void",
        "resolved": "(event: NonCancelableCustomEvent<SortingState<unknown>>) => void",
        "references": {
          "NonCancelableCustomEvent": {
            "location": "import",
            "path": "@awsui/components-react"
          },
          "TableProps": {
            "location": "import",
            "path": "@awsui/components-react/table"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      }
    }
  }; }
  static get states() { return {
    "selectedItems": {}
  }; }
  static get elementRef() { return "host"; }
}
